<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Input Data for SNaQ · PhyloNetworks.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="PhyloNetworks.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">PhyloNetworks.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../installation/">Installation</a></li><li class="is-active"><a class="tocitem" href>Input Data for SNaQ</a><ul class="internal"><li><a class="tocitem" href="#Tutorial-data:-gene-trees"><span>Tutorial data: gene trees</span></a></li><li><a class="tocitem" href="#Tutorial-data:-quartet-CFs"><span>Tutorial data: quartet CFs</span></a></li><li><a class="tocitem" href="#Tutorial-data:-starting-tree"><span>Tutorial data: starting tree</span></a></li></ul></li><li><a class="tocitem" href="../ticr_howtogetQuartetCFs/">TICR pipeline</a></li><li><a class="tocitem" href="../snaq_plot/">Network estimation and display</a></li><li><a class="tocitem" href="../dist_reroot/">Network comparison and manipulation</a></li><li><a class="tocitem" href="../fixednetworkoptim/">Candidate Networks</a></li><li><a class="tocitem" href="../expectedCFs/">Extract Expected CFs</a></li><li><a class="tocitem" href="../bootstrap/">Bootstrap</a></li><li><a class="tocitem" href="../multiplealleles/">Multiple Alleles</a></li><li><a class="tocitem" href="../trait_tree/">Continuous Trait Evolution</a></li><li><a class="tocitem" href="../parsimony/">Parsimony on networks</a></li><li><a class="tocitem" href="../fitDiscrete/">Discrete Trait Evolution</a></li><li><a class="tocitem" href="../nj/">Neighbour Joining</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/public/">Public</a></li><li><a class="tocitem" href="../../lib/internals/">Internals</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Input Data for SNaQ</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Input Data for SNaQ</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/crsl4/PhyloNetworks.jl/blob/master/docs/src/man/inputdata.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Input-for-SNaQ"><a class="docs-heading-anchor" href="#Input-for-SNaQ">Input for SNaQ</a><a id="Input-for-SNaQ-1"></a><a class="docs-heading-anchor-permalink" href="#Input-for-SNaQ" title="Permalink"></a></h1><p>SNaQ is a method implemented in the package to estimate a phylogenetic network from multiple molecular sequence alignments. There are two alternatives for the input data:</p><ol><li>A list of estimated gene trees for each locus, which can be obtained using <a href="http://mrbayes.sourceforge.net">MrBayes</a> or <a href="http://sco.h-its.org/exelixis/software.html">RAxML</a>. Or:</li><li>A table of concordance factors (CF), i.e. gene tree frequencies, for each 4-taxon subset. This table can be obtained from <a href="http://www.stat.wisc.edu/~ane/bucky/">BUCKy</a>, to account for gene tree uncertainty</li></ol><p>This <a href="https://github.com/nstenz/TICR">pipeline</a> can be used to obtain the table of quartet CF needed as input for SNaQ (see also the <a href="https://github.com/crsl4/PhyloNetworks.jl/wiki/TICR:-from-alignments-to-quartet-concordance-factors">wiki</a>.) It starts from the sequence alignments, runs MrBayes and then BUCKy (both parallelized), producing the table of estimated CFs and their credibility intervals. Additional details on this <a href="../ticr_howtogetQuartetCFs/#TICR-pipeline">TICR pipeline</a> describe how to insert data at various stages (e.g. after running MrBayes on each locus).</p><h2 id="Tutorial-data:-gene-trees"><a class="docs-heading-anchor" href="#Tutorial-data:-gene-trees">Tutorial data: gene trees</a><a id="Tutorial-data:-gene-trees-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial-data:-gene-trees" title="Permalink"></a></h2><p>We suggest that you create a special directory for running these examples, where input files can be downloaded and where output files will be created (with estimated networks for instance). Enter this directory and run Julia from there.</p><p>Suppose you have a file with a list of gene trees in parenthetical format called <code>raxmltrees.tre</code>. You can access the example file of input trees <a href="https://github.com/crsl4/PhyloNetworks/blob/master/examples/raxmltrees.tre">here</a> or <a href="https://raw.githubusercontent.com/crsl4/PhyloNetworks/master/examples/raxmltrees.tre">here</a> for easier download.</p><p>Do not copy-paste into a &quot;smart&quot; text-editor. Instead, save the file directly into your working directory using &quot;save link as&quot; or &quot;download linked file as&quot;. This file contains 30 gene trees, each in parenthetical format on 6 taxa like this (with rounded branch lengths):</p><p><code>(E:0.038,((A:0.014,B:0.010):0.010,(C:0.008,D:0.002):0.010):0.025,O:0.078);</code></p><p>If <code>raxmltrees.tre</code> is in your working directory, you can view its content within Julia:</p><pre><code class="language-julia">less(&quot;raxmltrees.tre&quot;)</code></pre><p>or like this, to view the version downloaded with the package:</p><pre><code class="language-julia">raxmltrees = joinpath(dirname(pathof(PhyloNetworks)), &quot;..&quot;,&quot;examples&quot;,&quot;raxmltrees.tre&quot;)
less(raxmltrees)</code></pre><p>Just type <code>q</code> to quit viewing this file. You could read in these 30 trees and visualize the third one (say) like this:</p><pre><code class="language-julia">using PhyloNetworks
raxmltrees = joinpath(dirname(pathof(PhyloNetworks)), &quot;..&quot;,&quot;examples&quot;,&quot;raxmltrees.tre&quot;);</code></pre><pre><code class="language-julia-repl">julia&gt; genetrees = readMultiTopology(raxmltrees);

julia&gt; genetrees[3]
HybridNetwork, Rooted Network
9 edges
10 nodes: 6 tips, 0 hybrid nodes, 4 internal tree nodes.
tip labels: E, A, B, C, ...
((E:0.015,(A:0.006,B:0.006):0.003):0.041,(C:0.006,D:0.0):0.041,O:0.052);</code></pre><p>To visualize any of these input trees, use the <a href="https://github.com/cecileane/PhyloPlots.jl">PhyloPlots</a> package:</p><pre><code class="language-julia">using PhyloPlots
plot(genetrees[3], :R); # tree for 3rd gene</code></pre><p><img src="../../assets/figures/inputdata_gene3.svg" alt="gene3"/></p><p>To read in all gene trees and directly summarize them by a list of quartet CFs (proportion of input trees with a given quartet):</p><pre><code class="language-julia-repl">julia&gt; q,t = countquartetsintrees(genetrees); # read in trees, calculate quartet CFs
Reading in trees, looking at 15 quartets in each...
0+------------------------------+100%
  ******************************

julia&gt; df = writeTableCF(q,t)   # data frame with observed CFs: gene frequencies
15×8 DataFrame. Omitted printing of 1 columns
│ Row │ t1     │ t2     │ t3     │ t4     │ CF12_34  │ CF13_24   │ CF14_23   │
│     │ String │ String │ String │ String │ Float64  │ Float64   │ Float64   │
├─────┼────────┼────────┼────────┼────────┼──────────┼───────────┼───────────┤
│ 1   │ A      │ B      │ C      │ D      │ 1.0      │ 0.0       │ 0.0       │
│ 2   │ A      │ B      │ C      │ E      │ 0.833333 │ 0.0333333 │ 0.133333  │
│ 3   │ A      │ B      │ D      │ E      │ 0.833333 │ 0.0333333 │ 0.133333  │
│ 4   │ A      │ C      │ D      │ E      │ 0.0      │ 0.0       │ 1.0       │
│ 5   │ B      │ C      │ D      │ E      │ 0.0      │ 0.0       │ 1.0       │
│ 6   │ A      │ B      │ C      │ O      │ 0.866667 │ 0.0       │ 0.133333  │
│ 7   │ A      │ B      │ D      │ O      │ 0.866667 │ 0.0       │ 0.133333  │
│ 8   │ A      │ C      │ D      │ O      │ 0.0      │ 0.0       │ 1.0       │
│ 9   │ B      │ C      │ D      │ O      │ 0.0      │ 0.0       │ 1.0       │
│ 10  │ A      │ B      │ E      │ O      │ 0.833333 │ 0.0666667 │ 0.1       │
│ 11  │ A      │ C      │ E      │ O      │ 0.533333 │ 0.333333  │ 0.133333  │
│ 12  │ B      │ C      │ E      │ O      │ 0.666667 │ 0.266667  │ 0.0666667 │
│ 13  │ A      │ D      │ E      │ O      │ 0.533333 │ 0.333333  │ 0.133333  │
│ 14  │ B      │ D      │ E      │ O      │ 0.666667 │ 0.266667  │ 0.0666667 │
│ 15  │ C      │ D      │ E      │ O      │ 1.0      │ 0.0       │ 0.0       │

julia&gt; using CSV

julia&gt; CSV.write(&quot;tableCF.csv&quot;, df); # to save the data frame to a file

julia&gt; raxmlCF = readTableCF(&quot;tableCF.csv&quot;) # read in the file and produces a &quot;DataCF&quot; object
30.0 gene trees per 4-taxon set
Object DataCF
number of quartets: 15</code></pre><p><code>less(&quot;tableCF.csv&quot;)</code> lets you see the content of the newly created file &quot;tableCF.csv&quot;, within Julia. Again, type <code>q</code> to quit viewing this file.</p><p>In this table, each 4-taxon set is listed in one row. The 3 &quot;CF&quot; columns gives the proportion of genes that has each of the 3 possible trees on these 4 taxa.</p><p>For more help on any function, type <code>?</code> to enter the help mode, then type the name of the function. For example: type <code>?</code> then <code>countquartetsintrees</code> for information on the various options of that function.</p><p>When there are many more taxa, the number of quartets might be very large and we might want to use a subset to speed things up. Here, if we wanted to use a random sample of 10 quartets instead of all quartets, we could do:</p><p><code>raxmlCF = readTrees2CF(raxmltrees, whichQ=&quot;rand&quot;, numQ=10, CFfile=&quot;tableCF10.txt&quot;)</code></p><p>Be careful to use a numQ value smaller than the total number of possible 4-taxon subsets, which is <em>n choose 4</em> on <em>n</em> taxa (e.g. 15 on 6 taxa). To get a predictable random sample, you may set the seed with <code>using Random; Random.seed!(12321)</code> (for instance) prior to sampling the quartets as above. The <code>readTrees2CF</code> is <em>much</em> slower than the function <code>countquartetsintrees</code> to read in trees and calculate the quartet CFs observed in the trees, when we want to get <em>all</em> quartet CFs. But for a small sample of quartets, then <code>readTrees2CF</code> is available.</p><h2 id="Tutorial-data:-quartet-CFs"><a class="docs-heading-anchor" href="#Tutorial-data:-quartet-CFs">Tutorial data: quartet CFs</a><a id="Tutorial-data:-quartet-CFs-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial-data:-quartet-CFs" title="Permalink"></a></h2><p>If we already have a table of quartet concordance factor (CF) values in a file <code>buckyCF.csv</code> in this format</p><table><tr><th style="text-align: left">Taxon1</th><th style="text-align: left">Taxon2</th><th style="text-align: left">Taxon3</th><th style="text-align: left">Taxon4</th><th style="text-align: left">CF12_34</th><th style="text-align: left">CF13_24</th><th style="text-align: left">CF14_23</th></tr><tr><td style="text-align: left">D</td><td style="text-align: left">A</td><td style="text-align: left">E</td><td style="text-align: left">O</td><td style="text-align: left">0.565</td><td style="text-align: left">0.0903</td><td style="text-align: left">0.3447</td></tr><tr><td style="text-align: left">...</td><td style="text-align: left"></td><td style="text-align: left"></td><td style="text-align: left"></td><td style="text-align: left"></td><td style="text-align: left"></td><td style="text-align: left">...</td></tr></table><p>we would read it in one step like this: <code>readTableCF(&quot;buckyCF.csv&quot;)</code>. An example file comes with the package, available <a href="https://github.com/crsl4/PhyloNetworks/blob/master/examples/buckyCF.csv">here</a> or <a href="https://raw.githubusercontent.com/crsl4/PhyloNetworks/master/examples/buckyCF.csv">here</a>.</p><pre><code class="language-julia-repl">julia&gt; buckyCFfile = joinpath(dirname(pathof(PhyloNetworks)), &quot;..&quot;,&quot;examples&quot;,&quot;buckyCF.csv&quot;);

julia&gt; buckyCF = readTableCF(buckyCFfile)
30.0 gene trees per 4-taxon set
Object DataCF
number of quartets: 15</code></pre><p>The same thing could be done in 2 steps: first to read the file and convert it to a &#39;DataFrame&#39; object, and then to convert this DataFrame into a DataCF object.</p><pre><code class="language-julia-repl">julia&gt; using CSV, DataFrames

julia&gt; dat = DataFrame!(CSV.File(buckyCFfile));

julia&gt; first(dat, 6) # to see the first 6 rows
6×14 DataFrame. Omitted printing of 8 columns
│ Row │ taxon1 │ taxon2 │ taxon3 │ taxon4 │ CF12.34    │ CF12.34_lo │
│     │ String │ String │ String │ String │ Float64    │ Float64    │
├─────┼────────┼────────┼────────┼────────┼────────────┼────────────┤
│ 1   │ D      │ A      │ E      │ O      │ 0.565033   │ 0.5        │
│ 2   │ D      │ A      │ E      │ B      │ 0.0005     │ 0.0        │
│ 3   │ C      │ A      │ E      │ B      │ 0.0005     │ 0.0        │
│ 4   │ C      │ A      │ E      │ O      │ 0.565033   │ 0.5        │
│ 5   │ D      │ A      │ O      │ B      │ 3.33333e-5 │ 0.0        │
│ 6   │ C      │ D      │ E      │ O      │ 0.999867   │ 1.0        │

julia&gt; buckyCF = readTableCF(dat)
30.0 gene trees per 4-taxon set
Object DataCF
number of quartets: 15

julia&gt; writeTableCF(buckyCF)
15×8 DataFrame. Omitted printing of 2 columns
│ Row │ t1     │ t2     │ t3     │ t4     │ CF12_34    │ CF13_24    │
│     │ String │ String │ String │ String │ Float64    │ Float64    │
├─────┼────────┼────────┼────────┼────────┼────────────┼────────────┤
│ 1   │ D      │ A      │ E      │ O      │ 0.565033   │ 0.0903     │
│ 2   │ D      │ A      │ E      │ B      │ 0.0005     │ 0.8599     │
│ 3   │ C      │ A      │ E      │ B      │ 0.0005     │ 0.8599     │
│ 4   │ C      │ A      │ E      │ O      │ 0.565033   │ 0.0903     │
│ 5   │ D      │ A      │ O      │ B      │ 3.33333e-5 │ 0.8885     │
│ 6   │ C      │ D      │ E      │ O      │ 0.999867   │ 6.66667e-5 │
│ 7   │ C      │ E      │ O      │ B      │ 0.0401667  │ 0.263067   │
│ 8   │ C      │ D      │ A      │ E      │ 1.0        │ 0.0        │
│ 9   │ C      │ D      │ O      │ B      │ 0.999867   │ 6.66667e-5 │
│ 10  │ C      │ D      │ E      │ B      │ 1.0        │ 0.0        │
│ 11  │ D      │ E      │ O      │ B      │ 0.0401667  │ 0.263067   │
│ 12  │ C      │ D      │ A      │ O      │ 0.999867   │ 6.66667e-5 │
│ 13  │ C      │ D      │ A      │ B      │ 1.0        │ 0.0        │
│ 14  │ A      │ E      │ O      │ B      │ 0.0731667  │ 0.0424667  │
│ 15  │ C      │ A      │ O      │ B      │ 3.33333e-5 │ 0.8885     │</code></pre><p>In the input file, columns need to be in the right order: with the first 4 columns giving the names of the taxa in each 4-taxon set. The CF values are assumed to be in columns named &quot;CF12_34&quot;, etc., or else in columns 5,6,7. If available, a column named &quot;ngenes&quot; will be taken to have the the number of genes for each 4-taxon subset.</p><h2 id="Tutorial-data:-starting-tree"><a class="docs-heading-anchor" href="#Tutorial-data:-starting-tree">Tutorial data: starting tree</a><a id="Tutorial-data:-starting-tree-1"></a><a class="docs-heading-anchor-permalink" href="#Tutorial-data:-starting-tree" title="Permalink"></a></h2><p>If we have a tree for the data set at hand, it can be used as a starting point for the optimization. From our gene trees, we estimated a species tree with <a href="https://github.com/smirarab/ASTRAL/blob/master/astral-tutorial.md">ASTRAL</a>. This tree comes with the package in file <code>astral.tre</code> <a href="https://github.com/crsl4/PhyloNetworks/blob/master/examples/astral.tre">here</a>. This file has 102 trees: 100 bootstrap species trees, followed by their greedy consensus, followed by the best tree on the original data. It&#39;s this last tree that we are most interested in. We can read it with</p><pre><code class="language-julia">astralfile = joinpath(dirname(pathof(PhyloNetworks)), &quot;..&quot;,&quot;examples&quot;,&quot;astral.tre&quot;);
astraltree = readMultiTopology(astralfile)[102] # 102th tree: last tree here
plot(astraltree, :R, showEdgeLength=true);</code></pre><p><img src="../../assets/figures/inputdata_astraltree.svg" alt="astraltree"/></p><p>To start its search, SNaQ will need a network of &quot;level 1&quot;. All trees and all networks with 1 hybridization are of level 1. To make sure that a network with 2 or more hybridizations is of level 1, we can read it in with <code>readTopologyLevel1</code> (which also unroots the tree, resolves polytomies, replaces missing branch lengths by 1 for starting values etc.):</p><pre><code class="language-julia">T=readTopologyLevel1(&quot;startNetwork.txt&quot;)</code></pre><p>(here <code>startNetwork.txt</code> is a hypothetical file: replace this by the name of a file that contains your network of interest.)</p><p>Next: <a href="../snaq_plot/#Getting-a-Network">Getting a Network</a></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../installation/">« Installation</a><a class="docs-footer-nextpage" href="../ticr_howtogetQuartetCFs/">TICR pipeline »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Monday 12 October 2020 02:12">Monday 12 October 2020</span>. Using Julia version 1.5.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
