<!DOCTYPE html>
<!--[if lt IE 7 ]><html class="no-js ie6"><![endif]-->
<!--[if IE 7 ]><html class="no-js ie7"><![endif]-->
<!--[if IE 8 ]><html class="no-js ie8"><![endif]-->
<!--[if IE 9 ]><html class="no-js ie9"><![endif]-->
<!--[if (gt IE 9)|!(IE)]><!--> <html class="no-js"> <!--<![endif]-->
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,user-scalable=no,initial-scale=1,maximum-scale=1">
    
      
        <title>Input Data for SNaQ - PhyloNetworks.jl</title>
      
      
      
      
        <meta name="author" content="Claudia Sol&iacute;s-Lemus">
      
    
    <meta property="og:url" content="None">
    <meta property="og:title" content="PhyloNetworks.jl">
    <meta property="og:image" content="None/../../snaq_small.png">
    <meta name="apple-mobile-web-app-title" content="PhyloNetworks.jl">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    
      <link rel="apple-touch-icon" href="../../snaq_small.png">
    
    
    <link rel="shortcut icon" type="image/x-icon" href="../../assets/images/favicon-e565ddfa3b.ico">
    <link rel="icon" type="image/x-icon" href="../../assets/images/favicon-e565ddfa3b.ico">
    <style>
      @font-face {
      	font-family: 'Icon';
      	src: url('../../assets/fonts/icon.eot?52m981');
      	src: url('../../assets/fonts/icon.eot?#iefix52m981')
               format('embedded-opentype'),
      		   url('../../assets/fonts/icon.woff?52m981')
               format('woff'),
      		   url('../../assets/fonts/icon.ttf?52m981')
               format('truetype'),
      		   url('../../assets/fonts/icon.svg?52m981#icon')
               format('svg');
      	font-weight: normal;
      	font-style: normal;
      }
    </style>
    <link rel="stylesheet" href="../../assets/stylesheets/application-a422ff04cc.css">
    
      <link rel="stylesheet" href="../../assets/stylesheets/palettes-05ab2406df.css">
    
    
      
      
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Ubuntu:400,700|Ubuntu+Mono">
      <style>
        body, input {
          font-family: 'Ubuntu', Helvetica, Arial, sans-serif;
        }
        pre, code {
          font-family: 'Ubuntu Mono', 'Courier New', 'Courier', monospace;
        }
      </style>
    
    
      <link rel="stylesheet" href="../../assets/Documenter.css">
    
    <script src="../../assets/javascripts/modernizr-4ab42b99fd.js"></script>
    
  </head>
  
  
  
  <body class="palette-primary-teal palette-accent-teal">
    
      
      
    
    <div class="backdrop">
      <div class="backdrop-paper"></div>
    </div>
    <input class="toggle" type="checkbox" id="toggle-drawer">
    <input class="toggle" type="checkbox" id="toggle-search">
    <label class="toggle-button overlay" for="toggle-drawer"></label>
    <header class="header">
      <nav aria-label="Header">
  <div class="bar default">
    <div class="button button-menu" role="button" aria-label="Menu">
      <label class="toggle-button icon icon-menu" for="toggle-drawer">
        <span></span>
      </label>
    </div>
    <div class="stretch">
      <div class="title">
        
          <span class="path">
            
              
                Manual <i class="icon icon-link"></i>
              
            
          </span>
        
        Input Data for SNaQ
      </div>
    </div>
    
    
    <div class="button button-search" role="button" aria-label="Search">
      <label class="toggle-button icon icon-search" title="Search" for="toggle-search"></label>
    </div>
  </div>
  <div class="bar search">
    <div class="button button-close" role="button" aria-label="Close">
      <label class="toggle-button icon icon-back" for="toggle-search"></label>
    </div>
    <div class="stretch">
      <div class="field">
        <input class="query" type="text" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck>
      </div>
    </div>
    <div class="button button-reset" role="button" aria-label="Search">
      <button class="toggle-button icon icon-close" id="reset-search"></button>
    </div>
  </div>
</nav>
    </header>
    <main class="main">
      
      <div class="drawer">
        <nav aria-label="Navigation">
  
  <a href="https://github.com/crsl4/PhyloNetworks.jl" class="project">
    <div class="banner">
      
        <div class="logo">
          <img src="../../snaq_small.png">
        </div>
      
      <div class="name">
        <strong>
          PhyloNetworks.jl
          <span class="version">
            
          </span>
        </strong>
        
          <br>
          crsl4/PhyloNetworks.jl
        
      </div>
    </div>
  </a>
  <div class="scrollable">
    <div class="wrapper">
      
        <ul class="repo">
          <li class="repo-download">
            
            <a href="https://github.com/crsl4/PhyloNetworks.jl/archive/master.zip" target="_blank" title="Download" data-action="download">
              <i class="icon icon-download"></i> Download
            </a>
          </li>
          <li class="repo-stars">
            <a href="https://github.com/crsl4/PhyloNetworks.jl/stargazers" target="_blank" title="Stargazers" data-action="star">
              <i class="icon icon-star"></i> Stars
              <span class="count">&ndash;</span>
            </a>
          </li>
        </ul>
        <hr>
      
      <div class="toc">
        <ul>
          
            
  <li>
    <a class="" title="Home" href="../..">
      Home
    </a>
    
  </li>

          
            
  <li>
    <span class="section">Manual</span>
    <ul>
      
        
  <li>
    <a class="" title="Installation" href="../installation/">
      Installation
    </a>
    
  </li>

      
        
  <li>
    <a class="current" title="Input Data for SNaQ" href="./">
      Input Data for SNaQ
    </a>
    
      
        
      
      
        <ul>
          
            <li class="anchor">
              <a title="Tutorial data: gene trees" href="#tutorial-data-gene-trees">
                Tutorial data: gene trees
              </a>
            </li>
          
            <li class="anchor">
              <a title="Tutorial data: quartet CFs" href="#tutorial-data-quartet-cfs">
                Tutorial data: quartet CFs
              </a>
            </li>
          
        </ul>
      
    
  </li>

      
        
  <li>
    <a class="" title="TICR pipeline" href="../ticr_howtogetQuartetCFs/">
      TICR pipeline
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Network estimation and display" href="../snaq_plot/">
      Network estimation and display
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Fixed Network" href="../fixednetworkoptim/">
      Fixed Network
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Extract Expected CFs" href="../expectedCFs/">
      Extract Expected CFs
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Bootstrap" href="../bootstrap/">
      Bootstrap
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Multiple Alleles" href="../multiplealleles/">
      Multiple Alleles
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Trait Evolution" href="../trait_tree/">
      Trait Evolution
    </a>
    
  </li>

      
    </ul>
  </li>

          
            
  <li>
    <span class="section">Library</span>
    <ul>
      
        
  <li>
    <a class="" title="Public" href="../../lib/public/">
      Public
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Internals" href="../../lib/internals/">
      Internals
    </a>
    
  </li>

      
    </ul>
  </li>

          
        </ul>
        
      </div>
    </div>
  </div>
</nav>
      </div>
      <article class="article">
        <div class="wrapper">
          
          <p><a id='Input-for-SNaQ-1'></a></p>
<h1 id="input-for-snaq">Input for SNaQ</h1>
<p>SNaQ is a method implemented in the package to estimate a phylogenetic network from multiple molecular sequence alignments. There are two alternatives for the input data:</p>
<ol>
<li>A list of estimated gene trees for each locus, which can be obtained using</li>
</ol>
<p><a href="http://mrbayes.sourceforge.net">MrBayes</a> or <a href="http://sco.h-its.org/exelixis/software.html">RAxML</a>. Or:</p>
<ol>
<li>A table of concordance factors (CF) for each 4-taxon subset which can be</li>
</ol>
<p>obtained from <a href="http://www.stat.wisc.edu/~ane/bucky/">BUCKy</a>,   to account for gene tree uncertainty</p>
<p>This <a href="https://github.com/nstenz/TICR">pipeline</a> can be used to obtain the table of  quartet CF needed as input for SNaQ. The pipeline starts from the sequence alignments, runs MrBayes and then BUCKy (both parallelized), producing the table of estimated CFs and their credibility intervals. Additional details on this <a href="../ticr_howtogetQuartetCFs/#TICR-pipeline-1">TICR pipeline</a> describe how to insert data at various stages (e.g. after running MrBayes on each locus).</p>
<p><a id='Tutorial-data:-gene-trees-1'></a></p>
<h2 id="tutorial-data-gene-trees">Tutorial data: gene trees</h2>
<!–The examples files for this section can be found within the PhyloNetworks folder, typically in your *HOME/.julia/v0.4/PhyloNetworks/examples/*. However, links to the files are also included below. –> We suggest that you create a special directory for running these examples, where input files can be downloaded and where output files will be created (with estimated networks for instance). Enter this directory and run Julia from there.

<p>Suppose you have a file with a list of gene trees in parenthetical format called <em>treefile.txt</em>. You can access the example file of input trees <a href="https://github.com/crsl4/PhyloNetworks/blob/master/examples/treefile.txt">here</a> or <a href="https://raw.githubusercontent.com/crsl4/PhyloNetworks/master/examples/treefile.txt">here</a> for easier download.</p>
<p>Do not copy-paste into a "smart" text-editor. Instead, save the file directly into your working directory using "save link as" or "download linked file as". This file contains 10 trees, each in parenthetical format on 6 taxa like this:</p>
<p>(6:2.728,((3:0.655,5:0.655):1.202,(1:0.881,(2:0.783,4:0.783):0.098):0.976):0.871);</p>
<p>If 'treefile.txt' is in your working directory, you can view its content within Julia:</p>
<pre><code class="julia">less(&quot;treefile.txt&quot;)
</code></pre>

<p>Just type <code>q</code> to quit viewing this file. You could read in these 10 trees and visualize the third one (say) like this:</p>
<pre><code class="julia">tentrees = readMultiTopology(&quot;treefile.txt&quot;)
tentrees[3]
plot(tentrees[3])
</code></pre>

<p>To read in all gene trees and directly summarize them by a list of quartet CFs (proportion of input trees with a given quartet):</p>
<pre><code class="julia">d=readTrees2CF(&quot;treefile.txt&quot;, CFfile=&quot;tableCFall.txt&quot;)
less(&quot;tableCFall.txt&quot;)
</code></pre>

<p><code>less("tableCFall.txt")</code> lets you see the content of the newly created file "tableCFall.txt", within Julia. Again, type <code>q</code> to quit viewing this file.</p>
<p>If instead of all the 4-taxon subsets, you just want to use a random sample of 10 4-taxon subsets:</p>
<pre><code class="julia">d=readTrees2CF(&quot;treefile.txt&quot;, whichQ=&quot;rand&quot;, numQ=10, CFfile=&quot;tableCF10.txt&quot;)
</code></pre>

<p>Be careful to use a numQ value smaller than the total number of possible 4-taxon subsets, which is <em>n choose 4</em> on <em>n</em> taxa (e.g. 15 on 6 taxa). To get a predictable random sample, you may set the seed with <code>srand(12321)</code> (for instance) prior to sampling the quartets as above.</p>
<p><a id='Tutorial-data:-quartet-CFs-1'></a></p>
<h2 id="tutorial-data-quartet-cfs">Tutorial data: quartet CFs</h2>
<p>If you already have a table of quartet concordance factor (CF) values in a file <em>tableCF.txt</em> in this format</p>
<table>
<thead>
<tr>
<th align="left">Taxon1</th>
<th align="left">Taxon2</th>
<th align="left">Taxon3</th>
<th align="left">Taxon4</th>
<th align="left">CF12_34</th>
<th align="left">CF13_24</th>
<th align="left">CF14_23</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">A</td>
<td align="left">B</td>
<td align="left">E</td>
<td align="left">D</td>
<td align="left">0.4</td>
<td align="left">0.6</td>
<td align="left">0.0</td>
</tr>
</tbody>
</table>
<p>you would read it like this:</p>
<pre><code class="julia">d=readTableCF(&quot;tableCF.txt&quot;) # one step only: read file and convert to 'DataCF' object
# or in 2 steps:
using DataFrames
dat = readtable(&quot;tableCF.txt&quot;) # read file into a 'DataFrame' object
d = readTableCF(dat)           # convert DataFrame into DataCF
</code></pre>

<p>You can access this example file <a href="https://github.com/crsl4/PhyloNetworks/blob/master/examples/tableCF.txt">here</a> or <a href="https://raw.githubusercontent.com/crsl4/PhyloNetworks/master/examples/tableCF.txt">here</a>.</p>
<p>Columns need to be in the right order. If you have the information on the number of genes used for each 4-taxon subset, you can add this information in a column named "ngenes".</p>
<p>If you have a tree <em>startTree.txt</em> in parenthetical format to use as starting point for the optimization, you can read it with</p>
<pre><code class="julia">T=readTopology(&quot;startTree.txt&quot;)
</code></pre>

<p>You can access this example file <a href="https://github.com/crsl4/PhyloNetworks/blob/master/examples/startTree.txt">here</a> (raw file <a href="https://raw.githubusercontent.com/crsl4/PhyloNetworks/master/examples/startTree.txt">here</a>).</p>
<p>If the topology <code>T</code> is to be used as a starting tree for SNaQ, it needs to be of level 1. To make sure that it is, you may read it in with <code>readTopologyLevel1</code> (which may also unroot the tree, resolve polytomies, replace missing branch lengths by 1 for starting values etc.):</p>
<pre><code class="julia">T=readTopologyLevel1(&quot;startTree.txt&quot;)
</code></pre>

<p>Note that all trees and all networks with 1 hybridization are of level 1.</p>
<p>Next: <a href="../snaq_plot/#Getting-a-Network-1">Getting a Network</a></p>
          <aside class="copyright" role="note">
            
            Documentation built with
            <a href="http://www.mkdocs.org" target="_blank">MkDocs</a>
            using the
            <a href="http://squidfunk.github.io/mkdocs-material/" target="_blank">
              Material
            </a>
            theme.
          </aside>
          
            <footer class="footer">
              
  <nav class="pagination" aria-label="Footer">
    <div class="previous">
      
        <a href="../installation/" title="Installation">
          <span class="direction">
            Previous
          </span>
          <div class="page">
            <div class="button button-previous" role="button" aria-label="Previous">
              <i class="icon icon-back"></i>
            </div>
            <div class="stretch">
              <div class="title">
                Installation
              </div>
            </div>
          </div>
        </a>
      
    </div>
    <div class="next">
      
        <a href="../ticr_howtogetQuartetCFs/" title="TICR pipeline">
          <span class="direction">
            Next
          </span>
          <div class="page">
            <div class="stretch">
              <div class="title">
                TICR pipeline
              </div>
            </div>
            <div class="button button-next" role="button" aria-label="Next">
              <i class="icon icon-forward"></i>
            </div>
          </div>
        </a>
      
    </div>
  </nav>

            </footer>
          
        </div>
      </article>
      <div class="results" role="status" aria-live="polite">
        <div class="scrollable">
          <div class="wrapper">
            <div class="meta"></div>
            <div class="list"></div>
          </div>
        </div>
      </div>
    </main>
    <script>
      var base_url = '../..';
      var repo_id  = 'crsl4/PhyloNetworks.jl';
    </script>
    <script src="../../assets/javascripts/application-997097ee0c.js"></script>
    
      <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
      <script src="../../assets/mathjaxhelper.js"></script>
    
    
  </body>
</html>