<!DOCTYPE html>
<!--[if lt IE 7 ]><html class="no-js ie6"><![endif]-->
<!--[if IE 7 ]><html class="no-js ie7"><![endif]-->
<!--[if IE 8 ]><html class="no-js ie8"><![endif]-->
<!--[if IE 9 ]><html class="no-js ie9"><![endif]-->
<!--[if (gt IE 9)|!(IE)]><!--> <html class="no-js"> <!--<![endif]-->
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,user-scalable=no,initial-scale=1,maximum-scale=1">
    
      
        <title>TICR pipeline - PhyloNetworks.jl</title>
      
      
      
      
        <meta name="author" content="Claudia Sol&iacute;s-Lemus">
      
    
    <meta property="og:url" content="None">
    <meta property="og:title" content="PhyloNetworks.jl">
    <meta property="og:image" content="None/../../snaq_small.png">
    <meta name="apple-mobile-web-app-title" content="PhyloNetworks.jl">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    
      <link rel="apple-touch-icon" href="../../snaq_small.png">
    
    
    <link rel="shortcut icon" type="image/x-icon" href="../../None">
    <link rel="icon" type="image/x-icon" href="../../None">
    <style>
      @font-face {
      	font-family: 'Icon';
      	src: url('../../assets/fonts/icon.eot?52m981');
      	src: url('../../assets/fonts/icon.eot?#iefix52m981')
               format('embedded-opentype'),
      		   url('../../assets/fonts/icon.woff?52m981')
               format('woff'),
      		   url('../../assets/fonts/icon.ttf?52m981')
               format('truetype'),
      		   url('../../assets/fonts/icon.svg?52m981#icon')
               format('svg');
      	font-weight: normal;
      	font-style: normal;
      }
    </style>
    <link rel="stylesheet" href="../../assets/stylesheets/application-0dac3e5884.css">
    
      <link rel="stylesheet" href="../../assets/stylesheets/palettes-05ab2406df.css">
    
    
      
      
      
      <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Ubuntu:400,700|Ubuntu+Mono">
      <style>
        body, input {
          font-family: 'Ubuntu', Helvetica, Arial, sans-serif;
        }
        pre, code {
          font-family: 'Ubuntu Mono', 'Courier New', 'Courier', monospace;
        }
      </style>
    
    
      <link rel="stylesheet" href="../../assets/Documenter.css">
    
    <script src="../../assets/javascripts/modernizr-4ab42b99fd.js"></script>
    
  </head>
  
  
  
  <body class="palette-primary-teal palette-accent-teal">
    
      
      
    
    <div class="backdrop">
      <div class="backdrop-paper"></div>
    </div>
    <input class="toggle" type="checkbox" id="toggle-drawer">
    <input class="toggle" type="checkbox" id="toggle-search">
    <label class="toggle-button overlay" for="toggle-drawer"></label>
    <header class="header">
      <nav aria-label="Header">
  <div class="bar default">
    <div class="button button-menu" role="button" aria-label="Menu">
      <label class="toggle-button icon icon-menu" for="toggle-drawer">
        <span></span>
      </label>
    </div>
    <div class="stretch">
      <div class="title">
        
          <span class="path">
            
              
                Manual <i class="icon icon-link"></i>
              
            
          </span>
        
        TICR pipeline
      </div>
    </div>
    
    
    <div class="button button-search" role="button" aria-label="Search">
      <label class="toggle-button icon icon-search" title="Search" for="toggle-search"></label>
    </div>
  </div>
  <div class="bar search">
    <div class="button button-close" role="button" aria-label="Close">
      <label class="toggle-button icon icon-back" for="toggle-search"></label>
    </div>
    <div class="stretch">
      <div class="field">
        <input class="query" type="text" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck>
      </div>
    </div>
    <div class="button button-reset" role="button" aria-label="Search">
      <button class="toggle-button icon icon-close" id="reset-search"></button>
    </div>
  </div>
</nav>
    </header>
    <main class="main">
      
      <div class="drawer">
        <nav aria-label="Navigation">
  
  <a href="https://github.com/crsl4/PhyloNetworks.jl" class="project">
    <div class="banner">
      
        <div class="logo">
          <img src="../../snaq_small.png">
        </div>
      
      <div class="name">
        <strong>
          PhyloNetworks.jl
          <span class="version">
            
          </span>
        </strong>
        
          <br>
          crsl4/PhyloNetworks.jl
        
      </div>
    </div>
  </a>
  <div class="scrollable">
    <div class="wrapper">
      
        <ul class="repo">
          <li class="repo-download">
            
            <a href="https://github.com/crsl4/PhyloNetworks.jl/archive/master.zip" target="_blank" title="Download" data-action="download">
              <i class="icon icon-download"></i> Download
            </a>
          </li>
          <li class="repo-stars">
            <a href="https://github.com/crsl4/PhyloNetworks.jl/stargazers" target="_blank" title="Stargazers" data-action="star">
              <i class="icon icon-star"></i> Stars
              <span class="count">&ndash;</span>
            </a>
          </li>
        </ul>
        <hr>
      
      <div class="toc">
        <ul>
          
            
  <li>
    <a class="" title="Home" href="../..">
      Home
    </a>
    
  </li>

          
            
  <li>
    <span class="section">Manual</span>
    <ul>
      
        
  <li>
    <a class="" title="Installation" href="../installation/">
      Installation
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Input Data for SNaQ" href="../inputdata/">
      Input Data for SNaQ
    </a>
    
  </li>

      
        
  <li>
    <a class="current" title="TICR pipeline" href="./">
      TICR pipeline
    </a>
    
      
        
      
      
        <ul>
          
            <li class="anchor">
              <a title="To run MrBayes: You already have alignments" href="#to-run-mrbayes-you-already-have-alignments">
                To run MrBayes: You already have alignments
              </a>
            </li>
          
            <li class="anchor">
              <a title="To run BUCKy: You already have MrBayes output" href="#to-run-bucky-you-already-have-mrbayes-output">
                To run BUCKy: You already have MrBayes output
              </a>
            </li>
          
        </ul>
      
    
  </li>

      
        
  <li>
    <a class="" title="Network estimation and display" href="../snaq_plot/">
      Network estimation and display
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Fixed Network" href="../fixednetworkoptim/">
      Fixed Network
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Extract Expected CFs" href="../expectedCFs/">
      Extract Expected CFs
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Bootstrap" href="../bootstrap/">
      Bootstrap
    </a>
    
  </li>

      
    </ul>
  </li>

          
            
  <li>
    <span class="section">Library</span>
    <ul>
      
        
  <li>
    <a class="" title="Public" href="../../lib/public/">
      Public
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Internals" href="../../lib/internals/">
      Internals
    </a>
    
  </li>

      
    </ul>
  </li>

          
        </ul>
        
      </div>
    </div>
  </div>
</nav>
      </div>
      <article class="article">
        <div class="wrapper">
          
          <p><a id='TICR-pipeline-1'></a></p>
<h1 id="ticr-pipeline">TICR pipeline</h1>
<p>This page provides details to insert data into the TICR pipeline at various stages. Full information and code for this pipeline is <a href="https://github.com/nstenz/TICR">here</a>.</p>
<p><a id='To-run-MrBayes:-You-already-have-alignments-1'></a></p>
<h2 id="to-run-mrbayes-you-already-have-alignments">To run MrBayes: You already have alignments</h2>
<p>If you don't need to run mdl.pl because you already have aligned gene sequences which you would like to run through MrBayes, you can simply create a tarball of the Nexus files (fasta won't work at this stage) you wish to use. This command assumes that you want to use all the files ending with ".nex" in the current directory, one file per locus:</p>
<p><code>tar czf my-genes.tar.gz *.nex</code></p>
<p>Once the tarball has been successfully generated, you can then specify this file as input for <a href="https://github.com/nstenz/TICR/blob/master/scripts/mb.pl">mb.pl</a> assuming you have a a valid MrBayes block located in the file "bayes.txt":</p>
<p><code>mb.pl my-genes.tar.gz -m bayes.txt -o my-genes-mb</code></p>
<p>If you get an error message like <code>mb.pl: Command not found</code>, it might be because <code>mb.pl</code> has no execute permission or your current directory is not in your path. An easy fix is to run this command instead:</p>
<p><code>perl mb.pl my-genes.tar.gz -m bayes.txt -o my-genes-mb</code></p>
<p>The resulting output tarball would now be located in my-genes-mb/my-genes.mb.tar, and can be used normally with <a href="https://github.com/nstenz/TICR/blob/master/scripts/bucky.pl">bucky.pl</a>, that is, like this:</p>
<p><code>bucky.pl my-genes-mb/my-genes.mb.tar -o mygenes-bucky</code></p>
<p>The output, with the table of concordance factors for all sets of 4 taxa, will be in a file named <code>my-genes.CFs.csv</code> inside a directory named <code>mygenes-bucky</code>. That's the file containing the quartet concordance factors to give to SNaQ as input.</p>
<p><a id='To-run-BUCKy:-You-already-have-MrBayes-output-1'></a></p>
<h2 id="to-run-bucky-you-already-have-mrbayes-output">To run BUCKy: You already have MrBayes output</h2>
<p><a id='To-run-mbsum-on-the-output-of-MrBayes-for-each-gene-1'></a></p>
<h3 id="to-run-mbsum-on-the-output-of-mrbayes-for-each-gene">To run mbsum on the output of MrBayes for each gene</h3>
<p>You must now run <code>mbsum</code> separately on each gene's MrBayes output. For a gene with output tree files named gene1.run1.t gene1.run2.t gene1.run3.t, and a desired burnin of 1000 trees per tree file:</p>
<p><code>mbsum -n 1000 -o gene1.in gene1.run1.t gene1.run2.t gene1.run3.t</code></p>
<p>This <code>mbsum</code> command will need to be executed for each gene. Now continue to the next section.</p>
<p><a id='To-run-bucky-on-all-4-taxon-sets:-you-already-have-the-mbsum-output-1'></a></p>
<h3 id="to-run-bucky-on-all-4-taxon-sets-you-already-have-the-mbsum-output">To run bucky on all 4-taxon sets: you already have the mbsum output</h3>
<p>If you have already run mbsum on each individual gene's MrBayes output, you can simply create a tarball containing all the mbsum output files. So if you had mbsum output in files named gene1.in, gene2.in, ... , gene100.in, you would want to run something similar to the following command:</p>
<p><code>tar czf my-genes-mbsum.tar.gz gene*.in</code></p>
<p>You can now use this tarball along with the -s option in <a href="https://github.com/nstenz/TICR/blob/master/scripts/bucky.pl">bucky.pl</a> like this:</p>
<p><code>bucky.pl my-genes-mbsum.tar.gz -s -o mygenes-bucky</code></p>
<p>Again, if you get an error like <code>bucky.pl: Command not found</code>, run instead</p>
<p><code>bucky.pl my-genes-mbsum.tar.gz -s -o mygenes-bucky</code></p>
<p>The output, with the table of concordance factors for all sets of 4 taxa, will be in a file named <code>my-genes.CFs.csv</code> inside directory <code>mygenes-bucky</code>. That's the file containing the quartet concordance factors to give to SNaQ as input.</p>
          <aside class="copyright" role="note">
            
            Documentation built with
            <a href="http://www.mkdocs.org" target="_blank">MkDocs</a>
            using the
            <a href="http://squidfunk.github.io/mkdocs-material/" target="_blank">
              Material
            </a>
            theme.
          </aside>
          
            <footer class="footer">
              
  <nav class="pagination" aria-label="Footer">
    <div class="previous">
      
        <a href="../inputdata/" title="Input Data for SNaQ">
          <span class="direction">
            Previous
          </span>
          <div class="page">
            <div class="button button-previous" role="button" aria-label="Previous">
              <i class="icon icon-back"></i>
            </div>
            <div class="stretch">
              <div class="title">
                Input Data for SNaQ
              </div>
            </div>
          </div>
        </a>
      
    </div>
    <div class="next">
      
        <a href="../snaq_plot/" title="Network estimation and display">
          <span class="direction">
            Next
          </span>
          <div class="page">
            <div class="stretch">
              <div class="title">
                Network estimation and display
              </div>
            </div>
            <div class="button button-next" role="button" aria-label="Next">
              <i class="icon icon-forward"></i>
            </div>
          </div>
        </a>
      
    </div>
  </nav>

            </footer>
          
        </div>
      </article>
      <div class="results" role="status" aria-live="polite">
        <div class="scrollable">
          <div class="wrapper">
            <div class="meta"></div>
            <div class="list"></div>
          </div>
        </div>
      </div>
    </main>
    <script>
      var base_url = '../..';
      var repo_id  = 'crsl4/PhyloNetworks.jl';
    </script>
    <script src="../../assets/javascripts/application-997097ee0c.js"></script>
    
      <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
      <script src="../../assets/mathjaxhelper.js"></script>
    
    
  </body>
</html>