



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="PhyloNetworks is a Julia package for the manipulation, visualization and inference of phylogenetic networks.">
      
      
      
        <meta name="author" content="Claudia Sol&iacute;s-Lemus">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-0.17.5, mkdocs-material-2.9.4">
    
    
      
        <title>Network comparison and manipulation - PhyloNetworks.jl</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/application.451f80e5.css">
      
        <link rel="stylesheet" href="../../assets/stylesheets/application-palette.22915126.css">
      
      
        
        
        <meta name="theme-color" content="#009688">
      
    
    
      <script src="../../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../../assets/fonts/material-icons.css">
    
      <link rel="stylesheet" href="../../assets/Documenter.css">
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="teal" data-md-color-accent="teal">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#comparing-and-manipulating-networks" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../.." title="PhyloNetworks.jl" class="md-header-nav__button md-logo">
          
            <img src="../../snaq_small.png" width="24" height="24">
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                PhyloNetworks.jl
              </span>
              <span class="md-header-nav__topic">
                Network comparison and manipulation
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/crsl4/PhyloNetworks.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="../.." title="PhyloNetworks.jl" class="md-nav__button md-logo">
      
        <img src="../../snaq_small.png" width="48" height="48">
      
    </a>
    PhyloNetworks.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/crsl4/PhyloNetworks.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../.." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" checked>
    
    <label class="md-nav__link" for="nav-2">
      Manual
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        Manual
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../installation/" title="Installation" class="md-nav__link">
      Installation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../inputdata/" title="Input Data for SNaQ" class="md-nav__link">
      Input Data for SNaQ
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../ticr_howtogetQuartetCFs/" title="TICR pipeline" class="md-nav__link">
      TICR pipeline
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../snaq_plot/" title="Network estimation and display" class="md-nav__link">
      Network estimation and display
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Network comparison and manipulation
      </label>
    
    <a href="./" title="Network comparison and manipulation" class="md-nav__link md-nav__link--active">
      Network comparison and manipulation
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#comparing-networks-trees" title="Comparing networks / trees" class="md-nav__link">
    Comparing networks / trees
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#re-rooting-trees-and-networks" title="Re-rooting trees and networks" class="md-nav__link">
    Re-rooting trees and networks
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#what-if-the-root-conflicts-with-the-direction-of-a-reticulation" title="What if the root conflicts with the direction of a reticulation?" class="md-nav__link">
    What if the root conflicts with the direction of a reticulation?
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#extracting-the-major-tree" title="Extracting the major tree" class="md-nav__link">
    Extracting the major tree
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#displayed-trees-and-subnetworks" title="Displayed trees and subnetworks" class="md-nav__link">
    Displayed trees and subnetworks
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../fixednetworkoptim/" title="Candidate Networks" class="md-nav__link">
      Candidate Networks
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../expectedCFs/" title="Extract Expected CFs" class="md-nav__link">
      Extract Expected CFs
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../bootstrap/" title="Bootstrap" class="md-nav__link">
      Bootstrap
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../multiplealleles/" title="Multiple Alleles" class="md-nav__link">
      Multiple Alleles
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../trait_tree/" title="Continuous Trait Evolution" class="md-nav__link">
      Continuous Trait Evolution
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../parsimony/" title="Parsimony on networks" class="md-nav__link">
      Parsimony on networks
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3">
    
    <label class="md-nav__link" for="nav-3">
      Library
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        Library
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../lib/public/" title="Public" class="md-nav__link">
      Public
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../lib/internals/" title="Internals" class="md-nav__link">
      Internals
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#comparing-networks-trees" title="Comparing networks / trees" class="md-nav__link">
    Comparing networks / trees
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#re-rooting-trees-and-networks" title="Re-rooting trees and networks" class="md-nav__link">
    Re-rooting trees and networks
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#what-if-the-root-conflicts-with-the-direction-of-a-reticulation" title="What if the root conflicts with the direction of a reticulation?" class="md-nav__link">
    What if the root conflicts with the direction of a reticulation?
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#extracting-the-major-tree" title="Extracting the major tree" class="md-nav__link">
    Extracting the major tree
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#displayed-trees-and-subnetworks" title="Displayed trees and subnetworks" class="md-nav__link">
    Displayed trees and subnetworks
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/crsl4/PhyloNetworks.jl/edit/master/docs/man/dist_reroot.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <p><a id='Comparing-and-manipulating-networks-1'></a></p>
<h1 id="comparing-and-manipulating-networks">Comparing and manipulating networks</h1>
<p>Examples below follow those in <a href="../snaq_plot/#Getting-a-Network-1">Getting a Network</a>.</p>
<p><a id='Comparing-networks-/-trees-1'></a></p>
<h2 id="comparing-networks-trees">Comparing networks / trees</h2>
<p>Is the SNaQ tree (network with h=0) the same as the ASTRAL tree? We can calculate their Robinson-Foulds distance:</p>
<div class="codehilite"><pre><span></span>julia&gt; hardwiredClusterDistance(astraltree, net0, false)
0
</pre></div>


<p>The last option <code>false</code> is to consider topologies as unrooted. The RF distance is 0, so the two unrooted topologies are the same. If we had considered them as rooted, with whatever root they currently have in their internal representation, we would find a difference:</p>
<div class="codehilite"><pre><span></span>julia&gt; hardwiredClusterDistance(astraltree, net0, true)
5
</pre></div>


<p><a id='Re-rooting-trees-and-networks-1'></a></p>
<h2 id="re-rooting-trees-and-networks">Re-rooting trees and networks</h2>
<p>We can re-root our networks with the outgroup, O, and then re-compare the ASTRAL tree and the SNaQ tree as rooted topologies (and find no difference):</p>
<div class="codehilite"><pre><span></span>julia&gt; rootatnode!(astraltree, &quot;O&quot;)
PhyloNetworks.HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: E, O, C, D, ...
(O,(((C,D):3.029,(B,A):1.324):0.464,E));

julia&gt; rootatnode!(net0, &quot;O&quot;)
PhyloNetworks.HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, B, A, ...
(O,(E,((B,A):1.396,(C,D):10.0):0.485));

julia&gt; hardwiredClusterDistance(astraltree, net0, true)
0
</pre></div>


<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">PhyloPlots</span><span class="p">,</span> <span class="n">RCall</span>
<span class="n">R</span><span class="s">&quot;name &lt;- function(x) file.path(&#39;..&#39;, &#39;assets&#39;, &#39;figures&#39;, x)&quot;</span>
<span class="n">R</span><span class="s">&quot;svg(name(&#39;net0_O.svg&#39;), width=4, height=4)&quot;</span>
<span class="n">R</span><span class="s">&quot;par(mar = c(0, 0, 0, 0))&quot;</span>
<span class="n">plot</span><span class="p">(</span><span class="n">net0</span><span class="p">,</span> <span class="o">:</span><span class="n">R</span><span class="p">);</span>
<span class="n">R</span><span class="s">&quot;dev.off()&quot;</span>
</pre></div>


<p><img alt="net0_O" src="../../assets/figures/net0_O.svg" /></p>
<p>Note that, as in previous chapters, we use the possibilities of <code>RCall</code> to save the plot. We only show this commands once, but they will be run behind the scene each time a plot is called.</p>
<p>After trees/networks are rooted with a correct outgroup, their visualization is more meaningful.</p>
<p>Networks can be re-rooted at a given node or along a given edge. Get help (type <code>?</code>) on the functions <code>rootatnode!</code> and <code>rootonedge!</code> for more info. There are examples in the <a href="../bootstrap/#Bootstrap-1">Bootstrap</a> section.</p>
<p>If the network is plotted with crossing edges, you may identify ways to rotate the children edges at some nodes to untangle some crossing edges. This can be done using the function <code>rotate!</code>. See an example in the <a href="../bootstrap/#Bootstrap-1">Bootstrap</a> section, or type <code>?</code> then <code>rotate!</code>.</p>
<p><a id='What-if-the-root-conflicts-with-the-direction-of-a-reticulation?-1'></a></p>
<h2 id="what-if-the-root-conflicts-with-the-direction-of-a-reticulation">What if the root conflicts with the direction of a reticulation?</h2>
<p>With 1 hybridization or more, the direction of hybrid edges constrain the position of the root. The root cannot be downstream of hybrid edges. Any hybrid node has to be younger than, or of the same age as both of its parents. So time has to flow "downwards" of any hybrid node, and the root cannot be placed "below" a hybrid node. An attempt to re-root the network at a position incompatible with hybrid edges will fail, with a <code>RootMismatch</code> error. To show an example, let's use the network below. We plotted the edge numbers, because we will want to use them later to place the root.</p>
<div class="codehilite"><pre><span></span><span class="n">net7taxa</span> <span class="o">=</span> <span class="n">readTopology</span><span class="p">(</span><span class="s">&quot;(C,D,((O,(E,#H7:::0.196):0.314):0.664,(B,((A1,A2))#H7:::0.804):10.0):10.0);&quot;</span><span class="p">)</span>
<span class="n">plot</span><span class="p">(</span><span class="n">net7taxa</span><span class="p">,</span> <span class="o">:</span><span class="n">R</span><span class="p">,</span> <span class="n">showGamma</span><span class="o">=</span><span class="kc">true</span><span class="p">,</span> <span class="n">showEdgeNumber</span><span class="o">=</span><span class="kc">true</span><span class="p">,</span> <span class="n">tipOffset</span><span class="o">=</span><span class="mf">0.2</span><span class="p">);</span>
</pre></div>


<p><img alt="reroot net7taxa 1" src="../../assets/figures/reroot_net7taxa_1.svg" /></p>
<p>Let's imagine that the A1 and A2 are our outgroups, and we estimated the network above. According to this network, time must flow from the hybrid node towards A1 and A2. So any attempt to reroot the network with A1 as outgroup, or with A2 as outgroup, or with the A clade (on edge 11), will fail with a <code>RootMismatch</code> error:</p>
<div class="codehilite"><pre><span></span><span class="n">rootatnode!</span><span class="p">(</span><span class="n">net7taxa</span><span class="p">,</span> <span class="s">&quot;A1&quot;</span><span class="p">);</span> <span class="c"># ERROR: RootMismatch: non-leaf node 5 had 0 children. ...</span>
<span class="n">rootatnode!</span><span class="p">(</span><span class="n">net7taxa</span><span class="p">,</span> <span class="s">&quot;A2&quot;</span><span class="p">);</span> <span class="c"># ERROR: RootMismatch (again)</span>
<span class="n">rootonedge!</span><span class="p">(</span><span class="n">net7taxa</span><span class="p">,</span> <span class="mi">11</span><span class="p">);</span>   <span class="c"># ERROR: RootMismatch (again)</span>
</pre></div>


<p>In this case, however, it is possible to root the network on either parent edge of the hybrid node. These edges have numbers 12 and 5, based on the plot above. We get these 2 rooted versions of the network:</p>
<div class="codehilite"><pre><span></span><span class="n">R</span><span class="s">&quot;layout(matrix(1:2,1,2))&quot;</span><span class="p">;</span>
<span class="n">rootonedge!</span><span class="p">(</span><span class="n">net7taxa</span><span class="p">,</span> <span class="mi">12</span><span class="p">);</span>
<span class="n">plot</span><span class="p">(</span><span class="n">net7taxa</span><span class="p">,</span> <span class="o">:</span><span class="n">R</span><span class="p">,</span> <span class="n">showGamma</span><span class="o">=</span><span class="kc">true</span><span class="p">,</span> <span class="n">tipOffset</span><span class="o">=</span><span class="mf">0.2</span><span class="p">);</span>
<span class="n">R</span><span class="s">&quot;mtext(&#39;rooted on hybrid edge 12 (major)&#39;, line=-1)&quot;</span>
<span class="n">rootonedge!</span><span class="p">(</span><span class="n">net7taxa</span><span class="p">,</span> <span class="mi">5</span><span class="p">);</span>
<span class="n">plot</span><span class="p">(</span><span class="n">net7taxa</span><span class="p">,</span> <span class="o">:</span><span class="n">R</span><span class="p">,</span> <span class="n">showGamma</span><span class="o">=</span><span class="kc">true</span><span class="p">,</span> <span class="n">tipOffset</span><span class="o">=</span><span class="mf">0.2</span><span class="p">);</span>
<span class="n">R</span><span class="s">&quot;mtext(&#39;rooted on hybrid edge 5 (minor)&#39;, line=-1)&quot;</span>
</pre></div>


<p><img alt="reroot net7taxa 2" src="../../assets/figures/reroot_net7taxa_2.svg" /></p>
<p>On the second plot, the A clade does not <em>appear</em> to be an outgroup, but this is just because the plot follows the major tree primarily, based the major hybrid edges (those with γ&gt;0.5). We can display the exact same network differently, by changing the γ inheritance values to invert the major/minor consideration of the hybrid edges.</p>
<div class="codehilite"><pre><span></span><span class="n">net7taxa</span><span class="o">.</span><span class="n">edge</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="c"># just to check that it&#39;s one of the 2 hybrid edges of interest</span>
<span class="n">setGamma!</span><span class="p">(</span><span class="n">net7taxa</span><span class="o">.</span><span class="n">edge</span><span class="p">[</span><span class="mi">5</span><span class="p">],</span> <span class="mf">0.501</span><span class="p">)</span> <span class="c"># switch major/minor edges</span>
<span class="n">plot</span><span class="p">(</span><span class="n">net7taxa</span><span class="p">,</span> <span class="o">:</span><span class="n">R</span><span class="p">,</span> <span class="n">tipOffset</span><span class="o">=</span><span class="mf">0.2</span><span class="p">);</span> <span class="c"># not showing gamma values, because we changed them artificially</span>
<span class="n">R</span><span class="s">&quot;mtext(&#39;rooted on hybrid edge 5 (considered major)&#39;, line=-1)&quot;</span>
</pre></div>


<p><img alt="reroot net7taxa 3" src="../../assets/figures/reroot_net7taxa_3.svg" /></p>
<p>Conclusion, in this particular example: it is possible to re-root the network to a place where the A clade is indeed an outgroup. But it did require some care, and we discovered that there are 2 acceptable rooting options. The first is more plausible, if we think that the <em>species tree</em> is the <em>major tree</em>, meaning that any gene flow or introgression event replaced less than 50% of the genes in the recipient population.</p>
<p>In other cases, it may not be possible to re-root the network with a known outgroup. It would be the case if A1 was the only outgroup, and if A2 was an ingroup taxon. In such a case, the outgroup knowledge tells us that our estimated network is wrong. One (or more) reticulation in the network must be incorrect. Its placement might be correct, but then its direction would be incorrect. If the network was estimated via <code>snaq!</code>, check tips about <a href="../fixednetworkoptim/#Candidate-networks-compatible-with-a-known-outgroup-1">Candidate networks compatible with a known outgroup</a>.</p>
<p><a id='Extracting-the-major-tree-1'></a></p>
<h2 id="extracting-the-major-tree">Extracting the major tree</h2>
<p>We can also compare the networks estimated with h=0 (net0) and h=1 (net1):</p>
<div class="codehilite"><pre><span></span><span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">rootatnode</span><span class="o">!(</span><span class="nt">net1</span><span class="o">,</span> <span class="s2">&quot;O&quot;</span><span class="o">);</span> <span class="err">#</span> <span class="nt">the</span> <span class="o">;</span> <span class="nt">suppresses</span> <span class="nt">screen</span> <span class="nt">output</span>
<span class="nt">PhyloNetworks</span><span class="p">.</span><span class="nc">HybridNetwork</span><span class="o">,</span> <span class="nt">Rooted</span> <span class="nt">Network</span>
<span class="nt">13</span> <span class="nt">edges</span>
<span class="nt">13</span> <span class="nt">nodes</span><span class="o">:</span> <span class="nt">6</span> <span class="nt">tips</span><span class="o">,</span> <span class="nt">1</span> <span class="nt">hybrid</span> <span class="nt">nodes</span><span class="o">,</span> <span class="nt">6</span> <span class="nt">internal</span> <span class="nt">tree</span> <span class="nt">nodes</span><span class="o">.</span>
<span class="nt">tip</span> <span class="nt">labels</span><span class="o">:</span> <span class="nt">C</span><span class="o">,</span> <span class="nt">D</span><span class="o">,</span> <span class="nt">O</span><span class="o">,</span> <span class="nt">E</span><span class="o">,</span> <span class="o">...</span>
<span class="o">(</span><span class="nt">O</span><span class="o">,((</span><span class="nt">E</span><span class="o">,</span><span class="p">#</span><span class="nn">H7</span><span class="o">:</span><span class="p">::</span><span class="nd">0</span><span class="p">.</span><span class="nc">196</span><span class="o">)</span><span class="p">:</span><span class="nd">0</span><span class="p">.</span><span class="nc">314</span><span class="o">,((</span><span class="nt">B</span><span class="o">,(</span><span class="nt">A</span><span class="o">)</span><span class="p">#</span><span class="nn">H7</span><span class="o">:</span><span class="p">::</span><span class="nd">0</span><span class="p">.</span><span class="nc">804</span><span class="o">)</span><span class="p">:</span><span class="nd">10</span><span class="p">.</span><span class="nc">0</span><span class="o">,(</span><span class="nt">C</span><span class="o">,</span><span class="nt">D</span><span class="o">)</span><span class="p">:</span><span class="nd">10</span><span class="p">.</span><span class="nc">0</span><span class="o">)</span><span class="p">:</span><span class="nd">0</span><span class="p">.</span><span class="nc">664</span><span class="o">));</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">hardwiredClusterDistance</span><span class="o">(</span><span class="nt">net0</span><span class="o">,</span> <span class="nt">net1</span><span class="o">,</span> <span class="nt">true</span><span class="o">)</span>
<span class="nt">2</span>
</pre></div>


<div class="codehilite"><pre><span></span><span class="n">plot</span><span class="p">(</span><span class="n">net1</span><span class="p">,</span> <span class="o">:</span><span class="n">R</span><span class="p">,</span> <span class="n">showGamma</span><span class="o">=</span><span class="kc">true</span><span class="p">);</span>
</pre></div>


<p><img alt="net1_O" src="../../assets/figures/net1_O.svg" /></p>
<p>They differ by 2 clusters: that's because A is of hybrid descent in net1, not in net0.</p>
<p>To beyond this hybrid difference, we can extract the major tree from the network with 1 hybridization, that is, delete the hybrid edge supported by less than 50% of genes. Then we can compare this tree with the ASTRAL/SNaQ tree net0.</p>
<div class="codehilite"><pre><span></span>julia&gt; tree1 = majorTree(net1); # major tree from net1

julia&gt; hardwiredClusterDistance(net0, tree1, true)
0
</pre></div>


<p>They are identical (at distance 0), so here the species network with 1 hybrid node is a refinement of the estimated species tree (this needs not be the case always).</p>
<p>Is the SNaQ network with 1 hybrid node the same as the true network, the one that was initially used to simulate the data?</p>
<p>(digression on the data: gene trees were simulated under the coalescent along some "true" network, then 500 base-pair alignments were simulated along each gene tree with the HKY model, gene trees were estimated from each alignment with RAxML, and these estimated gene trees served as input to both ASTRAL and SNaQ.)</p>
<p>The true network is shown below, correctly rooted at the outgroup O, and plotted with branch lengths proportional to their values in coalescence units:</p>
<div class="codehilite"><pre><span></span><span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">truenet</span> <span class="o">=</span> <span class="nt">readTopology</span><span class="o">(</span><span class="s2">&quot;((((D:0.4,C:0.4):4.8,((A:0.8,B:0.8):2.2)#H1:2.2::0.7):4.0,(#H1:0::0.3,E:3.0):6.2):2.0,O:11.2);&quot;</span><span class="o">);</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">hardwiredClusterDistance</span><span class="o">(</span><span class="nt">net1</span><span class="o">,</span> <span class="nt">truenet</span><span class="o">,</span> <span class="nt">true</span><span class="o">)</span>
<span class="nt">4</span>
</pre></div>


<div class="codehilite"><pre><span></span><span class="n">plot</span><span class="p">(</span><span class="n">truenet</span><span class="p">,</span> <span class="o">:</span><span class="n">R</span><span class="p">,</span> <span class="n">useEdgeLength</span><span class="o">=</span><span class="kc">true</span><span class="p">,</span> <span class="n">showGamma</span><span class="o">=</span><span class="kc">true</span><span class="p">);</span>
</pre></div>


<p><img alt="truenet" src="../../assets/figures/truenet.svg" /></p>
<p>Our estimated network is not the same as the true network:</p>
<ul>
<li>the underlying tree is correctly estimated</li>
<li>the origin of gene flow is correctly estimated: E</li>
<li>the target of gene flow is <em>not</em> correctly estimated: it was the lineage ancestral to (A,B), but it is estimated to be A only.</li>
</ul>
<p>For networks, the distance here is the hardwired cluster distance: the number of hardwired clusters found in one network and not in the other. The <strong>hardwired cluster</strong> associated with an edge is the set of <em>all</em> tips descendant from that edge, i.e. all tips that inherited at least <em>some</em> genetic material from that edge.</p>
<p><a id='Displayed-trees-and-subnetworks-1'></a></p>
<h2 id="displayed-trees-and-subnetworks">Displayed trees and subnetworks</h2>
<p>We can extract all trees displayed in a network. These trees are obtained by picking one parent hybrid edge at each hybrid node, and dropping the other parent hybrid edge. We can choose to pick the "important" hybrid edges only, with heritability γ at or above a threshold. Below we use a γ threshold of 0, so we get all displayed trees:</p>
<div class="codehilite"><pre><span></span>julia&gt; t = displayedTrees(net1, 0.0) # list of trees displayed in network
2-element Array{PhyloNetworks.HybridNetwork,1}:
 PhyloNetworks.HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,(((B,A):10.0,(C,D):10.0):0.664,E));

 PhyloNetworks.HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,((E,A):0.314,((C,D):10.0,B):0.664));

julia&gt; writeTopology(t[1], round=true)
&quot;(O,(((B,A):10.0,(C,D):10.0):0.664,E));&quot;

julia&gt; writeTopology(t[2], round=true)
&quot;(O,((E,A):0.314,((C,D):10.0,B):0.664));&quot;
</pre></div>


<p>If we decide to keep edges with γ&gt;0.2 only, then we are left with a single tree in the list (the major tree). This is because our example has 1 hybrid node with minor γ=0.196.</p>
<div class="codehilite"><pre><span></span>julia&gt; t = displayedTrees(net1, 0.2)
1-element Array{PhyloNetworks.HybridNetwork,1}:
 PhyloNetworks.HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,(((B,A):10.0,(C,D):10.0):0.664,E));
</pre></div>


<p>We can also delete all "non-important" reticulations, those with a minor heritability γ below some threshold. The function below changes our network <code>net1</code>, as indicated by its name ending with a <code>!</code>.</p>
<div class="codehilite"><pre><span></span><span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">deleteHybridThreshold</span><span class="o">!(</span><span class="nt">net1</span><span class="o">,</span> <span class="nt">0</span><span class="p">.</span><span class="nc">1</span><span class="o">)</span>
<span class="nt">PhyloNetworks</span><span class="p">.</span><span class="nc">HybridNetwork</span><span class="o">,</span> <span class="nt">Rooted</span> <span class="nt">Network</span>
<span class="nt">13</span> <span class="nt">edges</span>
<span class="nt">13</span> <span class="nt">nodes</span><span class="o">:</span> <span class="nt">6</span> <span class="nt">tips</span><span class="o">,</span> <span class="nt">1</span> <span class="nt">hybrid</span> <span class="nt">nodes</span><span class="o">,</span> <span class="nt">6</span> <span class="nt">internal</span> <span class="nt">tree</span> <span class="nt">nodes</span><span class="o">.</span>
<span class="nt">tip</span> <span class="nt">labels</span><span class="o">:</span> <span class="nt">C</span><span class="o">,</span> <span class="nt">D</span><span class="o">,</span> <span class="nt">O</span><span class="o">,</span> <span class="nt">E</span><span class="o">,</span> <span class="o">...</span>
<span class="o">(</span><span class="nt">O</span><span class="o">,((</span><span class="nt">E</span><span class="o">,</span><span class="p">#</span><span class="nn">H7</span><span class="o">:</span><span class="p">::</span><span class="nd">0</span><span class="p">.</span><span class="nc">196</span><span class="o">)</span><span class="p">:</span><span class="nd">0</span><span class="p">.</span><span class="nc">314</span><span class="o">,((</span><span class="nt">B</span><span class="o">,(</span><span class="nt">A</span><span class="o">)</span><span class="p">#</span><span class="nn">H7</span><span class="o">:</span><span class="p">::</span><span class="nd">0</span><span class="p">.</span><span class="nc">804</span><span class="o">)</span><span class="p">:</span><span class="nd">10</span><span class="p">.</span><span class="nc">0</span><span class="o">,(</span><span class="nt">C</span><span class="o">,</span><span class="nt">D</span><span class="o">)</span><span class="p">:</span><span class="nd">10</span><span class="p">.</span><span class="nc">0</span><span class="o">)</span><span class="p">:</span><span class="nd">0</span><span class="p">.</span><span class="nc">664</span><span class="o">));</span>
</pre></div>


<p>Nothing happened to our network: because its γ is above 0.1. But if we set the threshold to 0.3, then our reticulation disappears:</p>
<div class="codehilite"><pre><span></span>julia&gt; deleteHybridThreshold!(net1, 0.3)
PhyloNetworks.HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,(((B,A):10.0,(C,D):10.0):0.664,E));
</pre></div>


<p>See also function <code>displayedNetworkAt!</code> to get the network with a single reticulation of interest, and eliminate all other reticulations.</p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../snaq_plot/" title="Network estimation and display" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Network estimation and display
              </span>
            </div>
          </a>
        
        
          <a href="../fixednetworkoptim/" title="Candidate Networks" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Candidate Networks
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/application.30f6b8b1.js"></script>
      
      <script>app.initialize({version:"0.17.5",url:{base:"../.."}})</script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../../assets/mathjaxhelper.js"></script>
      
    
    
      
    
  </body>
</html>