<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Network comparison and manipulation · PhyloNetworks.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="PhyloNetworks.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">PhyloNetworks.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../installation/">Installation</a></li><li><a class="tocitem" href="../inputdata/">Input Data for SNaQ</a></li><li><a class="tocitem" href="../ticr_howtogetQuartetCFs/">TICR pipeline</a></li><li><a class="tocitem" href="../snaq_plot/">Network estimation and display</a></li><li class="is-active"><a class="tocitem" href>Network comparison and manipulation</a><ul class="internal"><li><a class="tocitem" href="#Comparing-networks-/-trees"><span>Comparing networks / trees</span></a></li><li><a class="tocitem" href="#Re-rooting-trees-and-networks"><span>Re-rooting trees and networks</span></a></li><li><a class="tocitem" href="#What-if-the-root-conflicts-with-the-direction-of-a-reticulation?"><span>What if the root conflicts with the direction of a reticulation?</span></a></li><li><a class="tocitem" href="#Extracting-the-major-tree"><span>Extracting the major tree</span></a></li><li><a class="tocitem" href="#Displayed-trees-and-subnetworks"><span>Displayed trees and subnetworks</span></a></li></ul></li><li><a class="tocitem" href="../fixednetworkoptim/">Candidate Networks</a></li><li><a class="tocitem" href="../expectedCFs/">Extract Expected CFs</a></li><li><a class="tocitem" href="../bootstrap/">Bootstrap</a></li><li><a class="tocitem" href="../multiplealleles/">Multiple Alleles</a></li><li><a class="tocitem" href="../trait_tree/">Continuous Trait Evolution</a></li><li><a class="tocitem" href="../parsimony/">Parsimony on networks</a></li><li><a class="tocitem" href="../fitDiscrete/">Discrete Trait Evolution</a></li><li><a class="tocitem" href="../nj/">Neighbour Joining</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/public/">Public</a></li><li><a class="tocitem" href="../../lib/internals/">Internals</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Network comparison and manipulation</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Network comparison and manipulation</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/crsl4/PhyloNetworks.jl/blob/master/docs/src/man/dist_reroot.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Comparing-and-manipulating-networks"><a class="docs-heading-anchor" href="#Comparing-and-manipulating-networks">Comparing and manipulating networks</a><a id="Comparing-and-manipulating-networks-1"></a><a class="docs-heading-anchor-permalink" href="#Comparing-and-manipulating-networks" title="Permalink"></a></h1><p>Examples below follow those in <a href="../snaq_plot/#Getting-a-Network">Getting a Network</a>.</p><h2 id="Comparing-networks-/-trees"><a class="docs-heading-anchor" href="#Comparing-networks-/-trees">Comparing networks / trees</a><a id="Comparing-networks-/-trees-1"></a><a class="docs-heading-anchor-permalink" href="#Comparing-networks-/-trees" title="Permalink"></a></h2><p>Is the SNaQ tree (network with h=0) the same as the ASTRAL tree? We can calculate their Robinson-Foulds distance:</p><pre><code class="language-julia-repl">julia&gt; hardwiredClusterDistance(astraltree, net0, false)
0</code></pre><p>The last option <code>false</code> is to consider topologies as unrooted. The RF distance is 0, so the two unrooted topologies are the same. If we had considered them as rooted, with whatever root they currently have in their internal representation, we would find a difference:</p><pre><code class="language-julia-repl">julia&gt; hardwiredClusterDistance(astraltree, net0, true)
5</code></pre><h2 id="Re-rooting-trees-and-networks"><a class="docs-heading-anchor" href="#Re-rooting-trees-and-networks">Re-rooting trees and networks</a><a id="Re-rooting-trees-and-networks-1"></a><a class="docs-heading-anchor-permalink" href="#Re-rooting-trees-and-networks" title="Permalink"></a></h2><p>We can re-root our networks with the outgroup, O, and then re-compare the ASTRAL tree and the SNaQ tree as rooted topologies (and find no difference):</p><pre><code class="language-julia-repl">julia&gt; rootatnode!(astraltree, &quot;O&quot;)
HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: E, O, C, D, ...
(O,(((C,D)100.0:3.029,(B,A)100.0:1.324)100.0:0.464,E));

julia&gt; rootatnode!(net0, &quot;O&quot;)
HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, B, A, ...
(O,(E,((B,A):1.396,(C,D):10.0):0.485));

julia&gt; hardwiredClusterDistance(astraltree, net0, true)
0</code></pre><pre><code class="language-julia">using PhyloPlots, RCall
R&quot;name &lt;- function(x) file.path(&#39;..&#39;, &#39;assets&#39;, &#39;figures&#39;, x)&quot;
R&quot;svg(name(&#39;net0_O.svg&#39;), width=4, height=4)&quot;
R&quot;par&quot;(mar=[0,0,0,0])
plot(net0, :R);
R&quot;dev.off()&quot;</code></pre><p><img src="../../assets/figures/net0_O.svg" alt="net0_O"/></p><p>Note that, as in previous chapters, we use the possibilities of <code>RCall</code> to save the plot. We only show this commands once, but they will be run behind the scene each time a plot is called.</p><p>After trees/networks are rooted with a correct outgroup, their visualization is more meaningful.</p><p>Networks can be re-rooted at a given node or along a given edge. Get help (type <code>?</code>) on the functions <code>rootatnode!</code> and <code>rootonedge!</code> for more info. There are examples in the <a href="../bootstrap/#Bootstrap">Bootstrap</a> section.</p><p>If the network is plotted with crossing edges, you may identify ways to rotate the children edges at some nodes to untangle some crossing edges. This can be done using the function <code>rotate!</code>. See an example in the <a href="../bootstrap/#Bootstrap">Bootstrap</a> section, or type <code>?</code> then <code>rotate!</code>.</p><h2 id="What-if-the-root-conflicts-with-the-direction-of-a-reticulation?"><a class="docs-heading-anchor" href="#What-if-the-root-conflicts-with-the-direction-of-a-reticulation?">What if the root conflicts with the direction of a reticulation?</a><a id="What-if-the-root-conflicts-with-the-direction-of-a-reticulation?-1"></a><a class="docs-heading-anchor-permalink" href="#What-if-the-root-conflicts-with-the-direction-of-a-reticulation?" title="Permalink"></a></h2><p>With 1 hybridization or more, the direction of hybrid edges constrain the position of the root. The root cannot be downstream of hybrid edges. Any hybrid node has to be younger than, or of the same age as both of its parents. So time has to flow &quot;downwards&quot; of any hybrid node, and the root cannot be placed &quot;below&quot; a hybrid node. An attempt to re-root the network at a position incompatible with hybrid edges will fail, with a <code>RootMismatch</code> error. To show an example, let&#39;s use the network below. We plotted the edge numbers, because we will want to use them later to place the root.</p><pre><code class="language-julia">net7taxa = readTopology(&quot;(C,D,((O,(E,#H7:::0.196):0.314):0.664,(B,((A1,A2))#H7:::0.804):10.0):10.0);&quot;)
plot(net7taxa, :R, showGamma=true, showEdgeNumber=true, tipOffset=0.2);</code></pre><p><img src="../../assets/figures/reroot_net7taxa_1.svg" alt="reroot net7taxa 1"/></p><p>Let&#39;s imagine that the A1 and A2 are our outgroups, and we estimated the network above. According to this network, time must flow from the hybrid node towards A1 and A2. So any attempt to reroot the network with A1 as outgroup, or with A2 as outgroup, or with the A clade (on edge 11), will fail with a <code>RootMismatch</code> error:</p><pre><code class="language-julia">rootatnode!(net7taxa, &quot;A1&quot;); # ERROR: RootMismatch: non-leaf node 5 had 0 children. ...
rootatnode!(net7taxa, &quot;A2&quot;); # ERROR: RootMismatch (again)
rootonedge!(net7taxa, 11);   # ERROR: RootMismatch (again)</code></pre><p>In this case, however, it is possible to root the network on either parent edge of the hybrid node. These edges have numbers 12 and 5, based on the plot above. We get these 2 rooted versions of the network:</p><pre><code class="language-julia">R&quot;layout(matrix(1:2,1,2))&quot;;
rootonedge!(net7taxa, 12);
plot(net7taxa, :R, showGamma=true, tipOffset=0.2);
R&quot;mtext&quot;(&quot;rooted on hybrid edge 12 (major)&quot;, line=-1)
rootonedge!(net7taxa, 5);
plot(net7taxa, :R, showGamma=true, tipOffset=0.2);
R&quot;mtext&quot;(&quot;rooted on hybrid edge 5 (minor)&quot;, line=-1);</code></pre><p><img src="../../assets/figures/reroot_net7taxa_2.svg" alt="reroot net7taxa 2"/></p><p>On the second plot, the A clade does not <em>appear</em> to be an outgroup, but this is just because the plot follows the major tree primarily, based the major hybrid edges (those with γ&gt;0.5). We can display the exact same network differently, by changing the γ inheritance values to invert the major/minor consideration of the hybrid edges.</p><pre><code class="language-julia">net7taxa.edge[5] # just to check that it&#39;s one of the 2 hybrid edges of interest
setGamma!(net7taxa.edge[5], 0.501) # switch major/minor edges
plot(net7taxa, :R, tipOffset=0.2); # not showing gamma values, because we changed them artificially
R&quot;mtext&quot;(&quot;rooted on hybrid edge 5 (considered major)&quot;, line=-1);</code></pre><p><img src="../../assets/figures/reroot_net7taxa_3.svg" alt="reroot net7taxa 3"/></p><p>Conclusion, in this particular example: it is possible to re-root the network to a place where the A clade is indeed an outgroup. But it did require some care, and we discovered that there are 2 acceptable rooting options. The first is more plausible, if we think that the <em>species tree</em> is the <em>major tree</em>, meaning that any gene flow or introgression event replaced less than 50% of the genes in the recipient population.</p><p>In other cases, it may not be possible to re-root the network with a known outgroup. It would be the case if A1 was the only outgroup, and if A2 was an ingroup taxon. In such a case, the outgroup knowledge tells us that our estimated network is wrong. One (or more) reticulation in the network must be incorrect. Its placement might be correct, but then its direction would be incorrect. If the network was estimated via <code>snaq!</code>, check tips about <a href="../fixednetworkoptim/#Candidate-networks-compatible-with-a-known-outgroup">Candidate networks compatible with a known outgroup</a>.</p><h2 id="Extracting-the-major-tree"><a class="docs-heading-anchor" href="#Extracting-the-major-tree">Extracting the major tree</a><a id="Extracting-the-major-tree-1"></a><a class="docs-heading-anchor-permalink" href="#Extracting-the-major-tree" title="Permalink"></a></h2><p>We can also compare the networks estimated with h=0 (net0) and h=1 (net1):</p><pre><code class="language-julia-repl">julia&gt; rootatnode!(net1, &quot;O&quot;); # the ; suppresses screen output
HybridNetwork, Rooted Network
13 edges
13 nodes: 6 tips, 1 hybrid nodes, 6 internal tree nodes.
tip labels: C, D, O, E, ...
(O,((E,#H7:::0.196):0.314,((B,(A)#H7:::0.804):10.0,(C,D):10.0):0.664));

julia&gt; hardwiredClusterDistance(net0, net1, true)
2</code></pre><pre><code class="language-julia">plot(net1, :R, showGamma=true);</code></pre><p><img src="../../assets/figures/net1_O.svg" alt="net1_O"/></p><p>They differ by 2 clusters: that&#39;s because A is of hybrid descent in net1, not in net0.</p><p>To beyond this hybrid difference, we can extract the major tree from the network with 1 hybridization, that is, delete the hybrid edge supported by less than 50% of genes. Then we can compare this tree with the ASTRAL/SNaQ tree net0.</p><pre><code class="language-julia-repl">julia&gt; tree1 = majorTree(net1); # major tree from net1

julia&gt; hardwiredClusterDistance(net0, tree1, true)
0</code></pre><p>They are identical (at distance 0), so here the species network with 1 hybrid node is a refinement of the estimated species tree (this needs not be the case always).</p><p>Is the SNaQ network with 1 hybrid node the same as the true network, the one that was initially used to simulate the data?</p><p>(digression on the data: gene trees were simulated under the coalescent along some &quot;true&quot; network, then 500 base-pair alignments were simulated along each gene tree with the HKY model, gene trees were estimated from each alignment with RAxML, and these estimated gene trees served as input to both ASTRAL and SNaQ.)</p><p>The true network is shown below, correctly rooted at the outgroup O, and plotted with branch lengths proportional to their values in coalescence units:</p><pre><code class="language-julia-repl">julia&gt; truenet = readTopology(&quot;((((D:0.4,C:0.4):4.8,((A:0.8,B:0.8):2.2)#H1:2.2::0.7):4.0,(#H1:0::0.3,E:3.0):6.2):2.0,O:11.2);&quot;);

julia&gt; hardwiredClusterDistance(net1, truenet, true)
4</code></pre><pre><code class="language-julia">plot(truenet, :R, useEdgeLength=true, showGamma=true);</code></pre><p><img src="../../assets/figures/truenet.svg" alt="truenet"/></p><p>Our estimated network is not the same as the true network:</p><ul><li>the underlying tree is correctly estimated</li><li>the origin of gene flow is correctly estimated: E</li><li>the target of gene flow is <em>not</em> correctly estimated: it was the lineage ancestral to (A,B), but it is estimated to be A only.</li></ul><p>For networks, the distance here is the hardwired cluster distance: the number of hardwired clusters found in one network and not in the other. The <strong>hardwired cluster</strong> associated with an edge is the set of <em>all</em> tips descendant from that edge, i.e. all tips that inherited at least <em>some</em> genetic material from that edge.</p><h2 id="Displayed-trees-and-subnetworks"><a class="docs-heading-anchor" href="#Displayed-trees-and-subnetworks">Displayed trees and subnetworks</a><a id="Displayed-trees-and-subnetworks-1"></a><a class="docs-heading-anchor-permalink" href="#Displayed-trees-and-subnetworks" title="Permalink"></a></h2><p>We can extract all trees displayed in a network. These trees are obtained by picking one parent hybrid edge at each hybrid node, and dropping the other parent hybrid edge. We can choose to pick the &quot;important&quot; hybrid edges only, with heritability γ at or above a threshold. Below we use a γ threshold of 0, so we get all displayed trees:</p><pre><code class="language-julia-repl">julia&gt; t = displayedTrees(net1, 0.0) # list of trees displayed in network
2-element Vector{HybridNetwork}:
 HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,(((B,A):10.0,(C,D):10.0):0.664,E));

 HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,((E,A):0.314,((C,D):10.0,B):0.664));

julia&gt; writeTopology(t[1], round=true)
&quot;(O,(((B,A):10.0,(C,D):10.0):0.664,E));&quot;

julia&gt; writeTopology(t[2], round=true)
&quot;(O,((E,A):0.314,((C,D):10.0,B):0.664));&quot;</code></pre><p>If we decide to keep edges with γ&gt;0.2 only, then we are left with a single tree in the list (the major tree). This is because our example has 1 hybrid node with minor γ=0.196.</p><pre><code class="language-julia-repl">julia&gt; t = displayedTrees(net1, 0.2)
1-element Vector{HybridNetwork}:
 HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,(((B,A):10.0,(C,D):10.0):0.664,E));</code></pre><p>We can also delete all &quot;non-important&quot; reticulations, those with a minor heritability γ below some threshold. The function below changes our network <code>net1</code>, as indicated by its name ending with a <code>!</code>.</p><pre><code class="language-julia-repl">julia&gt; deleteHybridThreshold!(net1, 0.1)
HybridNetwork, Rooted Network
13 edges
13 nodes: 6 tips, 1 hybrid nodes, 6 internal tree nodes.
tip labels: C, D, O, E, ...
(O,((E,#H7:::0.196):0.314,((B,(A)#H7:::0.804):10.0,(C,D):10.0):0.664));</code></pre><p>Nothing happened to our network: because its γ is above 0.1. But if we set the threshold to 0.3, then our reticulation disappears:</p><pre><code class="language-julia-repl">julia&gt; deleteHybridThreshold!(net1, 0.3)
HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,(((B,A):10.0,(C,D):10.0):0.664,E));</code></pre><p>See also function <code>displayedNetworkAt!</code> to get the network with a single reticulation of interest, and eliminate all other reticulations.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../snaq_plot/">« Network estimation and display</a><a class="docs-footer-nextpage" href="../fixednetworkoptim/">Candidate Networks »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 15 October 2021 22:11">Friday 15 October 2021</span>. Using Julia version 1.6.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
