<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>fitting DNA on a network · PhyloNetworks.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="PhyloNetworks.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">PhyloNetworks.jl</span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../installation/">Installation</a></li><li><a class="tocitem" href="../inputdata/">Input Data for SNaQ</a></li><li><a class="tocitem" href="../ticr_howtogetQuartetCFs/">TICR pipeline</a></li><li><a class="tocitem" href="../snaq_plot/">Network estimation and display</a></li><li><a class="tocitem" href="../dist_reroot/">Network comparison and manipulation</a></li><li><a class="tocitem" href="../fixednetworkoptim/">Candidate Networks</a></li><li><a class="tocitem" href="../expectedCFs/">Extract Expected CFs</a></li><li><a class="tocitem" href="../bootstrap/">Bootstrap</a></li><li><a class="tocitem" href="../multiplealleles/">Multiple Alleles</a></li><li><a class="tocitem" href="../trait_tree/">Continuous Trait Evolution</a></li><li><a class="tocitem" href="../parsimony/">Parsimony on networks</a></li><li><a class="tocitem" href="../fitDiscrete/">Discrete Trait Evolution</a></li><li><a class="tocitem" href="../nj/">Neighbour Joining</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/public/">Public</a></li><li><a class="tocitem" href="../../lib/internals/">Internals</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>fitting DNA on a network</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>fitting DNA on a network</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/crsl4/PhyloNetworks.jl/blob/master/docs/src/man/fitdiscreteDNA.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="fitting-DNA-on-a-network-1"><a class="docs-heading-anchor" href="#fitting-DNA-on-a-network-1">fitting DNA on a network</a><a class="docs-heading-anchor-permalink" href="#fitting-DNA-on-a-network-1" title="Permalink"></a></h1><p>The methods below model each DNA site as a trait, assuming that sites are unlinked, that is, they evolve independently of each other. In other words, this is a &quot;concatenation&quot; approach where sites from the same locus do not share information about their evolutionary path. This is appropriate if recombination is assumed to have occurred within genes.</p><h2 id="DNA-evolution:-data-and-models-1"><a class="docs-heading-anchor" href="#DNA-evolution:-data-and-models-1">DNA evolution: data and models</a><a class="docs-heading-anchor-permalink" href="#DNA-evolution:-data-and-models-1" title="Permalink"></a></h2><h3 id="reading-in-an-alignment-1"><a class="docs-heading-anchor" href="#reading-in-an-alignment-1">reading in an alignment</a><a class="docs-heading-anchor-permalink" href="#reading-in-an-alignment-1" title="Permalink"></a></h3><p>As for trait evolution, <a href="../../lib/public/#PhyloNetworks.fitdiscrete"><code>fitdiscrete</code></a> can be used. It can be given data in a variety of ways. For DNA, this is one way:</p><pre><code class="language-julia-repl">julia&gt; # read in network
       dna_net = readTopology(&quot;((((((((((((((Ae_caudata_Tr275:1.0,Ae_caudata_Tr276:1.0):1.0,Ae_caudata_Tr139:1.0):1.0)#H1:1.0::0.6,((((((Ae_longissima_Tr241:1.0,Ae_longissima_Tr242:1.0):1.0,Ae_longissima_Tr355:1.0):1.0,(Ae_sharonensis_Tr265:1.0,Ae_sharonensis_Tr264:1.0):1.0):1.0,((Ae_bicornis_Tr408:1.0,Ae_bicornis_Tr407:1.0):1.0,Ae_bicornis_Tr406:1.0):1.0):1.0,((Ae_searsii_Tr164:1.0,Ae_searsii_Tr165:1.0):1.0,Ae_searsii_Tr161:1.0):1.0):1.0)#H2:1.0::0.6):1.0,(((Ae_umbellulata_Tr266:1.0,Ae_umbellulata_Tr257:1.0):1.0,Ae_umbellulata_Tr268:1.0):1.0,#H1:1.0::0.4):1.0):1.0,((Ae_comosa_Tr271:1.0,Ae_comosa_Tr272:1.0):1.0,(((Ae_uniaristata_Tr403:1.0,Ae_uniaristata_Tr357:1.0):1.0,Ae_uniaristata_Tr402:1.0):1.0,Ae_uniaristata_Tr404:1.0):1.0):1.0):1.0,(((Ae_tauschii_Tr352:1.0,Ae_tauschii_Tr351:1.0):1.0,(Ae_tauschii_Tr180:1.0,Ae_tauschii_Tr125:1.0):1.0):1.0,(#H2:1.0::0.4,((((Ae_mutica_Tr237:1.0,Ae_mutica_Tr329:1.0):1.0,Ae_mutica_Tr244:1.0):1.0,Ae_mutica_Tr332:1.0):1.0)#H4:1.0::0.6):1.0):1.0):1.0,(((T_boeoticum_TS8:1.0,(T_boeoticum_TS10:1.0,T_boeoticum_TS3:1.0):1.0):1.0,T_boeoticum_TS4:1.0):1.0,((T_urartu_Tr315:1.0,T_urartu_Tr232:1.0):1.0,(T_urartu_Tr317:1.0,T_urartu_Tr309:1.0):1.0):1.0):1.0):1.0,(((((Ae_speltoides_Tr320:1.0,Ae_speltoides_Tr323:1.0):1.0,Ae_speltoides_Tr223:1.0):1.0,Ae_speltoides_Tr251:1.0):1.0):1.0,#H4:1.0::0.4):1.0):1.0):1.0,Ta_caputMedusae_TB2:1.0):1.0,S_vavilovii_Tr279:1.0):1.0,Er_bonaepartis_TB1:1.0):1.0,H_vulgare_HVens23:1.0);&quot;);

julia&gt; # read in alignment in FASTA format
       fastafile = joinpath(dirname(pathof(PhyloNetworks)), &quot;..&quot;,&quot;examples&quot;,&quot;Ae_bicornis_Tr406_Contig10132.aln&quot;);

julia&gt; dna_dat, dna_weights = readfastatodna(fastafile, true);

julia&gt; dna_dat
22×210 DataFrames.DataFrame. Omitted printing of 205 columns
│ Row │ taxon                │ x1        │ x2        │ x3        │ x4        │
│     │ String               │ BioSymbo… │ BioSymbo… │ BioSymbo… │ BioSymbo… │
├─────┼──────────────────────┼───────────┼───────────┼───────────┼───────────┤
│ 1   │ Ae_uniaristata_Tr357 │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 2   │ Ae_mutica_Tr237      │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 3   │ Ae_tauschii_Tr352    │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 4   │ Ae_searsii_Tr165     │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
│ 5   │ Ae_sharonensis_Tr265 │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 6   │ Ae_bicornis_Tr407    │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
│ 7   │ Ae_longissima_Tr241  │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
⋮
│ 15  │ Ae_longissima_Tr242  │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 16  │ Ae_comosa_Tr271      │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 17  │ Ae_searsii_Tr164     │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
│ 18  │ Ae_speltoides_Tr251  │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 19  │ Ae_bicornis_Tr408    │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 20  │ Ae_tauschii_Tr351    │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
│ 21  │ Ae_longissima_Tr355  │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
│ 22  │ Ae_caudata_Tr139     │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │

julia&gt; dna_weights
209-element Array{Float64,1}:
 23.0
 18.0
 13.0
 16.0
  1.0
  1.0
  1.0
  1.0
  1.0
  9.0
  ⋮
  1.0
  1.0
  1.0
  1.0
  1.0
  1.0
  1.0
  2.0
  3.0</code></pre><p>Here, <code>dna_dat</code> is a single data frame containing both species names and trait data (site patterns). The alignment was summarized by listing each observed site pattern only once in <code>dna_dat</code>. <code>dna_weights</code> is a vector of weights, containing the number of times that each site pattern was observed.</p><h3 id="sequence-substitution-models-1"><a class="docs-heading-anchor" href="#sequence-substitution-models-1">sequence substitution models</a><a class="docs-heading-anchor-permalink" href="#sequence-substitution-models-1" title="Permalink"></a></h3><p>DNA-specific substitution models have 4 states: the 4 nucleotides from <a href="https://github.com/BioJulia/BioSymbols.jl">BioSymbols</a> (listed <a href="http://biojulia.net/BioSymbols.jl/stable/nucleicacids/">here</a>). Each model has a relative and an absolute version.</p><ul><li><code>:JC69</code> Jukes &amp; Cantor 1969 model: one single rate for all transitions. The relative version has values -1 along the diagonal of the rate matrix (1 expected transition / unit of time). The absolute version has an extra parameter to scale the rate matrix.</li><li><code>:HKY85</code> Hasegawa, Kishino &amp; Yano 1985: treats transitions differently from transversions. The relative is scaled to predict an average of 1 transition / unit of time.</li></ul><p>We may allow for rate variation across sites using the <code>:RV</code> option.</p><h3 id="likelihood-of-a-fixed-network-1"><a class="docs-heading-anchor" href="#likelihood-of-a-fixed-network-1">likelihood of a fixed network</a><a class="docs-heading-anchor-permalink" href="#likelihood-of-a-fixed-network-1" title="Permalink"></a></h3><p>In the examples below, none of the rate parameters are optimized, so we get to see the default starting values.</p><pre><code class="language-julia-repl">julia&gt; d1 = fitdiscrete(dna_net, :JC69, dna_dat, dna_weights, :RV; optimizeQ=false, optimizeRVAS=false)
┌ Warning: the network contains taxa with no data: those will be pruned
└ @ PhyloNetworks ~/build/crsl4/PhyloNetworks.jl/src/traitsLikDiscrete.jl:824
PhyloNetworks.StatisticalSubstitutionModel:
Jukes and Cantor 69 Substitution Model,
relative rate version
off-diagonal rates equal to 1/3
rate matrix Q:
               A       C       G       T
       A       *  0.3333  0.3333  0.3333
       C  0.3333       *  0.3333  0.3333
       G  0.3333  0.3333       *  0.3333
       T  0.3333  0.3333  0.3333       *
209 traits, 22 species
variable rates across sites ~ discretized gamma with
 alpha=1.0
 4 categories
 rate multipliers: [0.14578, 0.51313, 1.07083, 2.27025]
on a network with 3 reticulations
log-likelihood: -14442.27642

julia&gt; d2 = fitdiscrete(dna_net, :HKY85, dna_dat, dna_weights, :RV; optimizeQ=false, optimizeRVAS=false)
┌ Warning: the network contains taxa with no data: those will be pruned
└ @ PhyloNetworks ~/build/crsl4/PhyloNetworks.jl/src/traitsLikDiscrete.jl:824
PhyloNetworks.StatisticalSubstitutionModel:
HKY85 Substitution Model base frequencies: [0.29100263035741913, 0.2682964567538295, 0.23638403218319665, 0.2043168807055547]
relative rate version with transition/tranversion ratio kappa = 1.0,
 scaled so that there is one substitution per unit time
rate matrix Q:
               A       C       G       T
       A       *  0.3598  0.3170  0.2740
       C  0.3902       *  0.3170  0.2740
       G  0.3902  0.3598       *  0.2740
       T  0.3902  0.3598  0.3170       *
209 traits, 22 species
variable rates across sites ~ discretized gamma with
 alpha=1.0
 4 categories
 rate multipliers: [0.14578, 0.51313, 1.07083, 2.27025]
on a network with 3 reticulations
log-likelihood: -14420.83584</code></pre><p>When allowing for rate variation across sites, the default α is 1.</p><p>In the more interesting examples below, we optimize the evolutionary rates and the way rates vary across sites (which is the default).</p><pre><code class="language-julia-repl">julia&gt; d3 = fitdiscrete(dna_net, :JC69, dna_dat, dna_weights, :RV; ftolAbs=0.1, xtolAbs=0.01)
┌ Warning: the network contains taxa with no data: those will be pruned
└ @ PhyloNetworks ~/build/crsl4/PhyloNetworks.jl/src/traitsLikDiscrete.jl:824
PhyloNetworks.StatisticalSubstitutionModel:
Jukes and Cantor 69 Substitution Model,
relative rate version
off-diagonal rates equal to 1/3
rate matrix Q:
               A       C       G       T
       A       *  0.3333  0.3333  0.3333
       C  0.3333       *  0.3333  0.3333
       G  0.3333  0.3333       *  0.3333
       T  0.3333  0.3333  0.3333       *
209 traits, 22 species
variable rates across sites ~ discretized gamma with
 alpha=0.0412500000000001
 4 categories
 rate multipliers: [0.0, 0.0, 0.00112, 3.99888]
on a network with 3 reticulations
log-likelihood: -3927.09709</code></pre><p>Lenient tolerance parameters <code>ftolAbs</code> etc. have been chosen here to make this example faster. Note that the fitted object contains a separate version of the input network, where any taxon without data has been pruned, and where branch length numbers may have been modified.</p><pre><code class="language-julia-repl">julia&gt; d3.net
HybridNetwork, Rooted Network
53 edges
51 nodes: 22 tips, 3 hybrid nodes, 26 internal tree nodes.
tip labels: Ae_caudata_Tr275, Ae_caudata_Tr139, Ae_longissima_Tr241, Ae_longissima_Tr242, ...
(H_vulgare_HVens23:1.0,((((Ae_speltoides_Tr251:2.0):1.0,#H4:1.0::0.4):1.0,((((((Ae_caudata_Tr139:1.0,Ae_caudata_Tr275:2.0):1.0)#H1:1.0::0.6,((((((Ae_longissima_Tr241:1.0,Ae_longissima_Tr242:1.0):1.0,Ae_longissima_Tr355:1.0):1.0,Ae_sharonensis_Tr265:2.0):1.0,((Ae_bicornis_Tr408:1.0,Ae_bicornis_Tr407:1.0):1.0,Ae_bicornis_Tr406:1.0):1.0):1.0,(Ae_searsii_Tr164:1.0,Ae_searsii_Tr165:1.0):2.0):1.0)#H2:1.0::0.6):1.0,#H1:2.0::0.4):1.0,((Ae_comosa_Tr271:1.0,Ae_comosa_Tr272:1.0):1.0,((Ae_uniaristata_Tr403:1.0,Ae_uniaristata_Tr357:1.0):1.0,Ae_uniaristata_Tr402:1.0):2.0):1.0):1.0,(((Ae_tauschii_Tr352:1.0,Ae_tauschii_Tr351:1.0):1.0,Ae_tauschii_Tr125:2.0):1.0,(#H2:1.0::0.4,(Ae_mutica_Tr237:4.0)#H4:1.0::0.6):1.0):1.0):2.0):1.0):4.0);</code></pre></article></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 15 May 2020 15:32">Friday 15 May 2020</span>. Using Julia version 1.4.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
