<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Pairwise and quartet data · PhyloNetworks.jl</title><meta name="title" content="Pairwise and quartet data · PhyloNetworks.jl"/><meta property="og:title" content="Pairwise and quartet data · PhyloNetworks.jl"/><meta property="twitter:title" content="Pairwise and quartet data · PhyloNetworks.jl"/><meta name="description" content="Documentation for PhyloNetworks.jl."/><meta property="og:description" content="Documentation for PhyloNetworks.jl."/><meta property="twitter:description" content="Documentation for PhyloNetworks.jl."/><meta property="og:url" content="https://juliaphylo.github.io/PhyloNetworks.jl/stable/man/expecteddata/"/><meta property="twitter:url" content="https://juliaphylo.github.io/PhyloNetworks.jl/stable/man/expecteddata/"/><link rel="canonical" href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/man/expecteddata/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="PhyloNetworks.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="PhyloNetworks.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PhyloNetworks.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../installation/">Installation</a></li><li><a class="tocitem" href="../introduction/">Introduction</a></li><li><a class="tocitem" href="../netmanipulation/">Network manipulation</a></li><li><a class="tocitem" href="../net_plot/">Network visualization</a></li><li><a class="tocitem" href="../dist_reroot/">Network comparison and manipulation</a></li><li><a class="tocitem" href="../network_support/">Network support</a></li><li><a class="tocitem" href="../parsimony/">Parsimony on networks</a></li><li><a class="tocitem" href="../nj/">Neighbour Joining</a></li><li class="is-active"><a class="tocitem" href>Pairwise and quartet data</a><ul class="internal"><li><a class="tocitem" href="#average-pairwise-distance"><span>average pairwise distance</span></a></li><li><a class="tocitem" href="#hamming-distance"><span>hamming distance</span></a></li><li><a class="tocitem" href="#expected-f2-statistics"><span>expected f2-statistics</span></a></li><li><a class="tocitem" href="#expected-f4-statistics"><span>expected f4-statistics</span></a></li><li><a class="tocitem" href="#expected-f3-statistics"><span>expected f3-statistics</span></a></li><li><a class="tocitem" href="#quartet-concordance-factors"><span>quartet concordance factors</span></a></li></ul></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/public/">Public</a></li><li><a class="tocitem" href="../../lib/internals/">Internals</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Pairwise and quartet data</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Pairwise and quartet data</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/PhyloNetworks.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/PhyloNetworks.jl/blob/master/docs/src/man/expecteddata.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Pairwise-and-quartet-data"><a class="docs-heading-anchor" href="#Pairwise-and-quartet-data">Pairwise and quartet data</a><a id="Pairwise-and-quartet-data-1"></a><a class="docs-heading-anchor-permalink" href="#Pairwise-and-quartet-data" title="Permalink"></a></h1><p>We show here some functionalities to calculate dissimilarities or distances between taxa, either expected from a given network, or observed in data. To calculate expectations under a network, this network&#39;s edges need to have branch lengths and γ inheritance values.</p><p>We use 2 example networks in this section: <code>net0</code> without reticulations (a tree) and <code>net2</code> with 2 reticulations. <code>net0</code> is in fact <code>net2</code>&#39;s major tree: obtained by deleting every minor hybrid edge.</p><pre><code class="language-julia hljs">net0 = readnewick(&quot;(O:5.5,((E:4.0,(D:3.0,(C:1.0,B:1.0):2.0):1.0):1.0,A:5.0):0.5);&quot;);
net2 = readnewick(&quot;(O:5.5,(((E:1.5)#H1:2.5::0.7,((#H1:0,D:1.5):1.5,
    ((C:1,B:1):1)#H2:1::0.6):1.0):1.0,(#H2:0,A:2):3):0.5);&quot;)</code></pre><p><img src="../../assets/figures/expectedata_fig_net02.svg" alt="net02-distquartet"/></p><h2 id="average-pairwise-distance"><a class="docs-heading-anchor" href="#average-pairwise-distance">average pairwise distance</a><a id="average-pairwise-distance-1"></a><a class="docs-heading-anchor-permalink" href="#average-pairwise-distance" title="Permalink"></a></h2><p>One distance between pairs of taxa, say between t1 and t2, is the average length of all &quot;up-down&quot; paths in the network to go from t1 to t2 (see <a href="https://doi.org/10.1007/s00285-022-01847-8">Xu &amp; Ané 2023</a> for example). In a tree, there is a single such path: going up from t1 to their most recent common ancestor, then down to t2. In a general network, there can be multiple paths. Each path has an inheritance weight: the product of the inheritance γ&#39;s of all edges in the path. Under a simple model, this is the proportion of genetic material that took this path. The average distance between t1 and t2 is the weighted average of these paths&#39; lengths, weighted by the paths&#39; inheritance.</p><p>It can be calculated with <a href="../../lib/public/#PhyloNetworks.pairwisetaxondistancematrix-Tuple{HybridNetwork}"><code>pairwisetaxondistancematrix</code></a>. On our tree <code>net0</code>, we get this:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; aveD_net0 = pairwisetaxondistancematrix(net0)</code><code class="nohighlight hljs ansi" style="display:block;">6×6 Matrix{Float64}:
  0.0  11.0  11.0  11.0  11.0  11.0
 11.0   0.0   8.0   8.0   8.0  10.0
 11.0   8.0   0.0   6.0   6.0  10.0
 11.0   8.0   6.0   0.0   2.0  10.0
 11.0   8.0   6.0   2.0   0.0  10.0
 11.0  10.0  10.0  10.0  10.0   0.0</code></pre><p>The order of rows and columns in this matrix is the same as given by <code>tiplabels</code>. For example, we can convert the distance matrix <code>aveD_net0</code> to a data frame with a column named for each taxon like this.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; taxonlist0 = tiplabels(net0)</code><code class="nohighlight hljs ansi" style="display:block;">6-element Vector{String}:
 &quot;O&quot;
 &quot;E&quot;
 &quot;D&quot;
 &quot;C&quot;
 &quot;B&quot;
 &quot;A&quot;</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; using DataFrames</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DataFrame(aveD_net0, taxonlist0)</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr1">6×6 DataFrame</span>
<span class="sgr1"> Row </span>│<span class="sgr1"> O        E        D        C        B        A       </span>
     │<span class="sgr90"> Float64  Float64  Float64  Float64  Float64  Float64 </span>
─────┼──────────────────────────────────────────────────────
   1 │     0.0     11.0     11.0     11.0     11.0     11.0
   2 │    11.0      0.0      8.0      8.0      8.0     10.0
   3 │    11.0      8.0      0.0      6.0      6.0     10.0
   4 │    11.0      8.0      6.0      0.0      2.0     10.0
   5 │    11.0      8.0      6.0      2.0      0.0     10.0
   6 │    11.0     10.0     10.0     10.0     10.0      0.0</code></pre><p>Using the network <code>net2</code>, we see that its reticulations bring E closer to D; and bring B &amp; C closer to A (and away from D &amp; E since A is distant from them):</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; aveD_net2 = pairwisetaxondistancematrix(net2);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; taxonlist2 = tiplabels(net2);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DataFrame(aveD_net2, taxonlist2)</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr1">6×6 DataFrame</span>
<span class="sgr1"> Row </span>│<span class="sgr1"> O        E        D        C        B        A       </span>
     │<span class="sgr90"> Float64  Float64  Float64  Float64  Float64  Float64 </span>
─────┼──────────────────────────────────────────────────────
   1 │     0.0    11.0      11.0    11.0     11.0      11.0
   2 │    11.0     0.0       6.5     8.44     8.44     10.0
   3 │    11.0     6.5       0.0     7.6      7.6      10.0
   4 │    11.0     8.44      7.6     0.0      2.0       7.6
   5 │    11.0     8.44      7.6     2.0      0.0       7.6
   6 │    11.0    10.0      10.0     7.6      7.6       0.0</code></pre><h2 id="hamming-distance"><a class="docs-heading-anchor" href="#hamming-distance">hamming distance</a><a id="hamming-distance-1"></a><a class="docs-heading-anchor-permalink" href="#hamming-distance" title="Permalink"></a></h2><p>To calculate pairwise distances observed in data, such as along a DNA alignment across multiple sites, we can use <a href="../../lib/internals/#PhyloNetworks.hammingdistancematrix-Union{Tuple{AbstractArray{Vector{T}, 1}}, Tuple{T}, Tuple{AbstractArray{Vector{T}, 1}, AbstractVector{&lt;:Real}}} where T"><code>PhyloNetworks.hammingdistancematrix</code></a>.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; dna_data = [
           [&#39;T&#39;,&#39;G&#39;,&#39;T&#39;,&#39;A&#39;,&#39;G&#39;], # taxon 1
           [&#39;T&#39;,&#39;G&#39;,&#39;A&#39;,&#39;A&#39;,&#39;G&#39;],
           [&#39;T&#39;,&#39;G&#39;,&#39;A&#39;,&#39;A&#39;,&#39;C&#39;],
           [&#39;T&#39;,&#39;G&#39;,&#39;A&#39;,missing,missing],
           [&#39;T&#39;,&#39;G&#39;,&#39;A&#39;,&#39;T&#39;,&#39;C&#39;], # taxon 5
       ];</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; d = PhyloNetworks.hammingdistancematrix(dna_data)</code><code class="nohighlight hljs ansi" style="display:block;">5×5 Matrix{Float64}:
 0.0  0.2  0.4  0.2  0.6
 0.2  0.0  0.2  0.0  0.4
 0.4  0.2  0.0  0.0  0.2
 0.2  0.0  0.0  0.0  0.0
 0.6  0.4  0.2  0.0  0.0</code></pre><p>Perhaps we want to give weights to each trait, such as if an alignment is summarized by keeping each site pattern once, weighted by the number of sites having this pattern:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; site_count = [3,2,1,1,1]; # invariable site patterns have higher counts</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; d = PhyloNetworks.hammingdistancematrix(dna_data, site_count)</code><code class="nohighlight hljs ansi" style="display:block;">5×5 Matrix{Float64}:
 0.0    0.125  0.25   0.125  0.375
 0.125  0.0    0.125  0.0    0.25
 0.25   0.125  0.0    0.0    0.125
 0.125  0.0    0.0    0.0    0.0
 0.375  0.25   0.125  0.0    0.0</code></pre><p>A Jukes-Cantor correction can be applied with <a href="../../lib/internals/#PhyloNetworks.distancecorrection_JC!-Tuple{AbstractMatrix, Integer}"><code>PhyloNetworks.distancecorrection_JC!</code></a>:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; PhyloNetworks.distancecorrection_JC!(d, 4) # 4 states</code><code class="nohighlight hljs ansi" style="display:block;">5×5 Matrix{Float64}:
 0.0       0.136741  0.304099  0.136741  0.51986
 0.136741  0.0       0.136741  0.0       0.304099
 0.304099  0.136741  0.0       0.0       0.136741
 0.136741  0.0       0.0       0.0       0.0
 0.51986   0.304099  0.136741  0.0       0.0</code></pre><h2 id="expected-f2-statistics"><a class="docs-heading-anchor" href="#expected-f2-statistics">expected f2-statistics</a><a id="expected-f2-statistics-1"></a><a class="docs-heading-anchor-permalink" href="#expected-f2-statistics" title="Permalink"></a></h2><p>The f2-statistic gives another measure of dissimilarity between pairs of taxa. The expected value of f2 between taxa t₁ and t₂ is</p><p class="math-container">\[f_2(t_1, t_2) = E[(X(t_1) - X(t_2))^2]\]</p><p>under a Brownian motion model for trait X evolving along the network, where X(t₁) and X(t₂) are the values of X for taxa t₁ and t₂.</p><p>In the network, branch lengths measures units of drifts when the data X are allele frequencies, and with a variance factor dependent on the allele frequency at the root. For background, see for example <a href="https://doi.org/10.1534/genetics.112.145037">Patterson et al. 2012</a> or <a href="https://doi.org/10.1111/1755-0998.13230">Lipson 2020</a>.</p><p>If the network is a tree, then this is exactly the average distance (or simply, the length of the unique path) between t₁ and t₂.</p><p>It can be calculated with <a href="../../lib/public/#PhyloNetworks.expectedf2matrix-Tuple{HybridNetwork}"><code>expectedf2matrix</code></a>. On our tree <code>net0</code>, we get an f2 matrix equal to the average distance matrix:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; f2D_net0 = expectedf2matrix(net0)</code><code class="nohighlight hljs ansi" style="display:block;">6×6 Matrix{Float64}:
  0.0  11.0  11.0  11.0  11.0  11.0
 11.0   0.0   8.0   8.0   8.0  10.0
 11.0   8.0   0.0   6.0   6.0  10.0
 11.0   8.0   6.0   0.0   2.0  10.0
 11.0   8.0   6.0   2.0   0.0  10.0
 11.0  10.0  10.0  10.0  10.0   0.0</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; f2D_net0 == aveD_net0</code><code class="nohighlight hljs ansi" style="display:block;">true</code></pre><p>Again, taxa are listed along rows and along columns in the same order as listed by <code>tiplabels()</code>.   On our network <code>net2</code>, the f2 and average distances differ:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; f2D_net2 = expectedf2matrix(net2);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DataFrame(f2D_net2, taxonlist2)</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr1">6×6 DataFrame</span>
<span class="sgr1"> Row </span>│<span class="sgr1"> O        E        D        C        B        A       </span>
     │<span class="sgr90"> Float64  Float64  Float64  Float64  Float64  Float64 </span>
─────┼──────────────────────────────────────────────────────
   1 │    0.0      9.95    11.0      9.56     9.56    11.0
   2 │    9.95     0.0      5.45     5.95     5.95     8.95
   3 │   11.0      5.45     0.0      6.16     6.16    10.0
   4 │    9.56     5.95     6.16     0.0      2.0      6.16
   5 │    9.56     5.95     6.16     2.0      0.0      6.16
   6 │   11.0      8.95    10.0      6.16     6.16     0.0</code></pre><p>In particular, we see that divergences between the outgroup O and tips B, C and E decrease because of the reticulations.</p><h2 id="expected-f4-statistics"><a class="docs-heading-anchor" href="#expected-f4-statistics">expected f4-statistics</a><a id="expected-f4-statistics-1"></a><a class="docs-heading-anchor-permalink" href="#expected-f4-statistics" title="Permalink"></a></h2><p>The f4-statistic gives a measure of dissimilarity between quartets of taxa. The expected value of f4 between taxa t₁, t₂, t₃ and t₄ is</p><p class="math-container">\[f_4(t_1, t_2; t_3, t_4) = E[(X(t_2) - X(t_1))(X(t_4) - X(t_3))]\]</p><p>under a Brownian motion model for trait X evolving along the network, where X(tᵢ) is the values of X for taxa tᵢ.</p><p>There are 4! = 24 f4 statistics relating 4 taxa, but all can be computed using only 2 of them, using the symmetries such as: <span>$f_4(t_1, t_2; t_3, t_4) = - f_4(t_1, t_2; t_4, t_3)$</span>; <span>$f_4(t_1, t_2; t_3, t_4) = f_4(t_3, t_4; t_1, t_2)$</span> and</p><p class="math-container">\[f_4(t_1, t_2; t_3, t_4) + f_4(t_1, t_3; t_4, t_2) + f_4(t_1, t_4; t_2, t_3) = 0.\]</p><p>The f4-statistics can be used to capture tree-likeness, as on a tree with no reticulation and quartet topology t₁t₂|t₃t₄, the 4-points condition writes:</p><p class="math-container">\[f_4(t_1, t_2; t_3, t_4) = 0
\quad
\text{and}
\quad
f_4(t_1, t_4; t_2, t_3) = - f_4(t_1, t_3; t_4, t_2) &gt; 0.\]</p><p>For background, see again for example <a href="https://doi.org/10.1534/genetics.112.145037">Patterson et al. 2012</a> or <a href="https://doi.org/10.1111/1755-0998.13230">Lipson 2020</a>.</p><p>The f4-statistics expected from a network can be calculated with <a href="../../lib/public/#PhyloNetworks.expectedf4table-Tuple{HybridNetwork}"><code>expectedf4table</code></a>.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; f4,t = expectedf4table(net0);</code><code class="nohighlight hljs ansi" style="display:block;">Calculation of expected f4 for 15 4-taxon sets...
0+---------------+100%
  ***************</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; t # taxa, but ordered alphabetically: not as in tiplabels(net0)</code><code class="nohighlight hljs ansi" style="display:block;">6-element Vector{String}:
 &quot;A&quot;
 &quot;B&quot;
 &quot;C&quot;
 &quot;D&quot;
 &quot;E&quot;
 &quot;O&quot;</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; first(f4,2) # first 2 4-taxon sets, each with 3 quartet f4s</code><code class="nohighlight hljs ansi" style="display:block;">2-element Vector{PhyloNetworks.QuartetT{StaticArraysCore.MVector{3, Float64}}}:
 4-taxon set number 1; taxon numbers: 1,2,3,4
data: [-2.0, 2.0, 0.0]
 4-taxon set number 2; taxon numbers: 1,2,3,5
data: [-3.0, 3.0, 0.0]</code></pre><p>The taxon numbers above are indices in the taxon list <code>t</code>. Here is a way to print all f4 statistics expected from our tree <code>net0</code>:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; for q in f4
           println(join(t[q.taxonnumber],&quot;,&quot;) * &quot;: &quot; * string(round.(q.data, sigdigits=2)))
       end</code><code class="nohighlight hljs ansi" style="display:block;">A,B,C,D: [-2.0, 2.0, 0.0]
A,B,C,E: [-3.0, 3.0, 0.0]
A,B,D,E: [-1.0, 1.0, 0.0]
A,C,D,E: [-1.0, 1.0, 0.0]
B,C,D,E: [0.0, -2.0, 2.0]
A,B,C,O: [-4.0, 4.0, 0.0]
A,B,D,O: [-2.0, 2.0, 0.0]
A,C,D,O: [-2.0, 2.0, 0.0]
B,C,D,O: [0.0, -2.0, 2.0]
A,B,E,O: [-1.0, 1.0, 0.0]
A,C,E,O: [-1.0, 1.0, 0.0]
B,C,E,O: [0.0, -3.0, 3.0]
A,D,E,O: [-1.0, 1.0, 0.0]
B,D,E,O: [0.0, -1.0, 1.0]
C,D,E,O: [0.0, -1.0, 1.0]</code></pre><p>For each set of 4 taxa,  the following f4 statistics are computed: f4(t1, t2; t3, t4), f4(t1, t3; t4, t2) and f4(t1, t4; t2, t3), so that they sum up to 0 as recalled above.</p><p>On a tree with a split <code>t1,t2|t4,t5</code>, the corresponding f4 value should be 0, and the other 2 should give ± the length of the internal path separating the 2 groups of 2 taxa. (Again, the branch lengths unit depends on the data being considered.)</p><p>Here for example, the first 4-taxon set is <code>A,B,C,D</code>. In our tree, <code>BC</code> is a clade, separated from <code>AD</code> by a branch of length 2. Accordingly, the third f4 value, corresponding to <code>AD|BC</code>, is 0; and the other two f4s are 2 or -2.</p><p><img src="../../assets/figures/expectedata_fig_net02.svg" alt="net02-distquartet-again"/></p><p>We can see how adding reticulations to our tree affects expected f4s. The columns names below indicate the specific order of taxa for each f4. This order matters for the sign.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; f4,t = expectedf4table(net2, showprogressbar=false); # same t: alphabetically</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; nt = tablequartetf4(f4, t);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; df = DataFrame(nt) # convert table to data frame</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr1">15×8 DataFrame</span>
<span class="sgr1"> Row </span>│<span class="sgr1"> qind   t1      t2      t3      t4      f4_12_34  f4_13_42  f4_14_23 </span>
     │<span class="sgr90"> Int64  String  String  String  String  Float64   Float64   Float64  </span>
─────┼─────────────────────────────────────────────────────────────────────
   1 │     1  A       B       C       D          -0.16      0.16      0.0
   2 │     2  A       B       C       E          -0.58      0.58      0.0
   3 │     3  A       B       D       E          -0.42     -1.75      2.17
   4 │     4  A       C       D       E          -0.42     -1.75      2.17
   5 │     5  B       C       D       E           0.0      -2.33      2.33
   6 │     6  A       B       C       O          -1.36      1.36      0.0
   7 │     7  A       B       D       O          -1.2       0.0       1.2
   8 │     8  A       C       D       O          -1.2       0.0       1.2
   9 │     9  B       C       D       O           0.0      -1.36      1.36
  10 │    10  A       B       E       O          -0.78     -0.42      1.2
  11 │    11  A       C       E       O          -0.78     -0.42      1.2
  12 │    12  B       C       E       O           0.0      -1.78      1.78
  13 │    13  A       D       E       O          -1.75      1.75      0.0
  14 │    14  B       D       E       O          -0.97      0.55      0.42
  15 │    15  C       D       E       O          -0.97      0.55      0.42</code></pre><p>We still have f4=0 on the 3rd column for first taxon set <code>A,B,C,D</code>, because <code>BC</code> are still sister in the network. But the last taxon set for example, <code>C,D,E,O</code>, has no 0 values of f4 due to the reticulation between ancestors of <code>D</code> and <code>E</code>: in the subnetwork for <code>C,D,E,O</code>, there is a cycle of 4 edges.</p><h2 id="expected-f3-statistics"><a class="docs-heading-anchor" href="#expected-f3-statistics">expected f3-statistics</a><a id="expected-f3-statistics-1"></a><a class="docs-heading-anchor-permalink" href="#expected-f3-statistics" title="Permalink"></a></h2><p>The f3-statistics can give a measure of similarity between pairs of taxa, relative to a given reference taxon tₒ. Using the same notations as above, the expected value of f3 between taxa t₁ and t₂ is:</p><p class="math-container">\[f_3(t_o; t_1, t_2) = E[(X(t_o) - X(t_1))(X(t_o) - X(t_2))]\]</p><p>We may calculate f3 statistics expected from a network using <a href="../../lib/internals/#PhyloNetworks.expectedf3matrix-Tuple{HybridNetwork, AbstractString}"><code>PhyloNetworks.expectedf3matrix</code></a>. Here we use the outgroup <code>O</code> as a reference taxon:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; f3_net2 = PhyloNetworks.expectedf3matrix(net2, &quot;O&quot;);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; DataFrame(f3_net2, taxonlist2)</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr1">6×6 DataFrame</span>
<span class="sgr1"> Row </span>│<span class="sgr1"> O        E        D        C        B        A       </span>
     │<span class="sgr90"> Float64  Float64  Float64  Float64  Float64  Float64 </span>
─────┼──────────────────────────────────────────────────────
   1 │     0.0     0.0      0.0      0.0      0.0       0.0
   2 │     0.0     0.0      7.75     6.78     6.78      6.0
   3 │     0.0     7.75     0.0      7.2      7.2       6.0
   4 │     0.0     6.78     7.2      0.0      8.56      7.2
   5 │     0.0     6.78     7.2      8.56     0.0       7.2
   6 │     0.0     6.0      6.0      7.2      7.2       0.0</code></pre><h2 id="quartet-concordance-factors"><a class="docs-heading-anchor" href="#quartet-concordance-factors">quartet concordance factors</a><a id="quartet-concordance-factors-1"></a><a class="docs-heading-anchor-permalink" href="#quartet-concordance-factors" title="Permalink"></a></h2><p>The concordance factor of a quartet <code>ab|cd</code> is the proportion of the genome whose genealogy has this unrooted topology.</p><p>Tools to calculate quartet concordance factors expected from a network under the coalescent model are provided in package <a href="https://github.com/JuliaPhylo/QuartetNetworkGoodnessFit.jl">QGoF</a>: see its documentation about <a href="https://juliaphylo.github.io/QuartetNetworkGoodnessFit.jl/stable/man/expected_qCFs/#expected-concordance-factors">expected concordance factors</a>.</p><p>To calculate quartet concordance factors observed in data, one option is to count the number of gene trees that display each quartet, using <a href="../../lib/public/#PhyloNetworks.countquartetsintrees"><code>countquartetsintrees</code></a> and <a href="../../lib/public/#PhyloNetworks.tablequartetCF-Union{Tuple{Array{PhyloNetworks.QuartetT{T}, 1}}, Tuple{T}, Tuple{Array{PhyloNetworks.QuartetT{T}, 1}, AbstractVector{&lt;:AbstractString}}} where T&lt;:Union{StaticArraysCore.StaticArray{Tuple{3}, T, 1} where T, StaticArraysCore.StaticArray{Tuple{4}, T, 1} where T, StaticArraysCore.StaticArray{Tuple{3, N}, T, 2} where {N, T}}"><code>tablequartetCF</code></a>.</p><p>In the example below, the 4th gene tree is missing taxon A, and the 6th gene tree has a polytomies (unresolved ABE clade), such as if a branch of low support was collapsed. The number of genes underlying each quartet is captured in the table below.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; sixgenetrees_nwk = [
         &quot;(E,((A,B),(C,D)),O);&quot;,&quot;(((A,B),(C,D)),(E,O));&quot;,&quot;(A,B,((C,D),(E,O)));&quot;,
         &quot;(B,((C,D),(E,O)));&quot;,&quot;((C,D),(A,(B,E)),O);&quot;,&quot;((C,D),(A,B,E),O);&quot;];</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; genetrees = readnewick.(sixgenetrees_nwk);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; q,t = countquartetsintrees(genetrees, showprogressbar=false);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; df = tablequartetCF(q,t) |&gt; DataFrame</code><code class="nohighlight hljs ansi" style="display:block;"><span class="sgr1">15×9 DataFrame</span>
<span class="sgr1"> Row </span>│<span class="sgr1"> qind   t1      t2      t3      t4      CF12_34   CF13_24   CF14_23  ngenes  </span>
     │<span class="sgr90"> Int64  String  String  String  String  Float64   Float64   Float64  Float64 </span>
─────┼─────────────────────────────────────────────────────────────────────────────
   1 │     1  A       B       C       D       1.0       0.0          0.0       5.0
   2 │     2  A       B       C       E       0.75      0.25         0.0       4.0
   3 │     3  A       B       D       E       0.75      0.25         0.0       4.0
   4 │     4  A       C       D       E       0.0       0.0          1.0       5.0
   5 │     5  B       C       D       E       0.0       0.0          1.0       6.0
   6 │     6  A       B       C       O       1.0       0.0          0.0       5.0
   7 │     7  A       B       D       O       1.0       0.0          0.0       5.0
   8 │     8  A       C       D       O       0.0       0.0          1.0       5.0
   9 │     9  B       C       D       O       0.0       0.0          1.0       6.0
  10 │    10  A       B       E       O       0.75      0.0          0.25      4.0
  11 │    11  A       C       E       O       0.6       0.4          0.0       5.0
  12 │    12  B       C       E       O       0.666667  0.333333     0.0       6.0
  13 │    13  A       D       E       O       0.6       0.4          0.0       5.0
  14 │    14  B       D       E       O       0.666667  0.333333     0.0       6.0
  15 │    15  C       D       E       O       1.0       0.0          0.0       6.0</code></pre><p>If low-support branches are not collapsed, this counting method does not account for gene tree estimation error. (It biases concordance factors towards 1/3 for each resolution of a 4-taxon tree: lack of knowledge is mistaken as lack of concordance). Estimation error can be accounted for in a Bayesian framework: see <a href="https://juliaphylo.github.io/PhyloUtilities/">PhyloUtilities</a> for a pipeline.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../nj/">« Neighbour Joining</a><a class="docs-footer-nextpage" href="../../lib/public/">Public »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.11.4 on <span class="colophon-date" title="Wednesday 28 May 2025 14:19">Wednesday 28 May 2025</span>. Using Julia version 1.11.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
