



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="PhyloNetworks is a Julia package for the manipulation, visualization and inference of phylogenetic networks.">
      
      
      
        <meta name="author" content="Claudia Sol&iacute;s-Lemus">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-0.17.3, mkdocs-material-2.7.1">
    
    
      
        <title>Public - PhyloNetworks.jl</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/application.78aab2dc.css">
      
        <link rel="stylesheet" href="../../assets/stylesheets/application-palette.6079476c.css">
      
    
    
      <script src="../../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    
    
      <link rel="stylesheet" href="../../assets/Documenter.css">
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="teal" data-md-color-accent="teal">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="drawer"></label>
    
      <a href="#public-documentation" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../.." title="PhyloNetworks.jl" class="md-header-nav__button md-logo">
          
            <img src="../../snaq_small.png" width="24" height="24">
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                PhyloNetworks.jl
              </span>
              <span class="md-header-nav__topic">
                Public
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/crsl4/PhyloNetworks.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="drawer">
    <span class="md-nav__button md-logo">
      
        <img src="../../snaq_small.png" width="48" height="48">
      
    </span>
    PhyloNetworks.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/crsl4/PhyloNetworks.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../.." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2">
    
    <label class="md-nav__link" for="nav-2">
      Manual
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        Manual
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../man/installation/" title="Installation" class="md-nav__link">
      Installation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../man/inputdata/" title="Input Data for SNaQ" class="md-nav__link">
      Input Data for SNaQ
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../man/ticr_howtogetQuartetCFs/" title="TICR pipeline" class="md-nav__link">
      TICR pipeline
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../man/snaq_plot/" title="Network estimation and display" class="md-nav__link">
      Network estimation and display
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../man/dist_reroot/" title="Network comparison and manipulation" class="md-nav__link">
      Network comparison and manipulation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../man/fixednetworkoptim/" title="Candidate Networks" class="md-nav__link">
      Candidate Networks
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../man/expectedCFs/" title="Extract Expected CFs" class="md-nav__link">
      Extract Expected CFs
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../man/bootstrap/" title="Bootstrap" class="md-nav__link">
      Bootstrap
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../man/multiplealleles/" title="Multiple Alleles" class="md-nav__link">
      Multiple Alleles
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../man/trait_tree/" title="Continuous Trait Evolution" class="md-nav__link">
      Continuous Trait Evolution
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3" checked>
    
    <label class="md-nav__link" for="nav-3">
      Library
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        Library
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="toc">
        Public
      </label>
    
    <a href="./" title="Public" class="md-nav__link md-nav__link--active">
      Public
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#index" title="Index" class="md-nav__link">
    Index
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#types" title="types" class="md-nav__link">
    types
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utilities" title="utilities" class="md-nav__link">
    utilities
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-and-topology-readwrite" title="data and topology read/write" class="md-nav__link">
    data and topology read/write
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#network-inference" title="network inference" class="md-nav__link">
    network inference
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#network-comparisons" title="network Comparisons" class="md-nav__link">
    network Comparisons
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#continuous-trait-evolution" title="continuous trait evolution" class="md-nav__link">
    continuous trait evolution
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#discrete-trait-evolution" title="discrete trait evolution" class="md-nav__link">
    discrete trait evolution
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../internals/" title="Internals" class="md-nav__link">
      Internals
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#index" title="Index" class="md-nav__link">
    Index
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#types" title="types" class="md-nav__link">
    types
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#utilities" title="utilities" class="md-nav__link">
    utilities
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-and-topology-readwrite" title="data and topology read/write" class="md-nav__link">
    data and topology read/write
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#network-inference" title="network inference" class="md-nav__link">
    network inference
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#network-comparisons" title="network Comparisons" class="md-nav__link">
    network Comparisons
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#continuous-trait-evolution" title="continuous trait evolution" class="md-nav__link">
    continuous trait evolution
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#discrete-trait-evolution" title="discrete trait evolution" class="md-nav__link">
    discrete trait evolution
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/crsl4/PhyloNetworks.jl/edit/master/docs/lib/public.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <p><a id='Public-Documentation-1'></a></p>
<h1 id="public-documentation">Public Documentation</h1>
<p>Documentation for <code>PhyloNetworks</code>'s public (exported) interface.</p>
<p>See <a href="../internals/#Internal-Documentation-1">Internal Documentation</a> for documentation on internal functions.</p>
<ul>
<li><a href="./#Public-Documentation-1">Public Documentation</a><ul>
<li><a href="./#Index-1">Index</a></li>
<li><a href="./#types-1">types</a></li>
<li><a href="./#utilities-1">utilities</a></li>
<li><a href="./#data-and-topology-read/write-1">data and topology read/write</a></li>
<li><a href="./#network-inference-1">network inference</a></li>
<li><a href="./#network-Comparisons-1">network Comparisons</a></li>
<li><a href="./#continuous-trait-evolution-1">continuous trait evolution</a></li>
<li><a href="./#discrete-trait-evolution-1">discrete trait evolution</a></li>
</ul>
</li>
</ul>
<p><a id='Index-1'></a></p>
<h2 id="index">Index</h2>
<ul>
<li><a href="./#PhyloNetworks.BinaryTraitSubstitutionModel"><code>PhyloNetworks.BinaryTraitSubstitutionModel</code></a></li>
<li><a href="./#PhyloNetworks.DataCF"><code>PhyloNetworks.DataCF</code></a></li>
<li><a href="./#PhyloNetworks.EqualRatesSubstitutionModel"><code>PhyloNetworks.EqualRatesSubstitutionModel</code></a></li>
<li><a href="./#PhyloNetworks.HybridNetwork"><code>PhyloNetworks.HybridNetwork</code></a></li>
<li><a href="./#PhyloNetworks.ParamsBM"><code>PhyloNetworks.ParamsBM</code></a></li>
<li><a href="./#PhyloNetworks.PhyloNetworkLinearModel"><code>PhyloNetworks.PhyloNetworkLinearModel</code></a></li>
<li><a href="./#PhyloNetworks.Quartet"><code>PhyloNetworks.Quartet</code></a></li>
<li><a href="./#PhyloNetworks.ReconstructedStates"><code>PhyloNetworks.ReconstructedStates</code></a></li>
<li><a href="./#PhyloNetworks.ShiftNet"><code>PhyloNetworks.ShiftNet</code></a></li>
<li><a href="./#PhyloNetworks.TraitSimulation"><code>PhyloNetworks.TraitSimulation</code></a></li>
<li><a href="./#PhyloNetworks.TraitSubstitutionModel"><code>PhyloNetworks.TraitSubstitutionModel</code></a></li>
<li><a href="./#PhyloNetworks.TwoBinaryTraitSubstitutionModel"><code>PhyloNetworks.TwoBinaryTraitSubstitutionModel</code></a></li>
<li><a href="./#Base.getindex-Tuple{PhyloNetworks.TraitSimulation,Symbol}"><code>Base.getindex</code></a></li>
<li><a href="./#PhyloNetworks.P"><code>PhyloNetworks.P</code></a></li>
<li><a href="./#PhyloNetworks.Q"><code>PhyloNetworks.Q</code></a></li>
<li><a href="./#PhyloNetworks.ancestralStateReconstruction"><code>PhyloNetworks.ancestralStateReconstruction</code></a></li>
<li><a href="./#PhyloNetworks.biconnectedComponents"><code>PhyloNetworks.biconnectedComponents</code></a></li>
<li><a href="./#PhyloNetworks.bootsnaq"><code>PhyloNetworks.bootsnaq</code></a></li>
<li><a href="./#PhyloNetworks.calibrateFromPairwiseDistances!"><code>PhyloNetworks.calibrateFromPairwiseDistances!</code></a></li>
<li><a href="./#PhyloNetworks.cladewiseorder!"><code>PhyloNetworks.cladewiseorder!</code></a></li>
<li><a href="./#PhyloNetworks.deleteHybridThreshold!"><code>PhyloNetworks.deleteHybridThreshold!</code></a></li>
<li><a href="./#PhyloNetworks.deleteleaf!"><code>PhyloNetworks.deleteleaf!</code></a></li>
<li><a href="./#PhyloNetworks.descendenceMatrix"><code>PhyloNetworks.descendenceMatrix</code></a></li>
<li><a href="./#PhyloNetworks.directEdges!"><code>PhyloNetworks.directEdges!</code></a></li>
<li><a href="./#PhyloNetworks.displayedNetworkAt!"><code>PhyloNetworks.displayedNetworkAt!</code></a></li>
<li><a href="./#PhyloNetworks.displayedTrees"><code>PhyloNetworks.displayedTrees</code></a></li>
<li><a href="./#PhyloNetworks.expectations"><code>PhyloNetworks.expectations</code></a></li>
<li><a href="./#PhyloNetworks.expectationsPlot"><code>PhyloNetworks.expectationsPlot</code></a></li>
<li><a href="./#PhyloNetworks.fittedQuartetCF"><code>PhyloNetworks.fittedQuartetCF</code></a></li>
<li><a href="./#PhyloNetworks.getNodeAges"><code>PhyloNetworks.getNodeAges</code></a></li>
<li><a href="./#PhyloNetworks.getShiftEdgeNumber"><code>PhyloNetworks.getShiftEdgeNumber</code></a></li>
<li><a href="./#PhyloNetworks.getShiftValue"><code>PhyloNetworks.getShiftValue</code></a></li>
<li><a href="./#PhyloNetworks.hardwiredCluster"><code>PhyloNetworks.hardwiredCluster</code></a></li>
<li><a href="./#PhyloNetworks.hardwiredClusterDistance"><code>PhyloNetworks.hardwiredClusterDistance</code></a></li>
<li><a href="./#PhyloNetworks.hardwiredClusters"><code>PhyloNetworks.hardwiredClusters</code></a></li>
<li><a href="./#PhyloNetworks.hybridBootstrapSupport"><code>PhyloNetworks.hybridBootstrapSupport</code></a></li>
<li><a href="./#PhyloNetworks.hybridatnode!"><code>PhyloNetworks.hybridatnode!</code></a></li>
<li><a href="./#PhyloNetworks.lambda_estim"><code>PhyloNetworks.lambda_estim</code></a></li>
<li><a href="./#PhyloNetworks.majorTree"><code>PhyloNetworks.majorTree</code></a></li>
<li><a href="./#PhyloNetworks.mapAllelesCFtable"><code>PhyloNetworks.mapAllelesCFtable</code></a></li>
<li><a href="./#PhyloNetworks.minorTreeAt"><code>PhyloNetworks.minorTreeAt</code></a></li>
<li><a href="./#PhyloNetworks.mu_estim"><code>PhyloNetworks.mu_estim</code></a></li>
<li><a href="./#PhyloNetworks.nStates"><code>PhyloNetworks.nStates</code></a></li>
<li><a href="./#PhyloNetworks.pairwiseTaxonDistanceMatrix"><code>PhyloNetworks.pairwiseTaxonDistanceMatrix</code></a></li>
<li><a href="./#PhyloNetworks.parsimonyGF"><code>PhyloNetworks.parsimonyGF</code></a></li>
<li><a href="./#PhyloNetworks.parsimonySoftwired"><code>PhyloNetworks.parsimonySoftwired</code></a></li>
<li><a href="./#PhyloNetworks.phyloNetworklm"><code>PhyloNetworks.phyloNetworklm</code></a></li>
<li><a href="./#PhyloNetworks.predint"><code>PhyloNetworks.predint</code></a></li>
<li><a href="./#PhyloNetworks.predintPlot"><code>PhyloNetworks.predintPlot</code></a></li>
<li><a href="./#PhyloNetworks.preorder!"><code>PhyloNetworks.preorder!</code></a></li>
<li><a href="./#PhyloNetworks.printEdges"><code>PhyloNetworks.printEdges</code></a></li>
<li><a href="./#PhyloNetworks.printNodes"><code>PhyloNetworks.printNodes</code></a></li>
<li><a href="./#PhyloNetworks.randomTrait"><code>PhyloNetworks.randomTrait</code></a></li>
<li><a href="./#PhyloNetworks.readBootstrapTrees"><code>PhyloNetworks.readBootstrapTrees</code></a></li>
<li><a href="./#PhyloNetworks.readInputTrees"><code>PhyloNetworks.readInputTrees</code></a></li>
<li><a href="./#PhyloNetworks.readMultiTopology"><code>PhyloNetworks.readMultiTopology</code></a></li>
<li><a href="./#PhyloNetworks.readSnaqNetwork"><code>PhyloNetworks.readSnaqNetwork</code></a></li>
<li><a href="./#PhyloNetworks.readTableCF"><code>PhyloNetworks.readTableCF</code></a></li>
<li><a href="./#PhyloNetworks.readTableCF!"><code>PhyloNetworks.readTableCF!</code></a></li>
<li><a href="./#PhyloNetworks.readTopology"><code>PhyloNetworks.readTopology</code></a></li>
<li><a href="./#PhyloNetworks.readTopologyLevel1"><code>PhyloNetworks.readTopologyLevel1</code></a></li>
<li><a href="./#PhyloNetworks.readTrees2CF"><code>PhyloNetworks.readTrees2CF</code></a></li>
<li><a href="./#PhyloNetworks.regressorHybrid"><code>PhyloNetworks.regressorHybrid</code></a></li>
<li><a href="./#PhyloNetworks.regressorShift"><code>PhyloNetworks.regressorShift</code></a></li>
<li><a href="./#PhyloNetworks.rootatnode!"><code>PhyloNetworks.rootatnode!</code></a></li>
<li><a href="./#PhyloNetworks.rootonedge!"><code>PhyloNetworks.rootonedge!</code></a></li>
<li><a href="./#PhyloNetworks.rotate!"><code>PhyloNetworks.rotate!</code></a></li>
<li><a href="./#PhyloNetworks.setGamma!"><code>PhyloNetworks.setGamma!</code></a></li>
<li><a href="./#PhyloNetworks.setLength!"><code>PhyloNetworks.setLength!</code></a></li>
<li><a href="./#PhyloNetworks.sharedPathMatrix"><code>PhyloNetworks.sharedPathMatrix</code></a></li>
<li><a href="./#PhyloNetworks.shiftHybrid"><code>PhyloNetworks.shiftHybrid</code></a></li>
<li><a href="./#PhyloNetworks.sigma2_estim"><code>PhyloNetworks.sigma2_estim</code></a></li>
<li><a href="./#PhyloNetworks.simulate"><code>PhyloNetworks.simulate</code></a></li>
<li><a href="./#PhyloNetworks.snaq!"><code>PhyloNetworks.snaq!</code></a></li>
<li><a href="./#PhyloNetworks.snaqDebug"><code>PhyloNetworks.snaqDebug</code></a></li>
<li><a href="./#PhyloNetworks.sorttaxa!"><code>PhyloNetworks.sorttaxa!</code></a></li>
<li><a href="./#PhyloNetworks.summarizeDataCF"><code>PhyloNetworks.summarizeDataCF</code></a></li>
<li><a href="./#PhyloNetworks.tipLabels"><code>PhyloNetworks.tipLabels</code></a></li>
<li><a href="./#PhyloNetworks.topologyMaxQPseudolik!"><code>PhyloNetworks.topologyMaxQPseudolik!</code></a></li>
<li><a href="./#PhyloNetworks.topologyQPseudolik!"><code>PhyloNetworks.topologyQPseudolik!</code></a></li>
<li><a href="./#PhyloNetworks.treeEdgesBootstrap"><code>PhyloNetworks.treeEdgesBootstrap</code></a></li>
<li><a href="./#PhyloNetworks.writeMultiTopology"><code>PhyloNetworks.writeMultiTopology</code></a></li>
<li><a href="./#PhyloNetworks.writeSubTree!"><code>PhyloNetworks.writeSubTree!</code></a></li>
<li><a href="./#PhyloNetworks.writeTopology"><code>PhyloNetworks.writeTopology</code></a></li>
</ul>
<p><a id='types-1'></a></p>
<h2 id="types">types</h2>
<p><a id='PhyloNetworks.BinaryTraitSubstitutionModel' href='#PhyloNetworks.BinaryTraitSubstitutionModel'>#</a>
<strong><code>PhyloNetworks.BinaryTraitSubstitutionModel</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>BinaryTraitSubstitutionModel(α, β [, label])
</pre></div>


<p><a href="./#PhyloNetworks.TraitSubstitutionModel"><code>TraitSubstitutionModel</code></a> for binary traits (with 2 states). Default labels are "0" and "1". α is the rate of transition from "0" to "1", and β from "1" to "0".</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L89-L95' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.DataCF' href='#PhyloNetworks.DataCF'>#</a>
<strong><code>PhyloNetworks.DataCF</code></strong> &mdash; <em>Type</em>.</p>
<p><code>DataCF type</code></p>
<p>type that contains the following attributes:</p>
<ul>
<li>quartet (vector of Quartets)</li>
<li>numQuartets</li>
<li>tree (vector of trees: empty if a table of CF was input instead of list of trees)</li>
<li>numTrees (-1 if a table CF was input instead of list of trees)</li>
<li>repSpecies (taxon names that were repeated in table of CF or input gene trees: used inside snaq for multiple alleles case)</li>
</ul>
<p>The list of Quartet may be accessed with the attribute .quartet. If the input was a list of trees, the HybridNetwork's can be accessed with the attribute .tree. For example, if the DataCF object is named d, d.quartet[1] will show the first quartet and d.tree[1] will print the first input tree.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/types.jl#L241-L256' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.EqualRatesSubstitutionModel' href='#PhyloNetworks.EqualRatesSubstitutionModel'>#</a>
<strong><code>PhyloNetworks.EqualRatesSubstitutionModel</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>EqualRatesSubstitutionModel(numberStates, α, labels)
</pre></div>


<p><a href="./#PhyloNetworks.TraitSubstitutionModel"><code>TraitSubstitutionModel</code></a> for traits with any number of states and equal substitution rates α between all states. Default labels are "1","2",...</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L219-L225' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.HybridNetwork' href='#PhyloNetworks.HybridNetwork'>#</a>
<strong><code>PhyloNetworks.HybridNetwork</code></strong> &mdash; <em>Type</em>.</p>
<p><code>HybridNetwork type</code> Explicit network or tree with the following attributes:</p>
<ul>
<li>numTaxa</li>
<li>numNodes (total number of nodes)</li>
<li>numEdges</li>
<li>numHybrids (number of hybrid nodes)</li>
<li>edge (array of Edges)</li>
<li>node (array of Nodes)</li>
<li>root (index of root in vector 'node'. May be artificial, for printing and traversal purposes only.)</li>
<li>hybrid (array of Nodes: those are are hybrid nodes)</li>
<li>leaf (array of Nodes: those that are leaves)</li>
<li>loglik (negative log pseudolik after estimation)</li>
<li>isRooted (true or false)</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/types.jl#L102-L117' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.ParamsBM' href='#PhyloNetworks.ParamsBM'>#</a>
<strong><code>PhyloNetworks.ParamsBM</code></strong> &mdash; <em>Type</em>.</p>
<p><code>ParamsBM &lt;: ParamsProcess</code></p>
<p>Type for a BM process on a network. Fields are <code>mu</code> (expectation), <code>sigma2</code> (variance), <code>randomRoot</code> (whether the root is random, default to <code>false</code>), and <code>varRoot</code> (if the root is random, the variance of the root, defalut to <code>NaN</code>).</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L763-L770' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.PhyloNetworkLinearModel' href='#PhyloNetworks.PhyloNetworkLinearModel'>#</a>
<strong><code>PhyloNetworks.PhyloNetworkLinearModel</code></strong> &mdash; <em>Type</em>.</p>
<p><code>PhyloNetworkLinearModel&lt;:LinPredModel</code></p>
<p>Regression object for a phylogenetic regression. Result of fitting function <a href="./#PhyloNetworks.phyloNetworklm"><code>phyloNetworklm</code></a>. Dominated by the <code>LinPredModel</code> class, from package <code>GLM</code>.</p>
<p>The following StatsBase functions can be applied to it: <code>coef</code>, <code>nobs</code>, <code>vcov</code>, <code>stderr</code>, <code>confint</code>, <code>coeftable</code>, <code>dof_residual</code>, <code>dof</code>, <code>deviance</code>, <code>residuals</code>, <code>model_response</code>, <code>predict</code>, <code>loglikelihood</code>, <code>nulldeviance</code>, <code>nullloglikelihood</code>, <code>r2</code>, <code>adjr2</code>, <code>aic</code>, <code>aicc</code>, <code>bic</code>.</p>
<p>The following StatsModels functions can also be applied to it: <code>ModelFrame</code>, <code>ModelMatrix</code>, <code>Formula</code>.</p>
<p>Estimated variance and mean of the BM process used can be retrieved with functions <a href="./#PhyloNetworks.sigma2_estim"><code>sigma2_estim</code></a> and <a href="./#PhyloNetworks.mu_estim"><code>mu_estim</code></a>.</p>
<p>If a Pagel's lambda model is fitted, the parameter can be retrieved with function <a href="./#PhyloNetworks.lambda_estim"><code>lambda_estim</code></a>.</p>
<p>An ancestral state reconstruction can be performed from this fitted object using function: <a href="./#PhyloNetworks.ancestralStateReconstruction"><code>ancestralStateReconstruction</code></a>.</p>
<p>The <code>PhyloNetworkLinearModel</code> object has fields: <code>lm</code>, <code>V</code>, <code>Vy</code>, <code>RL</code>, <code>Y</code>, <code>X</code>, <code>logdetVy</code>, <code>ind</code>, <code>msng</code>, <code>model</code>, <code>lambda</code>. Type in "?PhyloNetworkLinearModel.field" to get help on a specific field.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1034-L1059' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.Quartet' href='#PhyloNetworks.Quartet'>#</a>
<strong><code>PhyloNetworks.Quartet</code></strong> &mdash; <em>Type</em>.</p>
<p><code>Quartet type</code></p>
<p>type that saves the information on a given 4-taxon subset. It contains the following attributes:</p>
<ul>
<li>number</li>
<li>taxon (vector of taxon names)</li>
<li>obsCF (vector of observed CF)</li>
<li>logPseudoLik</li>
<li>ngenes (number of gene trees used to compute the observed CF: -1 if unknown)</li>
<li>qnet (internal topological structure that saves the expCF after snaq estimation to emphasize that the expCF depend on a specific network, not the data)</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/types.jl#L202-L213' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.ReconstructedStates' href='#PhyloNetworks.ReconstructedStates'>#</a>
<strong><code>PhyloNetworks.ReconstructedStates</code></strong> &mdash; <em>Type</em>.</p>
<p><code>ReconstructedStates</code></p>
<p>Type containing the inferred information about the law of the ancestral states given the observed tips values. The missing tips are considered as ancestral states.</p>
<p>The following functions can be applied to it: <a href="./#PhyloNetworks.expectations"><code>expectations</code></a> (vector of expectations at all nodes), <code>stderr</code> (the standard error), <code>predint</code> (the prediction interval).</p>
<p>The <code>ReconstructedStates</code> object has fields: <code>traits_nodes</code>, <code>variances_nodes</code>, <code>NodeNumbers</code>, <code>traits_tips</code>, <code>tipNumbers</code>, <code>model</code>. Type in "?ReconstructedStates.field" to get help on a specific field.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1836-L1848' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.ShiftNet' href='#PhyloNetworks.ShiftNet'>#</a>
<strong><code>PhyloNetworks.ShiftNet</code></strong> &mdash; <em>Type</em>.</p>
<p><code>ShiftNet</code></p>
<p>Shifts associated to a <a href="./#PhyloNetworks.HybridNetwork"><code>HybridNetwork</code></a> sorted in topological order. Its <code>shift</code> field is a vector of shift values, one for each node, corresponding to the shift on the parent edge of the node (which makes sense for tree nodes only: they have a single parent edge).</p>
<p>Two <code>ShiftNet</code> objects on the same network can be concatened with <code>*</code>.</p>
<p><code>ShiftNet(node::Vector{Node}, value::Vector{T} where T&lt;:Real, net::HybridNetwork; checkPreorder=true::Bool)</code></p>
<p>Constructor from a vector of nodes and associated values. The shifts are located on the edges above the nodes provided. Warning, shifts on hybrid edges are not allowed.</p>
<p><code>ShiftNet(edge::Vector{Edge}, value::Vector{T} where T&lt;:Real, net::HybridNetwork; checkPreorder=true::Bool)</code></p>
<p>Constructor from a vector of edges and associated values. Warning, shifts on hybrid edges are not allowed.</p>
<p>Extractors: <a href="./#PhyloNetworks.getShiftEdgeNumber"><code>getShiftEdgeNumber</code></a>, <a href="./#PhyloNetworks.getShiftValue"><code>getShiftValue</code></a></p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L626-L648' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.TraitSimulation' href='#PhyloNetworks.TraitSimulation'>#</a>
<strong><code>PhyloNetworks.TraitSimulation</code></strong> &mdash; <em>Type</em>.</p>
<p><code>TraitSimulation</code></p>
<p>Result of a trait simulation on an <a href="./#PhyloNetworks.HybridNetwork"><code>HybridNetwork</code></a> with function <a href="./#PhyloNetworks.simulate"><code>simulate</code></a>.</p>
<p>The following functions and extractors can be applied to it: <a href="./#PhyloNetworks.tipLabels"><code>tipLabels</code></a>, <code>obj[:Tips]</code>, <code>obj[:InternalNodes]</code> (see documentation for function <a href="./#Base.getindex-Tuple{PhyloNetworks.TraitSimulation,Symbol}"><code>getindex(::TraitSimulation, ::Symbol)</code></a>).</p>
<p>The <code>TraitSimulation</code> object has fields: <code>M</code>, <code>params</code>, <code>model</code>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L822-L830' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.TraitSubstitutionModel' href='#PhyloNetworks.TraitSubstitutionModel'>#</a>
<strong><code>PhyloNetworks.TraitSubstitutionModel</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>TraitSubstitutionModel
</pre></div>


<p>Abstract type for discrete trait substitution models, using a continous time Markov model on a phylogeny. Adapted from the substitutionModels module in BioJulia. The same Q and P function names are used for the transition rates and probabilities.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L1-L8' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.TwoBinaryTraitSubstitutionModel' href='#PhyloNetworks.TwoBinaryTraitSubstitutionModel'>#</a>
<strong><code>PhyloNetworks.TwoBinaryTraitSubstitutionModel</code></strong> &mdash; <em>Type</em>.</p>
<div class="codehilite"><pre><span></span>TwoBinaryTraitSubstitutionModel(rate [, label])
</pre></div>


<p><a href="./#PhyloNetworks.TraitSubstitutionModel"><code>TraitSubstitutionModel</code></a> for two binary traits, possibly correlated. Default labels are "x0", "x1" for trait 1, and "y0", "y1" for trait 2. If provided, <code>label</code> should be a vector of size 4, listing labels for trait 1 first then labels for trait 2. <code>rate</code> should be a vector of substitution rates of size 8. rate[1],...,rate[4] describe rates of changes in trait 1. rate[5],...,rate[8] describe rates of changes in trait 2.</p>
<p>In the transition matrix, trait combinations are listed in the following order: x0-y0, x0-y1, x1-y0, x1-y1.</p>
<p><strong>example</strong></p>
<div class="codehilite"><pre><span></span><span class="n">model</span> <span class="o">=</span> <span class="n">TwoBinaryTraitSubstitutionModel</span><span class="p">([</span><span class="mf">2.0</span><span class="p">,</span><span class="mf">1.2</span><span class="p">,</span><span class="mf">1.1</span><span class="p">,</span><span class="mf">2.2</span><span class="p">,</span><span class="mf">1.0</span><span class="p">,</span><span class="mf">3.1</span><span class="p">,</span><span class="mf">2.0</span><span class="p">,</span><span class="mf">1.1</span><span class="p">],</span>
        <span class="p">[</span><span class="s">&quot;carnivory&quot;</span><span class="p">,</span> <span class="s">&quot;noncarnivory&quot;</span><span class="p">,</span> <span class="s">&quot;wet&quot;</span><span class="p">,</span> <span class="s">&quot;dry&quot;</span><span class="p">]);</span>
<span class="n">model</span>
<span class="k">using</span> <span class="n">PhyloPlots</span>
<span class="n">plot</span><span class="p">(</span><span class="n">model</span><span class="p">)</span> <span class="c"># to visualize states and rates</span>
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L158-L181' class='documenter-source'>source</a><br></p>
<p><a id='utilities-1'></a></p>
<h2 id="utilities">utilities</h2>
<p><a id='PhyloNetworks.tipLabels' href='#PhyloNetworks.tipLabels'>#</a>
<strong><code>PhyloNetworks.tipLabels</code></strong> &mdash; <em>Function</em>.</p>
<p><code>tipLabels(net::HybridNetwork)</code></p>
<p>returns a vector of taxon names (at the leaves) from a HybridNetwork object</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/descriptive.jl#L4-L8' class='documenter-source'>source</a><br></p>
<p><code>tipLabels(obj::TraitSimulation)</code> returns a vector of taxon names (at the leaves) for a simulated object.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L844-L847' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.sorttaxa!' href='#PhyloNetworks.sorttaxa!'>#</a>
<strong><code>PhyloNetworks.sorttaxa!</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>sorttaxa!(DataFrame, columns)
</pre></div>


<p>Reorder the 4 taxa and reorders the observed concordance factors accordingly, on each row of the data frame. If <code>columns</code> is committed, taxon names are assumed to be in columns 1-4 and CFs are assumed to be in columns 5-6 with quartets in this order: 12_34, 13_24, 14_23. Does <strong>not</strong> reorder credibility interval values, if present.</p>
<div class="codehilite"><pre><span></span>sorttaxa!(DataCF)
sorttaxa!(Quartet, permutation_tax, permutation_cf)
</pre></div>


<p>Reorder the 4 taxa in each element of the DataCF <code>quartet</code>. For a given Quartet, reorder the 4 taxa in its fields <code>taxon</code> and <code>qnet.quartetTaxon</code> (if non-empty) and reorder the 3 concordance values accordingly, in <code>obsCF</code> and <code>qnet.expCF</code>.</p>
<p><code>permutation_tax</code> and <code>permutation_cf</code> should be vectors of short integers (Int8) of length 4 and 3 respectively, whose memory allocation gets reused. Their length is <em>not checked</em>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/auxiliary.jl#L1225-L1242' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.printEdges' href='#PhyloNetworks.printEdges'>#</a>
<strong><code>PhyloNetworks.printEdges</code></strong> &mdash; <em>Function</em>.</p>
<p><code>printEdges(net::HybridNetwork)</code></p>
<p>prints the information on the edges of net: edge number, node numbers of nodes attached to it, in which cycle it is contained (-1 if no cycle), can it contain root, is it an identifiable edge, length, is it hybrid, gamma value</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/auxiliary.jl#L652-L656' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.printNodes' href='#PhyloNetworks.printNodes'>#</a>
<strong><code>PhyloNetworks.printNodes</code></strong> &mdash; <em>Function</em>.</p>
<p><code>printNodes(net::HybridNetwork)</code></p>
<p>prints information on the nodes of net: node number, in which cycle it is contained (-1 if no cycle), is it hybrid, does it has hybrid edges, edges number attached to it</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/auxiliary.jl#L670-L674' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.summarizeDataCF' href='#PhyloNetworks.summarizeDataCF'>#</a>
<strong><code>PhyloNetworks.summarizeDataCF</code></strong> &mdash; <em>Function</em>.</p>
<p><code>summarizeDataCF(d::DataCF)</code></p>
<p>function to summarize the information contained in a DataCF object. It has the following optional arguments:</p>
<ul>
<li>filename: if provided, the summary will be saved in the filename, not to screen</li>
<li>pc (number between (0,1)): threshold of percentage of missing genes to identify 4-taxon subsets with fewer genes than the threshold</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readData.jl#L746-L752' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.directEdges!' href='#PhyloNetworks.directEdges!'>#</a>
<strong><code>PhyloNetworks.directEdges!</code></strong> &mdash; <em>Function</em>.</p>
<p><code>directEdges!(net::HybridNetwork; checkMajor=true::Bool)</code></p>
<p>Updates the edges' attribute <code>isChild1</code>, according to the root placement. Also updates edges' attribute <code>containRoot</code>, for other possible root placements compatible with the direction of existing hybrid edges. Relies on hybrid nodes having exactly 1 major hybrid parent edge, but checks for that if checkMajor=true.</p>
<p>Warning: Assumes that isChild1 is correct on hybrid edges (to avoid changing the identity of which nodes are hybrids and which are not).</p>
<p>Returns the network. Throws a 'RootMismatch' Exception if the root was found to conflict with the direction of any hybrid edge.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/manipulateNet.jl#L639-L653' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.preorder!' href='#PhyloNetworks.preorder!'>#</a>
<strong><code>PhyloNetworks.preorder!</code></strong> &mdash; <em>Function</em>.</p>
<p><code>preorder!(net::HybridNetwork)</code></p>
<p>Updates attribute net.nodes_changed in which the nodes are pre-ordered (also called topological sorting), such that each node is visited after its parent(s). The edges' direction needs to be correct before calling preorder!, using directEdges!</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/manipulateNet.jl#L772-L778' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.cladewiseorder!' href='#PhyloNetworks.cladewiseorder!'>#</a>
<strong><code>PhyloNetworks.cladewiseorder!</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>cladewiseorder!(net::HybridNetwork)
</pre></div>


<p>Updates attribute net.cladewiseorder_nodeIndex. Used for plotting the network. In the major tree, all nodes in a given clade are consecutive. On a tree, this function also provides a pre-ordering of the nodes. The edges' direction needs to be correct before calling <a href="./#PhyloNetworks.cladewiseorder!"><code>cladewiseorder!</code></a>, using <a href="./#PhyloNetworks.directEdges!"><code>directEdges!</code></a></p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/manipulateNet.jl#L813-L821' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.rootatnode!' href='#PhyloNetworks.rootatnode!'>#</a>
<strong><code>PhyloNetworks.rootatnode!</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">rootatnode</span><span class="o">!(</span><span class="nt">HybridNetwork</span><span class="o">,</span> <span class="nt">nodeNumber</span><span class="p">::</span><span class="nd">Integer</span><span class="o">;</span> <span class="nt">index</span><span class="o">=</span><span class="nt">false</span><span class="p">::</span><span class="nd">Bool</span><span class="o">)</span>
<span class="nt">rootatnode</span><span class="o">!(</span><span class="nt">HybridNetwork</span><span class="o">,</span> <span class="nt">Node</span><span class="o">)</span>
<span class="nt">rootatnode</span><span class="o">!(</span><span class="nt">HybridNetwork</span><span class="o">,</span> <span class="nt">nodeName</span><span class="p">::</span><span class="nd">AbstractString</span><span class="o">)</span>
</pre></div>


<p>Roots the network/tree object at the node with name 'nodeName' or number 'nodeNumber' (by default) or with index 'nodeNumber' if index=true. Attributes isChild1 and containRoot are updated along the way. Use <code>plot(net, showNodeNumber=true, showEdgeLength=false)</code> to visualize and identify a node of interest. (see package <a href="https://github.com/cecileane/PhyloPlots.jl">PhyloPlots</a>)</p>
<p>Returns the network.</p>
<p>Warnings:</p>
<ul>
<li>If the node is a leaf, the root will be placed along the edge adjacent to the leaf, with a message. This might add a new node.</li>
<li>
<p>If the desired root placement is incompatible with one or more hybrids, then</p>
<ul>
<li>a RootMismatch error is thrown</li>
<li>the input network will still have some attributes modified.</li>
</ul>
</li>
</ul>
<p>See also: <a href="./#PhyloNetworks.rootonedge!"><code>rootonedge!</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/manipulateNet.jl#L207-L231' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.rootonedge!' href='#PhyloNetworks.rootonedge!'>#</a>
<strong><code>PhyloNetworks.rootonedge!</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">rootonedge</span><span class="o">!(</span><span class="nt">HybridNetwork</span><span class="o">,</span> <span class="nt">edgeNumber</span><span class="p">::</span><span class="nd">Integer</span><span class="o">;</span> <span class="nt">index</span><span class="o">=</span><span class="nt">false</span><span class="p">::</span><span class="nd">Bool</span><span class="o">)</span>
<span class="nt">rootonedge</span><span class="o">!(</span><span class="nt">HybridNetwork</span><span class="o">,</span> <span class="nt">Edge</span><span class="o">)</span>
</pre></div>


<p>Roots the network/tree object along an edge with number 'edgeNumber' (by default) or with index 'edgeNumber if index=true. Attributes isChild1 and containRoot are updated along the way.</p>
<p>This adds a new node and a new edge to the network. Use <code>plot(net, showEdgeNumber=true, showEdgeLength=false)</code> to visualize and identify an edge of interest. (see package <a href="https://github.com/cecileane/PhyloPlots.jl">PhyloPlots</a>)</p>
<p>See also: <a href="./#PhyloNetworks.rootatnode!"><code>rootatnode!</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/manipulateNet.jl#L286-L300' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.hybridatnode!' href='#PhyloNetworks.hybridatnode!'>#</a>
<strong><code>PhyloNetworks.hybridatnode!</code></strong> &mdash; <em>Function</em>.</p>
<p><code>hybridatnode!(net::HybridNetwork, nodeNumber::Integer)</code></p>
<p>Changes the hybrid in a cycle to the node with number <code>nodeNumber</code>. This node must be in one (and only one) cycle, otherwise an error will be thrown.</p>
<p><strong>Example #"</strong></p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="n">net</span> <span class="o">=</span> <span class="n">readTopology</span><span class="p">(</span><span class="s">&quot;(A:1.0,((B:1.1,#H1:0.2::0.2):1.2,(((C:0.52,(E:0.5)#H2:0.02::0.7):0.6,(#H2:0.01::0.3,F:0.7):0.8):0.9,(D:0.8)#H1:0.3::0.8):1.3):0.7):0.1;&quot;</span><span class="p">);</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="k">using</span> <span class="n">PhyloPlots</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">plot</span><span class="p">(</span><span class="n">net</span><span class="p">,</span> <span class="n">showNodeNumber</span><span class="o">=</span><span class="kc">true</span><span class="p">)</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">hybridatnode!</span><span class="p">(</span><span class="n">net</span><span class="p">,</span> <span class="o">-</span><span class="mi">4</span><span class="p">)</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">plot</span><span class="p">(</span><span class="n">net</span><span class="p">)</span>
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/manipulateNet.jl#L92-L106' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.setLength!' href='#PhyloNetworks.setLength!'>#</a>
<strong><code>PhyloNetworks.setLength!</code></strong> &mdash; <em>Function</em>.</p>
<p><code>setLength!(Edge,new length)</code></p>
<p>set a new length for an object Edge. The new length needs to be positive. For example, if you have a HybridNetwork object net, and do printEdges(net), you can see the list of Edges and their lengths. You can then change the length of the 3rd edge with setLength!(net.edge[3],1.2). If <code>new length</code> is above 10, the value 10 will be used, as an upper limit to coalescent units that can be reliably estimated.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/auxiliary.jl#L817-L824' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.setGamma!' href='#PhyloNetworks.setGamma!'>#</a>
<strong><code>PhyloNetworks.setGamma!</code></strong> &mdash; <em>Function</em>.</p>
<p><code>setGamma!(Edge,new gamma)</code></p>
<p>Set inheritance probability γ for an edge, which must be a hybrid edge. The new γ needs to be in [0,1]. The γ of the "partner" hybrid edge is changed accordingly, to 1-γ. The field <code>isMajor</code> is also changed accordingly. If the new γ is approximately 0.5, <code>Edge</code> is set to the major parent, its partner is set to the minor parent.</p>
<p>If <code>net</code> is a HybridNetwork object, <code>printEdges(net)</code> will show the list of edges and their γ's. The γ of the third hybrid edge (say) can be changed to 0.2 with <code>setGamma!(net.edge[3],0.2)</code>. This will automatically set γ of the partner hybrid edge to 0.8.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/auxiliary.jl#L842-L855' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.deleteleaf!' href='#PhyloNetworks.deleteleaf!'>#</a>
<strong><code>PhyloNetworks.deleteleaf!</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">deleteleaf</span><span class="o">!(</span><span class="nt">HybridNetwork</span><span class="o">,</span> <span class="nt">leafName</span><span class="p">::</span><span class="nd">AbstractString</span><span class="o">;</span> <span class="nt">simplify</span><span class="o">=</span><span class="nt">true</span><span class="o">)</span>
<span class="nt">deleteleaf</span><span class="o">!(</span><span class="nt">HybridNetwork</span><span class="o">,</span> <span class="nt">Node</span><span class="o">;</span> <span class="nt">simplify</span><span class="o">=</span><span class="nt">true</span><span class="o">)</span>
<span class="nt">deleteleaf</span><span class="o">!(</span><span class="nt">HybridNetwork</span><span class="o">,</span> <span class="nt">Integer</span><span class="o">;</span> <span class="nt">index</span><span class="o">=</span><span class="nt">false</span><span class="o">,</span> <span class="nt">simplify</span><span class="o">=</span><span class="nt">true</span><span class="o">)</span>
</pre></div>


<p>Deletes a leaf node from the network, possibly from its name, number, or index in the network's array of nodes.</p>
<p>simplify: if true and if deleting the node results in 2 hybrid edges forming a cycle of k=2 nodes, then these hybrid edges are merged and simplified as a single tree edge.</p>
<p>The first 2 versions require that <code>node</code> is a leaf. The 3rd version does <strong>not</strong> require that <code>node</code> is a leaf. If <code>node</code> has degree 3 or more, nothing happens. If it has degree 1 or 2, it is deleted.</p>
<p>Warning: does <strong>not</strong> update attributes related to level-1 networks, such as inCycle, partition, gammaz, etc. Does not require branch lengths, and designed to work on networks of all levels.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/manipulateNet.jl#L914-L934' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.deleteHybridThreshold!' href='#PhyloNetworks.deleteHybridThreshold!'>#</a>
<strong><code>PhyloNetworks.deleteHybridThreshold!</code></strong> &mdash; <em>Function</em>.</p>
<p><code>deleteHybridThreshold!(net::HybridNetwork,gamma::Float64)</code></p>
<p>Deletes from a network all hybrid edges with heritability below a threshold gamma. Returns the network.</p>
<ul>
<li>if gamma&lt;0.5: deletes     minor hybrid edges with gamma value &lt;  threshold</li>
<li>if gamma=0.5: deletes all minor hybrid edges (i.e gamma value &lt;= threshold)</li>
</ul>
<p>Warning: assumes correct isMajor attributes.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/compareNetworks.jl#L261-L271' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.rotate!' href='#PhyloNetworks.rotate!'>#</a>
<strong><code>PhyloNetworks.rotate!</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span><span class="nt">rotate</span><span class="o">!(</span><span class="nt">net</span><span class="p">::</span><span class="nd">HybridNetwork</span><span class="o">,</span> <span class="nt">nodeNumber</span><span class="p">::</span><span class="nd">Integer</span><span class="o">;</span> <span class="nt">orderedEdgeNum</span><span class="p">::</span><span class="nd">Array</span><span class="p">{</span><span class="err">Int,1</span><span class="p">}</span><span class="o">)</span>
</pre></div>


<p>Rotates the order of the node's children edges. Useful for plotting, to remove crossing edges. If <code>node</code> is a tree node with no polytomy, the 2 children edges are switched and the optional argument <code>orderedEdgeNum</code> is ignored.</p>
<p>Use <code>plot(net, showNodeNumber=true, showEdgeNumber=false)</code> to map node and edge numbers on the network, as shown in the examples below. (see package <a href="https://github.com/cecileane/PhyloPlots.jl">PhyloPlots</a>)</p>
<p>Warning: assumes that edges are correctly directed (isChild1 updated). This is done by <code>plot(net)</code>. Otherwise run <code>directEdges!(net)</code>.</p>
<p><strong>Example #"</strong></p>
<div class="codehilite"><pre><span></span><span class="n">julia</span><span class="o">&gt;</span> <span class="n">net</span> <span class="o">=</span> <span class="n">readTopology</span><span class="p">(</span><span class="s">&quot;(A:1.0,((B:1.1,#H1:0.2::0.2):1.2,(((C:0.52,(E:0.5)#H2:0.02::0.7):0.6,(#H2:0.01::0.3,F:0.7):0.8):0.9,(D:0.8)#H1:0.3::0.8):1.3):0.7):0.1;&quot;</span><span class="p">);</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="k">using</span> <span class="n">PhyloPlots</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">plot</span><span class="p">(</span><span class="n">net</span><span class="p">,</span> <span class="n">showNodeNumber</span><span class="o">=</span><span class="kc">true</span><span class="p">)</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">rotate!</span><span class="p">(</span><span class="n">net</span><span class="p">,</span> <span class="o">-</span><span class="mi">4</span><span class="p">)</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">plot</span><span class="p">(</span><span class="n">net</span><span class="p">)</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">net</span><span class="o">=</span><span class="n">readTopology</span><span class="p">(</span><span class="s">&quot;(4,((1,(2)#H7:::0.864):2.069,(6,5):3.423):0.265,(3,#H7:::0.136):10.0);&quot;</span><span class="p">);</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">plot</span><span class="p">(</span><span class="n">net</span><span class="p">,</span> <span class="n">showNodeNumber</span><span class="o">=</span><span class="kc">true</span><span class="p">,</span> <span class="n">showEdgeNumber</span><span class="o">=</span><span class="kc">true</span><span class="p">)</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">rotate!</span><span class="p">(</span><span class="n">net</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">orderedEdgeNum</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">12</span><span class="p">,</span><span class="mi">9</span><span class="p">])</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">plot</span><span class="p">(</span><span class="n">net</span><span class="p">,</span> <span class="n">showNodeNumber</span><span class="o">=</span><span class="kc">true</span><span class="p">,</span> <span class="n">showEdgeNumber</span><span class="o">=</span><span class="kc">true</span><span class="p">)</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">rotate!</span><span class="p">(</span><span class="n">net</span><span class="p">,</span> <span class="o">-</span><span class="mi">3</span><span class="p">)</span>
<span class="n">julia</span><span class="o">&gt;</span> <span class="n">plot</span><span class="p">(</span><span class="n">net</span><span class="p">)</span>
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/manipulateNet.jl#L849-L879' class='documenter-source'>source</a><br></p>
<p><a id='Base.getindex-Tuple{PhyloNetworks.TraitSimulation,Symbol}' href='#Base.getindex-Tuple{PhyloNetworks.TraitSimulation,Symbol}'>#</a>
<strong><code>Base.getindex</code></strong> &mdash; <em>Method</em>.</p>
<div class="codehilite"><pre><span></span>getindex(collection, key...)
</pre></div>


<p>Retrieve the value(s) stored at the given key or index within a collection. The syntax <code>a[i,j,...]</code> is converted by the compiler to <code>getindex(a, i, j, ...)</code>.</p>
<p><strong>Example</strong></p>
<div class="codehilite"><pre><span></span>julia&gt; A = Dict(&quot;a&quot; =&gt; 1, &quot;b&quot; =&gt; 2)
Dict{String,Int64} with 2 entries:
  &quot;b&quot; =&gt; 2
  &quot;a&quot; =&gt; 1

julia&gt; getindex(A, &quot;a&quot;)
1
</pre></div>


<p><a target='_blank' href='https://github.com/JuliaLang/julia/blob/d386e40c17d43b79fc89d3e579fc04547241787c/base/docs/helpdb/Base.jl#L150-L166' class='documenter-source'>source</a><br></p>
<p><code>getindex(obj, d)</code></p>
<p>Getting submatrices of an object of type <a href="./#PhyloNetworks.TraitSimulation"><code>TraitSimulation</code></a>.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>obj::TraitSimulation</code>: the matrix from which to extract.</li>
<li>
<p><code>d::Symbol</code>: a symbol precising which sub-matrix to extract. Can be:</p>
<ul>
<li><code>:Tips</code> columns and/or rows corresponding to the tips</li>
<li><code>:InternalNodes</code> columns and/or rows corresponding to the internal nodes</li>
</ul>
</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1004-L1014' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.getNodeAges' href='#PhyloNetworks.getNodeAges'>#</a>
<strong><code>PhyloNetworks.getNodeAges</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>getNodeAges(net)
</pre></div>


<p>vector of node ages in pre-order, as in <code>nodes_changed</code>, which is assumed to have been calculated before.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/pairwiseDistanceLS.jl#L15-L20' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.pairwiseTaxonDistanceMatrix' href='#PhyloNetworks.pairwiseTaxonDistanceMatrix'>#</a>
<strong><code>PhyloNetworks.pairwiseTaxonDistanceMatrix</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>pairwiseTaxonDistanceMatrix(net; keepInternal=false,
                            checkPreorder=true, nodeAges=[])
pairwiseTaxonDistanceMatrix!(M, net, nodeAges)
</pre></div>


<p>Matrix of pairwise distances between nodes in the network:</p>
<ul>
<li>between all nodes (internal and leaves) if <code>keepInternal=true</code></li>
<li>between taxa only otherwise.</li>
</ul>
<p>Update <code>net.nodes_changed</code> to get a topological ordering if <code>checkPreorder</code> is true. The second form modifies <code>M</code> in place, assuming all nodes.</p>
<p>If <code>nodeAges</code> is provided: this vector should list node ages in the pre-order in which they are listed in nodes_changed (including leaves), and <strong>edge lengths</strong> in <code>net</code> <strong>are modified</strong> accordingly.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/pairwiseDistanceLS.jl#L41-L56' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.biconnectedComponents' href='#PhyloNetworks.biconnectedComponents'>#</a>
<strong><code>PhyloNetworks.biconnectedComponents</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>biconnectedComponents(network, ignoreTrivial=false)
</pre></div>


<p>Calculate biconnected components (aka "blobs") using Tarjan's algorithm: the output is an array of arrays of edges. These blobs are returned in post-order, but within a blob, edges are <em>not</em> necessarily sorted in topological order. If <code>ignoreTrivial</code> is true, trivial components (of a single edge) are not returned. The network is assumed to be connected.</p>
<p><strong>Warnings</strong>: for nodes, fields <code>k</code>, <code>inCycle</code>, and <code>prev</code> are modified during the algorithm. They are used to store the node's "index" (time of visitation), "lowpoint", and the node's "parent", as defined by the order in which nodes are visited.</p>
<p>References:</p>
<ul>
<li>p. 153 of <a href="http://epubs.siam.org/doi/pdf/10.1137/0201010">Tarjan (1972)</a>. Depth first search and linear graph algorithms, SIAM Journal on Computing, 1(2):146-160</li>
<li>on <a href="https://www.geeksforgeeks.org/biconnected-components/">geeksforgeeks</a>, there is an error (as of 2018-01-30): <code>elif v != parent[u] and low[u] &gt; disc[v]:</code> (python version) should be replaced by <code>elif v != parent[u] and disc[u] &gt; disc[v]:</code></li>
<li>nice explanation at this <a href="https://www.cs.cmu.edu/~avrim/451f12/lectures/biconnected.pdf">url</a></li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/biconnectedComponents.jl#L1-L28' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>biconnectedComponents(node, index, S, blobs, ignoreTrivial)
</pre></div>


<p>Helper recursive function starting at a node (not a network). <code>index</code> is an array containing a single integer, thus mutable: order in which nodes are visited.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/biconnectedComponents.jl#L53-L59' class='documenter-source'>source</a><br></p>
<p><a id='data-and-topology-read/write-1'></a></p>
<h2 id="data-and-topology-readwrite">data and topology read/write</h2>
<p><a id='PhyloNetworks.readTopology' href='#PhyloNetworks.readTopology'>#</a>
<strong><code>PhyloNetworks.readTopology</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>readTopology(file name)
readTopology(parenthetical description)
</pre></div>


<p>Read tree or network topology from parenthetical format (extended Newick).</p>
<p>Input: text file or parenthetical format directly. The file name may not start with a left parenthesis, otherwise the file name itself would be interpreted as the parenthetical description.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readwrite.jl#L447-L456' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.readTopologyLevel1' href='#PhyloNetworks.readTopologyLevel1'>#</a>
<strong><code>PhyloNetworks.readTopologyLevel1</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>readTopologyLevel1(filename)
readTopologyLevel1(parenthetical format)
</pre></div>


<p>same as readTopology, reads a tree or network from parenthetical format, but this function enforces the necessary conditions for any starting topology in SNaQ: non-intersecting cycles, no polytomies, unrooted. It sets any missing branch length to 1.0.</p>
<p>If the network has a bad diamond II (in which edge lengths are γ's are not identifiable) and if the edge below this diamond has a length <code>t</code> different from 0, then this length is set back to 0 and the major parent hybrid edge is lengthened by <code>t</code>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readwrite.jl#L877-L889' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.readInputTrees' href='#PhyloNetworks.readInputTrees'>#</a>
<strong><code>PhyloNetworks.readInputTrees</code></strong> &mdash; <em>Function</em>.</p>
<p><code>readInputTrees(file)</code></p>
<p>Read a text file with a list of trees/networks in parenthetical format (one tree per line) and transform them like <a href="./#PhyloNetworks.readTopologyLevel1"><code>readTopologyLevel1</code></a> does: to be unrooted, with resolved polytomies, missing branch lengths set to 1.0, etc. See <a href="./#PhyloNetworks.readMultiTopology"><code>readMultiTopology</code></a> to read multiple trees or networks with no modification.</p>
<p>Output: array of HybridNetwork objects.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readData.jl#L184-L194' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.readMultiTopology' href='#PhyloNetworks.readMultiTopology'>#</a>
<strong><code>PhyloNetworks.readMultiTopology</code></strong> &mdash; <em>Function</em>.</p>
<p><code>readMultiTopology(file)</code></p>
<p>Read a text file with a list of networks in parenthetical format (one per line). Each network is read with <a href="./#PhyloNetworks.readTopology"><code>readTopology</code></a>. Return an array of HybridNetwork object.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readwrite.jl#L1243-L1249' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.readSnaqNetwork' href='#PhyloNetworks.readSnaqNetwork'>#</a>
<strong><code>PhyloNetworks.readSnaqNetwork</code></strong> &mdash; <em>Function</em>.</p>
<p><code>readSnaqNetwork(output file)</code></p>
<p>function to read the estimated network from an .out file generated by the snaq function</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readwrite.jl#L1207-L1211' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.readTrees2CF' href='#PhyloNetworks.readTrees2CF'>#</a>
<strong><code>PhyloNetworks.readTrees2CF</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>readTrees2CF(treefile)
readTrees2CF(vector of trees)
</pre></div>


<p>Read trees in parenthetical format from a file, or take a vector of trees already read, and calculate the proportion of these trees having a given quartet (concordance factor: CF), for all quartets or for a sample of quartets. Optional arguments include:</p>
<ul>
<li>quartetfile: name of text file with list of 4-taxon subsets to be analyzed. If none is specified, the function will list all possible 4-taxon subsets.</li>
<li>whichQ="rand": to choose a random sample of 4-taxon subsets</li>
<li>numQ: size of random sample (ignored if whichQ is not set to "rand")</li>
<li>writeTab=false: does not write the observedCF to a table (default true)</li>
<li>CFfile: name of file to save the observedCF (default tableCF.txt)</li>
<li>writeQ=true: save intermediate files with the list of all 4-taxon subsets and chosen random sample (default false).</li>
<li>writeSummary: write descriptive stats of input data (default: true)</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readData.jl#L626-L642' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.readTableCF' href='#PhyloNetworks.readTableCF'>#</a>
<strong><code>PhyloNetworks.readTableCF</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>readTableCF(file)
readTableCF(data frame)
readTableCF!(data frame)
</pre></div>


<p>Read a file or DataFrame object containing a table of concordance factors (CF), with one row per 4-taxon set. The first 4 columns are assumed to give the labels of the 4 taxa in each set (tx1, tx2, tx3, tx4). Columns containing the CFs are assumed to be named 'CF12_34', 'CF13_24' and 'CF14_23', or else are assumed to be columns 5,6,7. If present, a column named 'ngenes' will be used to get the number of loci used to estimate the CFs for each 4-taxon set.</p>
<p>Optional arguments:</p>
<ul>
<li>summaryfile: if specified, a summary file will be created with that name.</li>
<li>delim (for the first form only): to specify how columns are delimited, with single quotes: delim=';'. Default is a <code>csv</code> file, i.e. <code>delim=','</code>.</li>
</ul>
<p>The last version modifies the input data frame, if species are represented by multiple alleles for instance (see <code>readTableCF!</code>).</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readData.jl#L42-L63' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.readTableCF!' href='#PhyloNetworks.readTableCF!'>#</a>
<strong><code>PhyloNetworks.readTableCF!</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>readTableCF!(data frame, columns)
</pre></div>


<p>Read in quartet CFs from data frame, assuming information is in columns numbered <code>columns</code>, of length <strong>7 or 8</strong>: 4 taxon labels then 3 CFs then ngenes possibly.</p>
<p>If some species appears more than once in the same 4-taxon set (e.g. t1,t1,t2,t3), then the data frame is modified to remove rows (4-taxon sets) that are uninformative about between-species relationships. This situation may occur if multiple individuals are sampled from the same species. A 4-taxon set is uninformative (and its row is removed) if one taxon is repeated 3 or 4 times (like t1,t1,t1,t1 or t1,t2,t2,t2). The list of species appearing twice in some 4-taxon sets is stored in the output DataCF object. For these species, the length of their external edge is identifiable (in coalescent units). If multiple rows correspond to the same 4-taxon set, these rows are merged and their CF values (and number of genes) are averaged.</p>
<div class="codehilite"><pre><span></span>readTableCF!(DataCF, data frame, columns)
</pre></div>


<p>Modify the <code>.quartet.obsCF</code> values in the <code>DataCF</code> object with those read from the data frame in columns numbered <code>columns</code>. <code>columns</code> should have <strong>3</strong> columns numbers for the 3 CFs in this order: 12_34, 13_24 and 14_23. Assume the same 4-taxon sets in <code>DataCF</code> and in the data frame, and in the same order, but this assumption is <em>not checked</em> (for speed, e.g. during bootstrapping).</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readData.jl#L143-L166' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.readBootstrapTrees' href='#PhyloNetworks.readBootstrapTrees'>#</a>
<strong><code>PhyloNetworks.readBootstrapTrees</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>readBootstrapTrees(filename)
</pre></div>


<p>input: name of file containing the path/name to multiple bootstrap files, one per line. Each bootstrap file corresponds to bootstrap trees from a single gene.</p>
<p>output: vector of vectors of trees.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/bootstrap.jl#L5-L12' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.writeSubTree!' href='#PhyloNetworks.writeSubTree!'>#</a>
<strong><code>PhyloNetworks.writeSubTree!</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>writeSubTree!(IO, network, dendroscope::Bool, names::Bool,
              round_branch_lengths::Bool, digits::Integer)
</pre></div>


<p>Write to IO the extended newick format (parenthetical description) of a network. If written for dendroscope, inheritance γ's are not written. If <code>names</code> is true, taxon names are written, otherwise taxon numbers are written instead of taxon names. If unspecified, branch lengths and γ's are rounded to 3 digits.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readwrite.jl#L926-L936' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>writeSubTree!(IO, node, edge, dendroscope::Bool, names::Bool,
              round_branch_lengths::Bool, digits::Integer)
</pre></div>


<p>Write the extended newick format of the sub-network rooted at <code>node</code> and assuming that <code>edge</code> is a parent of <code>node</code>.</p>
<p>If <code>parent</code> is <code>nothing</code>, the edge attribute <code>isChild1</code> is used and assumed to be correct to write the subtree rooted at <code>node</code>. This is useful to write a subtree starting at a non-root node. Example:</p>
<div class="codehilite"><pre><span></span><span class="n">net</span> <span class="o">=</span> <span class="n">readTopology</span><span class="p">(</span><span class="s">&quot;(((A,(B)#H1:::0.9),(C,#H1:::0.1)),D);&quot;</span><span class="p">)</span>
<span class="n">directEdges!</span><span class="p">(</span><span class="n">net</span><span class="p">)</span>
<span class="n">s</span> <span class="o">=</span> <span class="kt">IOBuffer</span><span class="p">()</span>
<span class="n">PhyloNetworks</span><span class="o">.</span><span class="n">writeSubTree!</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">net</span><span class="o">.</span><span class="n">node</span><span class="p">[</span><span class="mi">7</span><span class="p">],</span> <span class="nb">nothing</span><span class="p">,</span> <span class="kc">false</span><span class="p">,</span> <span class="kc">true</span><span class="p">)</span>
<span class="n">String</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>


<p>Used by <a href="./#PhyloNetworks.writeTopology"><code>writeTopology</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readwrite.jl#L957-L978' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.writeTopology' href='#PhyloNetworks.writeTopology'>#</a>
<strong><code>PhyloNetworks.writeTopology</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>writeTopology(net)
writeTopology(net, filename)
writeTopology(net, IO)
</pre></div>


<p>Write the parenthetical extended Newick format of a network, as a string, to a file or to an IO buffer / stream. Optional arguments (default values):</p>
<ul>
<li>di (false): write in format for Dendroscope</li>
<li>round (false): rounds branch lengths and heritabilities γ</li>
<li>digits (3): digits after the decimal place for rounding</li>
<li>append (false): if true, appends to the file</li>
</ul>
<p>If the current root placement is not admissible, other placements are tried. The network is updated with this new root placement, if successful.</p>
<p>Uses lower-level function <a href="./#PhyloNetworks.writeSubTree!"><code>writeSubTree!</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readwrite.jl#L1316-L1334' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.writeMultiTopology' href='#PhyloNetworks.writeMultiTopology'>#</a>
<strong><code>PhyloNetworks.writeMultiTopology</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>writeMultiTopology(nets, file_name; append=false)
writeMultiTopology(nets, IO)
</pre></div>


<p>Write an array of networks in parenthetical extended Newick format, one network per line. Use the option append=true to append to the file. Otherwise, the default is to create a new file or overwrite it, if it already existed. Each network is written with <code>writeTopology</code>.</p>
<p><strong>Examples #"</strong></p>
<div class="codehilite"><pre><span></span><span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">net</span> <span class="o">=</span> <span class="cp">[</span><span class="nx">readTopology</span><span class="p">(</span><span class="s2">&quot;(D,((A,(B)#H7:::0.864):2.069,(F,E):3.423):0.265,(C,#H7:::0.1361111):10);&quot;</span><span class="p">),</span>
              <span class="nx">readTopology</span><span class="p">(</span><span class="s2">&quot;(A,(B,C));&quot;</span><span class="p">),</span><span class="nx">readTopology</span><span class="p">(</span><span class="s2">&quot;(E,F);&quot;</span><span class="p">),</span><span class="nx">readTopology</span><span class="p">(</span><span class="s2">&quot;(G,H,F);&quot;</span><span class="p">)</span><span class="cp">]</span><span class="o">;</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">writeMultiTopology</span><span class="o">(</span><span class="nt">net</span><span class="o">,</span> <span class="s2">&quot;fournets.net&quot;</span><span class="o">)</span> <span class="err">#</span> <span class="nt">to</span> <span class="o">(</span><span class="nt">over</span><span class="o">)</span><span class="nt">write</span> <span class="nt">to</span> <span class="nt">file</span> <span class="s2">&quot;fournets.net&quot;</span>
<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">writeMultiTopology</span><span class="o">(</span><span class="nt">net</span><span class="o">,</span> <span class="s2">&quot;fournets.net&quot;</span><span class="o">,</span> <span class="nt">append</span><span class="o">=</span><span class="nt">true</span><span class="o">)</span> <span class="err">#</span> <span class="nt">to</span> <span class="nt">append</span> <span class="nt">to</span> <span class="nt">this</span> <span class="nt">file</span>
<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">writeMultiTopology</span><span class="o">(</span><span class="nt">net</span><span class="o">,</span> <span class="nt">STDOUT</span><span class="o">)</span>         <span class="err">#</span> <span class="nt">to</span> <span class="nt">write</span> <span class="nt">to</span> <span class="nt">the</span> <span class="nt">screen</span> <span class="o">(</span><span class="nt">standard</span> <span class="nt">out</span><span class="o">)</span>
<span class="o">(</span><span class="nt">D</span><span class="o">,((</span><span class="nt">A</span><span class="o">,(</span><span class="nt">B</span><span class="o">)</span><span class="p">#</span><span class="nn">H7</span><span class="o">:</span><span class="p">::</span><span class="nd">0</span><span class="p">.</span><span class="nc">864</span><span class="o">)</span><span class="p">:</span><span class="nd">2</span><span class="p">.</span><span class="nc">069</span><span class="o">,(</span><span class="nt">F</span><span class="o">,</span><span class="nt">E</span><span class="o">)</span><span class="p">:</span><span class="nd">3</span><span class="p">.</span><span class="nc">423</span><span class="o">)</span><span class="p">:</span><span class="nd">0</span><span class="p">.</span><span class="nc">265</span><span class="o">,(</span><span class="nt">C</span><span class="o">,</span><span class="p">#</span><span class="nn">H7</span><span class="o">:</span><span class="p">::</span><span class="nd">0</span><span class="p">.</span><span class="nc">1361111</span><span class="o">)</span><span class="p">:</span><span class="nd">10</span><span class="p">.</span><span class="nc">0</span><span class="o">);</span>
<span class="o">(</span><span class="nt">A</span><span class="o">,(</span><span class="nt">B</span><span class="o">,</span><span class="nt">C</span><span class="o">));</span>
<span class="o">(</span><span class="nt">E</span><span class="o">,</span><span class="nt">F</span><span class="o">);</span>
<span class="o">(</span><span class="nt">G</span><span class="o">,</span><span class="nt">H</span><span class="o">,</span><span class="nt">F</span><span class="o">);</span>
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/readwrite.jl#L1271-L1293' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.mapAllelesCFtable' href='#PhyloNetworks.mapAllelesCFtable'>#</a>
<strong><code>PhyloNetworks.mapAllelesCFtable</code></strong> &mdash; <em>Function</em>.</p>
<p><code>mapAllelesCFtable(mapping file, CF file; filename, columns, delim)</code></p>
<p>Create a new DataFrame containing the same concordance factors as in the input CF file, but with modified taxon names. Each allele name in the input CF table is replaced by the species name that the allele maps onto, based on the mapping file. The mapping file should have column names: allele and species.</p>
<p>Optional arguments:</p>
<ul>
<li>file name to write/save resulting CF table. If not specified, then the output data frame is not saved to a file.</li>
<li>column numbers for the taxon names. 1-4 by default.</li>
<li>delim=',': how columns are delimited. comma by default: for csv files.</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/multipleAlleles.jl#L7-L20' class='documenter-source'>source</a><br></p>
<p><a id='network-inference-1'></a></p>
<h2 id="network-inference">network inference</h2>
<p><a id='PhyloNetworks.snaq!' href='#PhyloNetworks.snaq!'>#</a>
<strong><code>PhyloNetworks.snaq!</code></strong> &mdash; <em>Function</em>.</p>
<p><code>snaq!(T::HybridNetwork, d::DataCF)</code></p>
<p>Estimate the network (or tree) to fit observed quartet concordance factors (CFs) stored in a DataCF object, using maximum pseudo-likelihood. A level-1 network is assumed. The search starts from topology <code>T</code>, which can be a tree or a network with no more than <code>hmax</code> hybrid nodes. The function name ends with ! because it modifies the CF data <code>d</code> by updating its attributes <code>expCF</code>: CFs expected under the network model. It does <em>not</em> modify <code>T</code>. The quartet pseudo-deviance is the negative log pseudo-likelihood, up to an additive constant, such that a perfect fit corresponds to a deviance of 0.0.</p>
<p>Output:</p>
<ul>
<li>estimated network in file <code>.out</code> (also in <code>.log</code>): best network overall and list of networks from each individual run.</li>
<li>the best network and modifications of it, in file <code>.networks</code>. All networks in this file have the same undirected topology as the best network, but have different hybrid/gene flow directions. These other networks are reported with their pseudo-likelihood scores, because  non-identifiability issues can cause them to have very similar scores, and because  SNaQ was shown to estimate the undirected topology accurately but not the direction of  hybridization in cases of near non-identifiability.</li>
<li>if any error occurred, file <code>.err</code> provides information (seed) to reproduce the error.</li>
</ul>
<p>There are many optional arguments, including</p>
<ul>
<li>hmax: maximum number of hybridizations allowed (default 1)</li>
<li>verbose: if true, print information about the numerical optimization</li>
<li>runs: number of independent starting points for the search (default 10)</li>
<li>outgroup: outgroup taxon to root the estimated topology at the very end</li>
<li>filename: root name for the output files. Default is "snaq". If empty (""), files are <em>not</em> created, progress log goes to the screen only (standard out).</li>
<li>seed: seed to replicate a given search</li>
</ul>
<p>The following optional arguments control when to stop the optimization of branch lengths and γ's on each individual candidate network. Defaults are in parentheses:</p>
<ul>
<li>ftolRel (1e-6) and ftolAbs (1e-6): relative and absolute differences of the network score between the current and proposed parameters,</li>
<li>xtolRel (1e-2) and xtolAbs (1e-3): relative and absolute differences between the current and proposed parameters.</li>
</ul>
<p>Greater values will result in a less thorough but faster search. These parameters are used when evaluating candidate networks only. The following optional arguments control when to stop proposing new network topologies:</p>
<ul>
<li>Nfail (75): maximum number of times that new topologies are proposed and rejected (in a row).</li>
<li>liktolAbs (1e-6): the proposed network is accepted if its score is better than the current score by at least liktolAbs.</li>
</ul>
<p>Lower values of Nfail and greater values of liktolAbs and ftolAbs would result in a less thorough but faster search.</p>
<p>At the end, branch lengths and γ's are optimized on the last "best" network with different and very thorough tolerance parameters: 1e-12 for ftolRel, 1e-10 for ftolAbs, xtolRel, xtolAbs.</p>
<p>See also: <code>topologyMaxQPseudolik!</code> to optimize parameters on a fixed topology, and <code>topologyQPseudolik!</code> to get the deviance (pseudo log-likelihood up to a constant) of a fixed topology with fixed parameters.</p>
<p>Reference: Claudia Solís-Lemus and Cécile Ané (2016). Inferring phylogenetic networks with maximum pseudolikelihood under incomplete lineage sorting. <a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1005896">PLoS Genetics</a> 12(3):e1005896</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/optimization.jl#L1853-L1920' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.snaqDebug' href='#PhyloNetworks.snaqDebug'>#</a>
<strong><code>PhyloNetworks.snaqDebug</code></strong> &mdash; <em>Function</em>.</p>
<p><code>snaqDebug(currT::HybridNetwork,d::DataCF)</code></p>
<p>function to replicate a given run that produces error according to the .err file generated by snaq. The same settings used in that run should be used in this function, specially the seed. See the readme file online for more details.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/optimization.jl#L1957-L1962' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.topologyMaxQPseudolik!' href='#PhyloNetworks.topologyMaxQPseudolik!'>#</a>
<strong><code>PhyloNetworks.topologyMaxQPseudolik!</code></strong> &mdash; <em>Function</em>.</p>
<p><code>topologyMaxQPseudolik!(net::HybridNetwork, d::DataCF)</code></p>
<p>Estimate the branch lengths and inheritance probabilities (γ's) for a given network topology. The network is <em>not</em> modified, only the object <code>d</code> is, with updated expected concordance factors.</p>
<p>Ouput: new network, with optimized parameters (branch lengths and gammas). The maximized quartet pseudo-deviance is the negative log pseudo-likelihood, up to an additive constant, such that a perfect fit corresponds to a deviance of 0.0. This is also an attribute of the network, which can be accessed with <code>net.loglik</code>.</p>
<p>Optional arguments (default value):</p>
<ul>
<li>verbose (false): if true, information on the numerical optimization is printed to screen</li>
<li>ftolRel (1e-5), ftolAbs (1e-6), xtolRel (1e-3), xtolAbs (1e-4): absolute and relative tolerance values for the pseudo-deviance function and the parameters</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/optimization.jl#L404-L421' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.topologyQPseudolik!' href='#PhyloNetworks.topologyQPseudolik!'>#</a>
<strong><code>PhyloNetworks.topologyQPseudolik!</code></strong> &mdash; <em>Function</em>.</p>
<p><code>topologyQPseudolik!(net::HybridNetwork, d::DataCF)</code></p>
<p>Calculate the quartet pseudo-deviance of a given network/tree for DataCF <code>d</code>. This is the negative log pseudo-likelihood, up to an additive constant, such that a perfect fit corresponds to a deviance of 0.0.</p>
<p>Be careful if the net object does not have all internal branch lengths specified because then the pseudolikelihood will be meaningless.</p>
<p>The loglik attribute of the network is undated, and <code>d</code> is updated with the expected concordance factors under the input network.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/pseudolik.jl#L1291-L1304' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.fittedQuartetCF' href='#PhyloNetworks.fittedQuartetCF'>#</a>
<strong><code>PhyloNetworks.fittedQuartetCF</code></strong> &mdash; <em>Function</em>.</p>
<p><code>fittedQuartetCF(d::DataCF, format::Symbol)</code></p>
<p>return a data frame with the observed and expected quartet concordance factors after estimation of a network with snaq(T,d). The format can be :wide (default) or :long.</p>
<ul>
<li>if wide, the output has one row per 4-taxon set, and each row has 10 columns: 4 columns for the taxon names, 3 columns for the observed CFs and 3 columns for the expected CF.</li>
<li>if long, the output has one row per quartet, i.e. 3 rows per 4-taxon sets, and 7 columns: 4 columns for the taxon names, one column to give the quartet resolution, one column for the observed CF and the last column for the expected CF.</li>
</ul>
<p>see also: <code>topologyQPseudolik!</code> and <code>topologyMaxQPseudolik!</code> to update the fitted CF expected under a specific network, inside the DataCF object <code>d</code>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/descriptive.jl#L15-L30' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.bootsnaq' href='#PhyloNetworks.bootsnaq'>#</a>
<strong><code>PhyloNetworks.bootsnaq</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>bootsnaq(T::HybridNetwork, df::DataFrame)
bootsnaq(T::HybridNetwork, vector of tree lists)
</pre></div>


<p>Bootstrap analysis for SNaQ. Bootstrap data can be quartet concordance factors (CF), drawn from sampling uniformly in their credibility intervals, as given in the data frame <code>df</code>. Alternatively, bootstrap data can be gene trees sampled from a vector of tree lists: one list of bootstrap trees per locus (see <code>readBootstrapTrees</code> to generate this, from a file containing a list of bootstrap files: one per locus).</p>
<p>From each bootstrap replicate, a network is estimated with snaq!, with a search starting from topology <code>T</code>. Optional arguments include the following, with default values in parentheses:</p>
<ul>
<li>hmax (1): max number of reticulations in the estimated networks</li>
<li>nrep (10): number of bootstrap replicates.</li>
<li>runs (10): number of independent optimization runs for each replicate</li>
<li>filename ("bootsnaq"): root name for output files. No output files if "".</li>
<li>seed (0 to get a random seed from the clock): seed for random number generator</li>
<li>otherNet (empty): another starting topology so that each replicate will start prcnet% runs on otherNet and (1-prcnet)% runs on T</li>
<li>prcnet (0): percentage of runs starting on otherNet; error if different than 0.0, and otherNet not specified.</li>
<li>ftolRel, ftolAbs, xtolRel, xtolAbs, liktolAbs, Nfail: see <code>snaq!</code>, same defaults.</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/bootstrap.jl#L293-L318' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.calibrateFromPairwiseDistances!' href='#PhyloNetworks.calibrateFromPairwiseDistances!'>#</a>
<strong><code>PhyloNetworks.calibrateFromPairwiseDistances!</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>calibrateFromPairwiseDistances!(net, distances::Matrix{Float64},
    taxon_names::Vector{String})
</pre></div>


<p>Calibrate the network to match (as best as possible) input pairwise distances between taxa, such as observed from sequence data. <code>taxon_names</code> should provide the list of taxa, in the same order in which they they are considered in the <code>distances</code> matrix. The optimization criterion is the sum of squares between the observed distances, and the distances from the network (weighted average of tree distances, weighted by γ's). The network's edge lengths are modified.</p>
<p>Warning: for many networks, mutiple calibrations can fit the pairwise distance data equally well (lack of identifiability). This function will output <em>one</em> of these equally good calibrations.</p>
<p>optional arguments (default):</p>
<ul>
<li>checkPreorder (true)</li>
<li>forceMinorLength0 (false) to force minor hybrid edges to have a length of 0</li>
<li>NLoptMethod (:LD_MMA) for the optimization algorithm. Other options include :LN_COBYLA (derivative-free); see NLopt package.</li>
<li>tolerance values to control when the optimization is stopped: ftolRel (1e-12), ftolAbs (1e-10) on the criterion, and xtolRel (1e-10), xtolAbs (1e-10) on branch lengths / divergence times.</li>
<li>verbose (false)</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/pairwiseDistanceLS.jl#L209-L235' class='documenter-source'>source</a><br></p>
<p><a id='network-Comparisons-1'></a></p>
<h2 id="network-comparisons">network Comparisons</h2>
<p><a id='PhyloNetworks.majorTree' href='#PhyloNetworks.majorTree'>#</a>
<strong><code>PhyloNetworks.majorTree</code></strong> &mdash; <em>Function</em>.</p>
<p><code>majorTree(net::HybridNetwork)</code></p>
<p>Warning: assumes correct isMajor attributes.</p>
<p>Extracts the major tree displayed in a network, keeping the major edge and dropping the minor edge at each hybrid node. Returns a HybridNetwork object.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/compareNetworks.jl#L330-L337' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.minorTreeAt' href='#PhyloNetworks.minorTreeAt'>#</a>
<strong><code>PhyloNetworks.minorTreeAt</code></strong> &mdash; <em>Function</em>.</p>
<p><code>minorTreeAt(net::HybridNetwork, hybindex::Integer)</code></p>
<p>Warning: assumes correct isMajor attributes.</p>
<p>Extracts the tree displayed in the network, following the major hybrid edge at each hybrid node, except at the ith hybrid node (i=hybindex), where the minor hybrid edge is kept instead of the major hybrid edge.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/compareNetworks.jl#L359-L365' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.displayedTrees' href='#PhyloNetworks.displayedTrees'>#</a>
<strong><code>PhyloNetworks.displayedTrees</code></strong> &mdash; <em>Function</em>.</p>
<p><code>displayedTrees(net::HybridNetwork, gamma::Float64)</code></p>
<p>Warning: assumes correct isMajor attributes.</p>
<p>Extracts all trees displayed in a network, following hybrid edges with heritability &gt;= gamma threshold (or &gt;0.5 if threshold=0.5) and ignoring any hybrid edge with heritability lower than gamma. Returns an array of trees, as HybridNetwork objects.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/compareNetworks.jl#L312-L321' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.displayedNetworkAt!' href='#PhyloNetworks.displayedNetworkAt!'>#</a>
<strong><code>PhyloNetworks.displayedNetworkAt!</code></strong> &mdash; <em>Function</em>.</p>
<p><code>displayedNetworkAt!(net::HybridNetwork, node::Node)</code></p>
<p>Warning: assumes correct isMajor attributes.</p>
<p>Deletes all the minor hybrid edges, except at input node. The network is left with a single hybridization, and otherwise displays the same major tree as before.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/compareNetworks.jl#L377-L383' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.hardwiredClusters' href='#PhyloNetworks.hardwiredClusters'>#</a>
<strong><code>PhyloNetworks.hardwiredClusters</code></strong> &mdash; <em>Function</em>.</p>
<p><code>hardwiredClusters(net::HybridNetwork, S::Union{Vector{String},Vector{Int}})</code></p>
<p>Returns a matrix describing all the hardwired clusters in a network. Warnings: Clusters are rooted, so the root must be correct.           Allows for missing taxa, with entries all 0.</p>
<p>Each row corresponds to one internal edge, that is, external edges are excluded. If the root is a leaf node, the external edge to that leaf is included (first row). Both parent hybrid edges to a given hybrid node only contribute a single row (they share the same hardwired cluster).</p>
<ul>
<li>first column: edge number</li>
<li>next columns: 0/1 values. 1=descendant of edge, 0=not a descendant, or missing taxon.</li>
<li>last column:  10/11 values. 10=tree edge, 11=hybrid edge</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/compareNetworks.jl#L62-L77' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.hardwiredCluster' href='#PhyloNetworks.hardwiredCluster'>#</a>
<strong><code>PhyloNetworks.hardwiredCluster</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>hardwiredCluster(edge::Edge,taxa::Union{Vector{String},Vector{Int}})
hardwiredCluster!(v::Vector{Bool},edge::Edge,taxa::Union{Vector{String},Vector{Int}})
hardwiredCluster!(v::Vector{Bool},edge::Edge,taxa::Union{Vector{String},Vector{Int}},
                  visited::Vector{Int})
</pre></div>


<p>Calculate the hardwired cluster of <code>node</code>, coded a vector of booleans: true for taxa that are descendent of nodes, false for other taxa (including missing taxa).</p>
<p>The node should belong in a rooted network for which isChild1 is up-to-date. Run directEdges! beforehand. This is very important, otherwise one might enter an infinite loop, and the function does not test for this.</p>
<p>visited: vector of node numbers, of all visited nodes.</p>
<p><strong>Examples: #"</strong></p>
<div class="codehilite"><pre><span></span>julia&gt; net5 = &quot;(A,((B,#H1),(((C,(E)#H2),(#H2,F)),(D)#H1)));&quot; |&gt; readTopology |&gt; directEdges! ;

julia&gt; taxa = net5 |&gt; tipLabels # ABC EF D
6-element Array{String,1}:
 &quot;A&quot;
 &quot;B&quot;
 &quot;C&quot;
 &quot;E&quot;
 &quot;F&quot;
 &quot;D&quot;

julia&gt; hardwiredCluster(net5.edge[12], taxa) # descendants of 12th edge = CEF
6-element Array{Bool,1}:
 false
 false
  true
  true
  true
 false
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/compareNetworks.jl#L145-L182' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.hardwiredClusterDistance' href='#PhyloNetworks.hardwiredClusterDistance'>#</a>
<strong><code>PhyloNetworks.hardwiredClusterDistance</code></strong> &mdash; <em>Function</em>.</p>
<p><code>hardwiredClusterDistance(net1::HybridNetwork, net2::HybridNetwork, rooted::Bool)</code></p>
<p>Takes 2 networks and returns their hardwired cluster distance, that is, the number of hardwired clusters found in one network and not in the other. Note that this is not a distance per se on the full space of hybrid networks: there are pairs of different networks for which this measure is 0. But it is a distance on some network subspaces.</p>
<p>If the 2 networks are trees, this is the Robinson-Foulds distance. If rooted=false, the trees are considered unrooted.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/compareNetworks.jl#L396-L407' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.treeEdgesBootstrap' href='#PhyloNetworks.treeEdgesBootstrap'>#</a>
<strong><code>PhyloNetworks.treeEdgesBootstrap</code></strong> &mdash; <em>Function</em>.</p>
<p><code>treeEdgesBootstrap(boot_net::Vector{HybridNetwork}, ref_net::HybridNetwork)</code></p>
<p>read a list of bootstrap networks (<code>boot_net</code>) and a reference network (<code>ref_net</code>), and calculate the bootstrap support of the tree edges in the reference network. All minor hybrid edges (γ&lt;0.5) are removed to extract the major tree from each network. All remaining edges are tree edges, each associated with a bipartition.</p>
<p>output:</p>
<ul>
<li>a data frame with one row per tree edge and two columns: edge number, bootstrap support (as a percentage)</li>
<li>the major tree from the reference network, where minor hybrid edges (with γ&lt;0.5) have been removed.</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/bootstrap.jl#L475-L489' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.hybridBootstrapSupport' href='#PhyloNetworks.hybridBootstrapSupport'>#</a>
<strong><code>PhyloNetworks.hybridBootstrapSupport</code></strong> &mdash; <em>Function</em>.</p>
<p><code>hybridBootstrapSupport(boot_net::Vector{HybridNetwork}, ref_net::HybridNetwork; rooted=false)</code></p>
<p>Match hybrid nodes in a reference network with those in an array of networks, like bootstrap networks. All networks must be fully resolved, and on the same taxon set. If <code>rooted=true</code>, all networks are assumed to have been properly rooted beforehand. Otherwise, the origin of each hybrid edge is considered as an unrooted bipartition (default).</p>
<p>Two hybrid edges in two networks are said to match if they share the same "hybrid" clade (or recipient) and the same "donor clade", which is a sister to the hybrid clade in the network. Since a hybrid clade has 2 parent edges, it is sister to two clades simultaneously: one is its major sister (following the major hybrid edge with γ&gt;0.5) and one is its minor sister (following the major hybrid edge with γ&lt;0.5).</p>
<p>To calculate these hybrid and sister clades at a given hybrid node, all other hybrid edges are first removed from the network. Then, the hybrid clade is the hardwired cluster (descendants) of either hybrid edge and major/minor clade is the hardwired cluster of the sibling edge of the major/minor hybrid parent. If <code>rooted=false</code>, sister clades are considered as bipartitions.</p>
<p>Output:</p>
<ol>
<li>a "node" data frame (see below)</li>
<li>an "edge" data frame (see below)</li>
<li>a "clade" data frame to describe the make up of all clades found as hybrids or sisters,</li>
</ol>
<p>starting with a column <code>taxa</code> that lists all taxa. All other columns correspond to a given   clade and contain true/false values. <code>true</code> means that a given taxon belongs in a given clade.   For a clade named <code>H1</code>, for instance, and if the data frame was named <code>cla</code>, the   list of taxa in this clade can be obtained with <code>cla[:taxa][cla[:H1]]</code>.</p>
<ol>
<li>an array of gamma values, with one row for each bootstrap network and two columns (major/minor)</li>
</ol>
<p>for each hybrid edge in the reference network. If this hybrid edge was found in the bootstrap network   (i.e. same hybrid and sister clades, after removal of all other hybrid nodes),   its bootstrap gamma value is recorded here. Otherwise, the gamma entry is 0.0.</p>
<ol>
<li>a vector with the number of each hybrid edge in the reference network, in the same order</li>
</ol>
<p>as for the columns in the array of gamma values above.</p>
<p>The "node" data frame has one row per clade and 9 columns giving:</p>
<ul>
<li><strong>clade</strong>: the clade's name, like the taxon name (if a hybrid is a single taxon) or the hybrid tag (like 'H1') in the reference network</li>
<li><strong>node</strong>: the node number in the reference network. missing if the clade is not in this network.</li>
<li><strong>hybridnode</strong>: typically the same node number as above, except for hybrid clades in the reference network. For those, the hybrid node number is listed here.</li>
<li><strong>edge</strong>: number of the parent edge, parent to the node in column 2, if found in the ref network. missing otherwise.</li>
<li><strong>BS_hybrid</strong>: percentage of bootstrap networks in which the clade is found to be a hybrid clade.</li>
<li><strong>BS_sister</strong>: percentage of bootstrap networks in which the clade is found to be sister to some hybrid clade (sum of the next 2 columns)</li>
<li><strong>BS_major_sister</strong>: percentage of bootstrap networks in which the clade is found to be the major sister to some hybrid clade</li>
<li><strong>BS_minor_sister</strong>: same as previous, but minor</li>
<li><strong>BS_hybrid_samesisters</strong>: percentage of bootstrap networks in which the clade is found to be a hybrid and with the same set of sister clades as in the reference network. Applies to hybrid clades found in the reference network only, missing for all other clades.</li>
</ul>
<p>The "edge" data frame has one row for each pair of clades, and 8 columns:</p>
<ul>
<li><strong>edge</strong>: hybrid edge number, if the edge appears in the reference network. missing otherwise.</li>
<li><strong>hybrid_clade</strong>: name of the clade found to be a hybrid, descendent of 'edge'</li>
<li><strong>hybrid</strong>: node number of that clade, if it appears in the reference network. missing otherwise.</li>
<li><strong>sister_clade</strong>: name of the clade that is sister to 'edge', i.e. be sister to a hybrid</li>
<li><strong>sister</strong>: node number of that clade, if in the ref network.</li>
<li><strong>BS_hybrid_edge</strong>: percentage of bootstrap networks in which 'edge' is found to be a hybrid  edge, i.e. when the clade in the 'hybrid' column is found to be a hybrid and the clade in  the 'sister' column is one of its sisters.</li>
<li><strong>BS_major</strong>: percentage of bootstrap networks in which 'edge' is found to be a major hybrid  edge, i.e. when 'hybrid' is found to be a hybrid clade and 'sister' is found to be its  major sister.</li>
<li><strong>BS_minor</strong>: same as previous, but minor</li>
</ul>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/bootstrap.jl#L626-L696' class='documenter-source'>source</a><br></p>
<p><a id='continuous-trait-evolution-1'></a></p>
<h2 id="continuous-trait-evolution">continuous trait evolution</h2>
<p><a id='PhyloNetworks.simulate' href='#PhyloNetworks.simulate'>#</a>
<strong><code>PhyloNetworks.simulate</code></strong> &mdash; <em>Function</em>.</p>
<p><code>simulate(net::HybridNetwork, params::ParamsProcess, checkPreorder=true::Bool)</code></p>
<p>Simualte some traits on <code>net</code> using the parameters <code>params</code>. For now, only parameters of type <a href="./#PhyloNetworks.ParamsBM"><code>ParamsBM</code></a> (Brownian Motion) are accepted.</p>
<p>Assumes that the network is in the pre-order. If <code>checkPreorder=true</code> (default), then it runs function <code>preoder</code> on the network beforehand.</p>
<p>Returns an object of type <a href="./#PhyloNetworks.TraitSimulation"><code>TraitSimulation</code></a>.</p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span>julia&gt; phy = readTopology(joinpath(Pkg.dir(&quot;PhyloNetworks&quot;), &quot;examples&quot;, &quot;carnivores_tree.txt&quot;));

julia&gt; par = ParamsBM(1, 0.1) # BM with expectation 1 and variance 0.1.
PhyloNetworks.ParamsBM:
Parameters of a BM with fixed root:
mu: 1
Sigma2: 0.1


julia&gt; srand(17920921); # Seed for reproducibility

julia&gt; sim = simulate(phy, par) # Simulate on the tree.
PhyloNetworks.TraitSimulation:
Trait simulation results on a network with 16 tips, using a BM model, with parameters:
mu: 1
Sigma2: 0.1


julia&gt; traits = sim[:Tips] # Extract simulated values at the tips.
16-element Array{Float64,1}:
  2.17618
  1.03308
  3.04898
  3.03796
  2.1897
  4.03159
  4.64773
 -0.877285
  4.62512
 -0.511167
  1.35604
 -0.103112
 -2.08847
  2.63991
  2.80512
  3.19109
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L853-L904' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.shiftHybrid' href='#PhyloNetworks.shiftHybrid'>#</a>
<strong><code>PhyloNetworks.shiftHybrid</code></strong> &mdash; <em>Function</em>.</p>
<p><code>shiftHybrid(value::Vector{T} where T&lt;:Real, net::HybridNetwork; checkPreorder=true::Bool)</code></p>
<p>Construct an object <a href="./#PhyloNetworks.ShiftNet"><code>ShiftNet</code></a> with shifts on all the edges below hybrid nodes, with values provided. The vector of values must have the same length as the number of hybrids in the network.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L684-L691' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.getShiftEdgeNumber' href='#PhyloNetworks.getShiftEdgeNumber'>#</a>
<strong><code>PhyloNetworks.getShiftEdgeNumber</code></strong> &mdash; <em>Function</em>.</p>
<p><code>getShiftEdgeNumber(shift::ShiftNet)</code></p>
<p>Get the edge numbers where the shifts are located, for an object <a href="../@ef"><code>ShiftNet</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L702-L706' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.getShiftValue' href='#PhyloNetworks.getShiftValue'>#</a>
<strong><code>PhyloNetworks.getShiftValue</code></strong> &mdash; <em>Function</em>.</p>
<p><code>getShiftValue(shift::ShiftNet)</code></p>
<p>Get the values of the shifts, for an object <a href="../@ef"><code>ShiftNet</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L718-L722' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.phyloNetworklm' href='#PhyloNetworks.phyloNetworklm'>#</a>
<strong><code>PhyloNetworks.phyloNetworklm</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>phyloNetworklm(f, fr, net, model=&quot;BM&quot;,
    fTolRel=1e^-10, fTolAbs=1e^-10, xTolRel=1e^-10, xTolAbs=1e^-10,
    startingValue=0.5)`
</pre></div>


<p>Phylogenetic regression, using the correlation structure induced by the network.</p>
<p>Returns an object of class <a href="./#PhyloNetworks.PhyloNetworkLinearModel"><code>PhyloNetworkLinearModel</code></a>. See documentation for this type and example to see all the functions that can be applied to it.</p>
<p><strong>Arguments</strong></p>
<ul>
<li><code>f::Formula</code>: formula to use for the regression (see the <code>DataFrame</code> package)</li>
<li><code>fr::AbstractDataFrame</code>: DataFrame containing the data and regressors at the tips. It should have an extra column labelled "tipNames", that gives the names of the taxa for each observation.</li>
<li><code>net::HybridNetwork</code>: phylogenetic network to use. Should have labelled tips.</li>
<li><code>model::AbstractString="BM"</code>: the model to use, "BM" (default) or "lambda" (for Pagel's lambda).</li>
<li><code>no_names::Bool=false</code>: if <code>true</code>, force the function to ignore the tips names. The data is then assumed to be in the same order as the tips of the network. Default to false, setting it to true is dangerous, and strongly discouraged.</li>
</ul>
<p>If <code>model="lambda"</code>, there are a some parameters to control the optimization in lambda:</p>
<ul>
<li><code>fTolRel::AbstractFloat=1e-10</code>: relative tolerance on the likelihood value for the optimization in lambda.</li>
<li><code>fTolAbs::AbstractFloat=1e-10</code>: absolute tolerance on the likelihood value for the optimization in lambda.</li>
<li><code>xTolRel::AbstractFloat=1e-10</code>: relative tolerance on the parameter value for the optimization in lambda.</li>
<li><code>xTolAbs::AbstractFloat=1e-10</code>: absolute tolerance on the parameter value for the optimization in lambda.</li>
<li><code>startingValue::Real=0.5</code>: the starting value for the parameter in the optimization in lambda.</li>
</ul>
<p><strong>See also</strong></p>
<p>Type <a href="./#PhyloNetworks.PhyloNetworkLinearModel"><code>PhyloNetworkLinearModel</code></a>, Function <a href="./#PhyloNetworks.ancestralStateReconstruction"><code>ancestralStateReconstruction</code></a></p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span>julia&gt; phy = readTopology(joinpath(Pkg.dir(&quot;PhyloNetworks&quot;), &quot;examples&quot;, &quot;caudata_tree.txt&quot;));

julia&gt; using CSV # to read data file, next

julia&gt; dat = CSV.read(joinpath(Pkg.dir(&quot;PhyloNetworks&quot;), &quot;examples&quot;, &quot;caudata_trait.txt&quot;));

julia&gt; using StatsModels # for stat model formulas

julia&gt; fitBM = phyloNetworklm(@formula(trait ~ 1), dat, phy);

julia&gt; fitBM # Shows a summary
StatsModels.DataFrameRegressionModel{PhyloNetworks.PhyloNetworkLinearModel,Array{Float64,2}}

Formula: trait ~ +1

Model: BM

Parameter(s) Estimates:
Sigma2: 0.00294521

Coefficients:
             Estimate Std.Error t value Pr(&gt;|t|)
(Intercept)     4.679  0.330627 14.1519   &lt;1e-31

Log Likelihood: -78.9611507833
AIC: 161.9223015666

julia&gt; round(sigma2_estim(fitBM), 6) # rounding for jldoctest convenience
0.002945

julia&gt; round(mu_estim(fitBM), 4)
4.679

julia&gt; using StatsBase # for aic() stderr() loglikelihood() etc.

julia&gt; round(loglikelihood(fitBM), 10)
-78.9611507833

julia&gt; round(aic(fitBM), 10)
161.9223015666

julia&gt; round(aicc(fitBM), 10)
161.9841572367

julia&gt; round(bic(fitBM), 10)
168.4887090241

julia&gt; coef(fitBM)
1-element Array{Float64,1}:
 4.679

julia&gt; confint(fitBM)
1×2 Array{Float64,2}:
 4.02696  5.33104

julia&gt; abs(round(r2(fitBM), 10)) # absolute value for jldoctest convenience
0.0

julia&gt; abs(round(adjr2(fitBM), 10))
0.0

julia&gt; vcov(fitBM)
1×1 Array{Float64,2}:
 0.109314

julia&gt; residuals(fitBM)
197-element Array{Float64,1}:
 -0.237648
 -0.357937
 -0.159387
 -0.691868
 -0.323977
 -0.270452
 -0.673486
 -0.584654
 -0.279882
 -0.302175
  ⋮
 -0.777026
 -0.385121
 -0.443444
 -0.327303
 -0.525953
 -0.673486
 -0.603158
 -0.211712
 -0.439833

julia&gt; model_response(fitBM)
197-element Array{Float64,1}:
 4.44135
 4.32106
 4.51961
 3.98713
 4.35502
 4.40855
 4.00551
 4.09434
 4.39912
 4.37682
 ⋮
 3.90197
 4.29388
 4.23555
 4.3517
 4.15305
 4.00551
 4.07584
 4.46729
 4.23917

julia&gt; predict(fitBM)
197-element Array{Float64,1}:
 4.679
 4.679
 4.679
 4.679
 4.679
 4.679
 4.679
 4.679
 4.679
 4.679
 ⋮
 4.679
 4.679
 4.679
 4.679
 4.679
 4.679
 4.679
 4.679
 4.679
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1353-L1518' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.sigma2_estim' href='#PhyloNetworks.sigma2_estim'>#</a>
<strong><code>PhyloNetworks.sigma2_estim</code></strong> &mdash; <em>Function</em>.</p>
<p><code>sigma2_estim(m::PhyloNetworkLinearModel)</code> Estimated variance for a fitted object.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1661-L1664' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.mu_estim' href='#PhyloNetworks.mu_estim'>#</a>
<strong><code>PhyloNetworks.mu_estim</code></strong> &mdash; <em>Function</em>.</p>
<p><code>mu_estim(m::PhyloNetworkLinearModel)</code> Estimated root value for a fitted object.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1670-L1673' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.lambda_estim' href='#PhyloNetworks.lambda_estim'>#</a>
<strong><code>PhyloNetworks.lambda_estim</code></strong> &mdash; <em>Function</em>.</p>
<p><code>lambda_estim(m::PhyloNetworkLinearModel)</code> Estimated lambda parameter for a fitted object.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1693-L1696' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.ancestralStateReconstruction' href='#PhyloNetworks.ancestralStateReconstruction'>#</a>
<strong><code>PhyloNetworks.ancestralStateReconstruction</code></strong> &mdash; <em>Function</em>.</p>
<p><code>ancestralStateReconstruction(net::HybridNetwork, Y::Vector, params::ParamsBM)</code></p>
<p>Compute the conditional expectations and variances of the ancestral (un-observed) traits values at the internal nodes of the phylogenetic network (<code>net</code>), given the values of the traits at the tips of the network (<code>Y</code>) and some known parameters of the process used for trait evolution (<code>params</code>, only BM with fixed root works for now).</p>
<p>This function assumes that the parameters of the process are known. For a more general function, see <code>ancestralStateReconstruction(obj::PhyloNetworkLinearModel[, X_n::Matrix])</code>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1967-L1979' class='documenter-source'>source</a><br></p>
<p><code>ancestralStateReconstruction(obj::PhyloNetworkLinearModel[, X_n::Matrix])</code></p>
<p>Function to find the ancestral traits reconstruction on a network, given an object fitted by function <a href="./#PhyloNetworks.phyloNetworklm"><code>phyloNetworklm</code></a>. By default, the function assumes that the regressor is just an intercept. If the value of the regressor for all the ancestral states is known, it can be entered in X_n, a matrix with as many columns as the number of predictors used, and as many lines as the number of unknown nodes or tips.</p>
<p>Returns an object of type <a href="./#PhyloNetworks.ReconstructedStates"><code>ReconstructedStates</code></a>. See documentation for this type and examples for functions that can be applied to it.</p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span>julia<span class="o">&gt;</span> using CSV <span class="c1"># to read data file</span>

julia<span class="o">&gt;</span> phy <span class="o">=</span> readTopology<span class="p">(</span>joinpath<span class="p">(</span>Pkg.dir<span class="p">(</span><span class="s">&quot;PhyloNetworks&quot;</span><span class="p">),</span> <span class="s">&quot;examples&quot;</span><span class="p">,</span> <span class="s">&quot;carnivores_tree.txt&quot;</span><span class="p">));</span>

julia<span class="o">&gt;</span> dat <span class="o">=</span> CSV.read<span class="p">(</span>joinpath<span class="p">(</span>Pkg.dir<span class="p">(</span><span class="s">&quot;PhyloNetworks&quot;</span><span class="p">),</span> <span class="s">&quot;examples&quot;</span><span class="p">,</span> <span class="s">&quot;carnivores_trait.txt&quot;</span><span class="p">));</span>

julia<span class="o">&gt;</span> using StatsModels <span class="c1"># for statistical model formulas</span>

julia<span class="o">&gt;</span> fitBM <span class="o">=</span> phyloNetworklm<span class="p">(</span><span class="o">@</span>formula<span class="p">(</span>trait <span class="o">~</span> <span class="m">1</span><span class="p">),</span> dat<span class="p">,</span> phy<span class="p">);</span>

julia<span class="o">&gt;</span> ancStates <span class="o">=</span> ancestralStateReconstruction<span class="p">(</span>fitBM<span class="p">)</span> <span class="c1"># Should produce a warning, as variance is unknown.</span>
WARNING<span class="o">:</span> These prediction intervals show uncertainty <span class="kr">in</span> ancestral values<span class="p">,</span>
assuming that the estimated variance rate of evolution is correct.
Additional uncertainty <span class="kr">in</span> the estimation of this variance rate is
ignored<span class="p">,</span> so prediction intervals should be larger.
PhyloNetworks.ReconstructedStates<span class="o">:</span>
     Node index     Pred.       Min. Max. <span class="p">(</span><span class="m">95</span>%<span class="p">)</span>
           <span class="m">-5.0</span>   <span class="m">1.32139</span>  <span class="m">-0.288423</span>     <span class="m">2.9312</span>
           <span class="m">-8.0</span>   <span class="m">1.03258</span>  <span class="m">-0.539072</span>    <span class="m">2.60423</span>
           <span class="m">-7.0</span>   <span class="m">1.41575</span> <span class="m">-0.0934395</span>    <span class="m">2.92495</span>
           <span class="m">-6.0</span>   <span class="m">1.39417</span> <span class="m">-0.0643135</span>    <span class="m">2.85265</span>
           <span class="m">-4.0</span>   <span class="m">1.39961</span> <span class="m">-0.0603343</span>    <span class="m">2.85955</span>
           <span class="m">-3.0</span>   <span class="m">1.51341</span>  <span class="m">-0.179626</span>    <span class="m">3.20644</span>
          <span class="m">-13.0</span>    <span class="m">5.3192</span>    <span class="m">3.96695</span>    <span class="m">6.67145</span>
          <span class="m">-12.0</span>   <span class="m">4.51176</span>    <span class="m">2.94268</span>    <span class="m">6.08085</span>
          <span class="m">-16.0</span>   <span class="m">1.50947</span>  <span class="m">0.0290151</span>    <span class="m">2.98992</span>
          <span class="m">-15.0</span>   <span class="m">1.67425</span>   <span class="m">0.241696</span>    <span class="m">3.10679</span>
          <span class="m">-14.0</span>   <span class="m">1.80309</span>   <span class="m">0.355568</span>     <span class="m">3.2506</span>
          <span class="m">-11.0</span>    <span class="m">2.7351</span>    <span class="m">1.21896</span>    <span class="m">4.25123</span>
          <span class="m">-10.0</span>   <span class="m">2.73217</span>    <span class="m">1.16545</span>    <span class="m">4.29889</span>
           <span class="m">-9.0</span>   <span class="m">2.41132</span>   <span class="m">0.639075</span>    <span class="m">4.18357</span>
           <span class="m">-2.0</span>   <span class="m">2.04138</span> <span class="m">-0.0340955</span>    <span class="m">4.11686</span>
           <span class="m">14.0</span>   <span class="m">1.64289</span>    <span class="m">1.64289</span>    <span class="m">1.64289</span>
            <span class="m">8.0</span>   <span class="m">1.67724</span>    <span class="m">1.67724</span>    <span class="m">1.67724</span>
            <span class="m">5.0</span>  <span class="m">0.331568</span>   <span class="m">0.331568</span>   <span class="m">0.331568</span>
            <span class="m">2.0</span>   <span class="m">2.27395</span>    <span class="m">2.27395</span>    <span class="m">2.27395</span>
            <span class="m">4.0</span>  <span class="m">0.275237</span>   <span class="m">0.275237</span>   <span class="m">0.275237</span>
            <span class="m">6.0</span>   <span class="m">3.39094</span>    <span class="m">3.39094</span>    <span class="m">3.39094</span>
           <span class="m">13.0</span>  <span class="m">0.355799</span>   <span class="m">0.355799</span>   <span class="m">0.355799</span>
           <span class="m">15.0</span>  <span class="m">0.542565</span>   <span class="m">0.542565</span>   <span class="m">0.542565</span>
            <span class="m">7.0</span>  <span class="m">0.773436</span>   <span class="m">0.773436</span>   <span class="m">0.773436</span>
           <span class="m">10.0</span>   <span class="m">6.94985</span>    <span class="m">6.94985</span>    <span class="m">6.94985</span>
           <span class="m">11.0</span>   <span class="m">4.78323</span>    <span class="m">4.78323</span>    <span class="m">4.78323</span>
           <span class="m">12.0</span>   <span class="m">5.33016</span>    <span class="m">5.33016</span>    <span class="m">5.33016</span>
            <span class="m">1.0</span> <span class="m">-0.122604</span>  <span class="m">-0.122604</span>  <span class="m">-0.122604</span>
           <span class="m">16.0</span>   <span class="m">0.73989</span>    <span class="m">0.73989</span>    <span class="m">0.73989</span>
            <span class="m">9.0</span>   <span class="m">4.84236</span>    <span class="m">4.84236</span>    <span class="m">4.84236</span>
            <span class="m">3.0</span>    <span class="m">1.0695</span>     <span class="m">1.0695</span>     <span class="m">1.0695</span>


julia<span class="o">&gt;</span> expectations<span class="p">(</span>ancStates<span class="p">)</span>
<span class="m">31</span>×<span class="m">2</span> DataFrames.DataFrame
│ Row │ nodeNumber │ condExpectation │
├─────┼────────────┼─────────────────┤
│ <span class="m">1</span>   │ <span class="m">-5</span>         │ <span class="m">1.32139</span>         │
│ <span class="m">2</span>   │ <span class="m">-8</span>         │ <span class="m">1.03258</span>         │
│ <span class="m">3</span>   │ <span class="m">-7</span>         │ <span class="m">1.41575</span>         │
│ <span class="m">4</span>   │ <span class="m">-6</span>         │ <span class="m">1.39417</span>         │
│ <span class="m">5</span>   │ <span class="m">-4</span>         │ <span class="m">1.39961</span>         │
│ <span class="m">6</span>   │ <span class="m">-3</span>         │ <span class="m">1.51341</span>         │
│ <span class="m">7</span>   │ <span class="m">-13</span>        │ <span class="m">5.3192</span>          │
│ <span class="m">8</span>   │ <span class="m">-12</span>        │ <span class="m">4.51176</span>         │
⋮
│ <span class="m">23</span>  │ <span class="m">15</span>         │ <span class="m">0.542565</span>        │
│ <span class="m">24</span>  │ <span class="m">7</span>          │ <span class="m">0.773436</span>        │
│ <span class="m">25</span>  │ <span class="m">10</span>         │ <span class="m">6.94985</span>         │
│ <span class="m">26</span>  │ <span class="m">11</span>         │ <span class="m">4.78323</span>         │
│ <span class="m">27</span>  │ <span class="m">12</span>         │ <span class="m">5.33016</span>         │
│ <span class="m">28</span>  │ <span class="m">1</span>          │ <span class="m">-0.122604</span>       │
│ <span class="m">29</span>  │ <span class="m">16</span>         │ <span class="m">0.73989</span>         │
│ <span class="m">30</span>  │ <span class="m">9</span>          │ <span class="m">4.84236</span>         │
│ <span class="m">31</span>  │ <span class="m">3</span>          │ <span class="m">1.0695</span>          │

julia<span class="o">&gt;</span> predint<span class="p">(</span>ancStates<span class="p">)</span>
<span class="m">31</span>×<span class="m">2</span> Array<span class="p">{</span>Float64<span class="p">,</span><span class="m">2</span><span class="p">}</span><span class="o">:</span>
 <span class="m">-0.288423</span>    <span class="m">2.9312</span>
 <span class="m">-0.539072</span>    <span class="m">2.60423</span>
 <span class="m">-0.0934395</span>   <span class="m">2.92495</span>
 <span class="m">-0.0643135</span>   <span class="m">2.85265</span>
 <span class="m">-0.0603343</span>   <span class="m">2.85955</span>
 <span class="m">-0.179626</span>    <span class="m">3.20644</span>
  <span class="m">3.96695</span>     <span class="m">6.67145</span>
  <span class="m">2.94268</span>     <span class="m">6.08085</span>
  <span class="m">0.0290151</span>   <span class="m">2.98992</span>
  <span class="m">0.241696</span>    <span class="m">3.10679</span>
  ⋮
  <span class="m">0.542565</span>    <span class="m">0.542565</span>
  <span class="m">0.773436</span>    <span class="m">0.773436</span>
  <span class="m">6.94985</span>     <span class="m">6.94985</span>
  <span class="m">4.78323</span>     <span class="m">4.78323</span>
  <span class="m">5.33016</span>     <span class="m">5.33016</span>
 <span class="m">-0.122604</span>   <span class="m">-0.122604</span>
  <span class="m">0.73989</span>     <span class="m">0.73989</span>
  <span class="m">4.84236</span>     <span class="m">4.84236</span>
  <span class="m">1.0695</span>      <span class="m">1.0695</span>

julia<span class="o">&gt;</span> expectationsPlot<span class="p">(</span>ancStates<span class="p">)</span> <span class="c1"># format the ancestral states</span>
<span class="m">31</span>×<span class="m">2</span> DataFrames.DataFrame
│ Row │ nodeNumber │ PredInt │
├─────┼────────────┼─────────┤
│ <span class="m">1</span>   │ <span class="m">-5</span>         │ <span class="m">1.32</span>    │
│ <span class="m">2</span>   │ <span class="m">-8</span>         │ <span class="m">1.03</span>    │
│ <span class="m">3</span>   │ <span class="m">-7</span>         │ <span class="m">1.42</span>    │
│ <span class="m">4</span>   │ <span class="m">-6</span>         │ <span class="m">1.39</span>    │
│ <span class="m">5</span>   │ <span class="m">-4</span>         │ <span class="m">1.4</span>     │
│ <span class="m">6</span>   │ <span class="m">-3</span>         │ <span class="m">1.51</span>    │
│ <span class="m">7</span>   │ <span class="m">-13</span>        │ <span class="m">5.32</span>    │
│ <span class="m">8</span>   │ <span class="m">-12</span>        │ <span class="m">4.51</span>    │
⋮
│ <span class="m">23</span>  │ <span class="m">15</span>         │ <span class="m">0.54</span>    │
│ <span class="m">24</span>  │ <span class="m">7</span>          │ <span class="m">0.77</span>    │
│ <span class="m">25</span>  │ <span class="m">10</span>         │ <span class="m">6.95</span>    │
│ <span class="m">26</span>  │ <span class="m">11</span>         │ <span class="m">4.78</span>    │
│ <span class="m">27</span>  │ <span class="m">12</span>         │ <span class="m">5.33</span>    │
│ <span class="m">28</span>  │ <span class="m">1</span>          │ <span class="m">-0.12</span>   │
│ <span class="m">29</span>  │ <span class="m">16</span>         │ <span class="m">0.74</span>    │
│ <span class="m">30</span>  │ <span class="m">9</span>          │ <span class="m">4.84</span>    │
│ <span class="m">31</span>  │ <span class="m">3</span>          │ <span class="m">1.07</span>    │

julia<span class="o">&gt;</span> using PhyloPlots <span class="c1"># next: plot ancestral states on the tree</span>

julia<span class="o">&gt;</span> plot<span class="p">(</span>phy<span class="p">,</span> <span class="o">:</span>RCall<span class="p">,</span> nodeLabel <span class="o">=</span> expectationsPlot<span class="p">(</span>ancStates<span class="p">));</span>

julia<span class="o">&gt;</span> predintPlot<span class="p">(</span>ancStates<span class="p">)</span>
<span class="m">31</span>×<span class="m">2</span> DataFrames.DataFrame
│ Row │ nodeNumber │ PredInt       │
├─────┼────────────┼───────────────┤
│ <span class="m">1</span>   │ <span class="m">-5</span>         │ <span class="p">[</span><span class="m">-0.29</span><span class="p">,</span> <span class="m">2.93</span><span class="p">]</span> │
│ <span class="m">2</span>   │ <span class="m">-8</span>         │ <span class="p">[</span><span class="m">-0.54</span><span class="p">,</span> <span class="m">2.6</span><span class="p">]</span>  │
│ <span class="m">3</span>   │ <span class="m">-7</span>         │ <span class="p">[</span><span class="m">-0.09</span><span class="p">,</span> <span class="m">2.92</span><span class="p">]</span> │
│ <span class="m">4</span>   │ <span class="m">-6</span>         │ <span class="p">[</span><span class="m">-0.06</span><span class="p">,</span> <span class="m">2.85</span><span class="p">]</span> │
│ <span class="m">5</span>   │ <span class="m">-4</span>         │ <span class="p">[</span><span class="m">-0.06</span><span class="p">,</span> <span class="m">2.86</span><span class="p">]</span> │
│ <span class="m">6</span>   │ <span class="m">-3</span>         │ <span class="p">[</span><span class="m">-0.18</span><span class="p">,</span> <span class="m">3.21</span><span class="p">]</span> │
│ <span class="m">7</span>   │ <span class="m">-13</span>        │ <span class="p">[</span><span class="m">3.97</span><span class="p">,</span> <span class="m">6.67</span><span class="p">]</span>  │
│ <span class="m">8</span>   │ <span class="m">-12</span>        │ <span class="p">[</span><span class="m">2.94</span><span class="p">,</span> <span class="m">6.08</span><span class="p">]</span>  │
⋮
│ <span class="m">23</span>  │ <span class="m">15</span>         │ <span class="m">0.54</span>          │
│ <span class="m">24</span>  │ <span class="m">7</span>          │ <span class="m">0.77</span>          │
│ <span class="m">25</span>  │ <span class="m">10</span>         │ <span class="m">6.95</span>          │
│ <span class="m">26</span>  │ <span class="m">11</span>         │ <span class="m">4.78</span>          │
│ <span class="m">27</span>  │ <span class="m">12</span>         │ <span class="m">5.33</span>          │
│ <span class="m">28</span>  │ <span class="m">1</span>          │ <span class="m">-0.12</span>         │
│ <span class="m">29</span>  │ <span class="m">16</span>         │ <span class="m">0.74</span>          │
│ <span class="m">30</span>  │ <span class="m">9</span>          │ <span class="m">4.84</span>          │
│ <span class="m">31</span>  │ <span class="m">3</span>          │ <span class="m">1.07</span>          │

julia<span class="o">&gt;</span> plot<span class="p">(</span>phy<span class="p">,</span> <span class="o">:</span>RCall<span class="p">,</span> nodeLabel <span class="o">=</span> predintPlot<span class="p">(</span>ancStates<span class="p">));</span>

julia<span class="o">&gt;</span> using Missings<span class="p">;</span> <span class="c1"># preparation to mask 2 values next, for taxa in rows 2 and 5</span>

julia<span class="o">&gt;</span> dat<span class="p">[</span><span class="o">:</span>trait<span class="p">]</span> <span class="o">=</span> allowmissing<span class="p">(</span>dat<span class="p">[</span><span class="o">:</span>trait<span class="p">]);</span>

julia<span class="o">&gt;</span> dat<span class="p">[[</span><span class="m">2</span><span class="p">,</span> <span class="m">5</span><span class="p">],</span> <span class="o">:</span>trait<span class="p">]</span> <span class="o">=</span> <span class="kp">missing</span><span class="p">;</span> <span class="c1"># missing values allowed to fit model</span>

julia<span class="o">&gt;</span> fitBM <span class="o">=</span> phyloNetworklm<span class="p">(</span><span class="o">@</span>formula<span class="p">(</span>trait <span class="o">~</span> <span class="m">1</span><span class="p">),</span> dat<span class="p">,</span> phy<span class="p">);</span>

julia<span class="o">&gt;</span> ancStates <span class="o">=</span> ancestralStateReconstruction<span class="p">(</span>fitBM<span class="p">);</span>
WARNING<span class="o">:</span> These prediction intervals show uncertainty <span class="kr">in</span> ancestral values<span class="p">,</span>
assuming that the estimated variance rate of evolution is correct.
Additional uncertainty <span class="kr">in</span> the estimation of this variance rate is
ignored<span class="p">,</span> so prediction intervals should be larger.

julia<span class="o">&gt;</span> expectations<span class="p">(</span>ancStates<span class="p">)</span>
<span class="m">31</span>×<span class="m">2</span> DataFrames.DataFrame
│ Row │ nodeNumber │ condExpectation │
├─────┼────────────┼─────────────────┤
│ <span class="m">1</span>   │ <span class="m">-5</span>         │ <span class="m">1.42724</span>         │
│ <span class="m">2</span>   │ <span class="m">-8</span>         │ <span class="m">1.35185</span>         │
│ <span class="m">3</span>   │ <span class="m">-7</span>         │ <span class="m">1.61993</span>         │
│ <span class="m">4</span>   │ <span class="m">-6</span>         │ <span class="m">1.54198</span>         │
│ <span class="m">5</span>   │ <span class="m">-4</span>         │ <span class="m">1.53916</span>         │
│ <span class="m">6</span>   │ <span class="m">-3</span>         │ <span class="m">1.64984</span>         │
│ <span class="m">7</span>   │ <span class="m">-13</span>        │ <span class="m">5.33508</span>         │
│ <span class="m">8</span>   │ <span class="m">-12</span>        │ <span class="m">4.55109</span>         │
⋮
│ <span class="m">23</span>  │ <span class="m">15</span>         │ <span class="m">0.542565</span>        │
│ <span class="m">24</span>  │ <span class="m">7</span>          │ <span class="m">0.773436</span>        │
│ <span class="m">25</span>  │ <span class="m">10</span>         │ <span class="m">6.94985</span>         │
│ <span class="m">26</span>  │ <span class="m">11</span>         │ <span class="m">4.78323</span>         │
│ <span class="m">27</span>  │ <span class="m">12</span>         │ <span class="m">5.33016</span>         │
│ <span class="m">28</span>  │ <span class="m">1</span>          │ <span class="m">-0.122604</span>       │
│ <span class="m">29</span>  │ <span class="m">16</span>         │ <span class="m">0.73989</span>         │
│ <span class="m">30</span>  │ <span class="m">9</span>          │ <span class="m">4.84236</span>         │
│ <span class="m">31</span>  │ <span class="m">3</span>          │ <span class="m">1.0695</span>          │

julia<span class="o">&gt;</span> predint<span class="p">(</span>ancStates<span class="p">)</span>
<span class="m">31</span>×<span class="m">2</span> Array<span class="p">{</span>Float64<span class="p">,</span><span class="m">2</span><span class="p">}</span><span class="o">:</span>
 <span class="m">-0.31245</span>     <span class="m">3.16694</span>
 <span class="m">-0.625798</span>    <span class="m">3.3295</span>
 <span class="m">-0.110165</span>    <span class="m">3.35002</span>
 <span class="m">-0.0710391</span>   <span class="m">3.15501</span>
 <span class="m">-0.0675924</span>   <span class="m">3.14591</span>
 <span class="m">-0.197236</span>    <span class="m">3.49692</span>
  <span class="m">3.89644</span>     <span class="m">6.77373</span>
  <span class="m">2.8741</span>      <span class="m">6.22808</span>
 <span class="m">-0.0358627</span>   <span class="m">3.12834</span>
  <span class="m">0.182594</span>    <span class="m">3.2534</span>
  ⋮
  <span class="m">0.542565</span>    <span class="m">0.542565</span>
  <span class="m">0.773436</span>    <span class="m">0.773436</span>
  <span class="m">6.94985</span>     <span class="m">6.94985</span>
  <span class="m">4.78323</span>     <span class="m">4.78323</span>
  <span class="m">5.33016</span>     <span class="m">5.33016</span>
 <span class="m">-0.122604</span>   <span class="m">-0.122604</span>
  <span class="m">0.73989</span>     <span class="m">0.73989</span>
  <span class="m">4.84236</span>     <span class="m">4.84236</span>
  <span class="m">1.0695</span>      <span class="m">1.0695</span>

julia<span class="o">&gt;</span> expectationsPlot<span class="p">(</span>ancStates<span class="p">)</span> <span class="c1"># format node &lt;-&gt; ancestral state</span>
<span class="m">31</span>×<span class="m">2</span> DataFrames.DataFrame
│ Row │ nodeNumber │ PredInt │
├─────┼────────────┼─────────┤
│ <span class="m">1</span>   │ <span class="m">-5</span>         │ <span class="m">1.43</span>    │
│ <span class="m">2</span>   │ <span class="m">-8</span>         │ <span class="m">1.35</span>    │
│ <span class="m">3</span>   │ <span class="m">-7</span>         │ <span class="m">1.62</span>    │
│ <span class="m">4</span>   │ <span class="m">-6</span>         │ <span class="m">1.54</span>    │
│ <span class="m">5</span>   │ <span class="m">-4</span>         │ <span class="m">1.54</span>    │
│ <span class="m">6</span>   │ <span class="m">-3</span>         │ <span class="m">1.65</span>    │
│ <span class="m">7</span>   │ <span class="m">-13</span>        │ <span class="m">5.34</span>    │
│ <span class="m">8</span>   │ <span class="m">-12</span>        │ <span class="m">4.55</span>    │
⋮
│ <span class="m">23</span>  │ <span class="m">15</span>         │ <span class="m">0.54</span>    │
│ <span class="m">24</span>  │ <span class="m">7</span>          │ <span class="m">0.77</span>    │
│ <span class="m">25</span>  │ <span class="m">10</span>         │ <span class="m">6.95</span>    │
│ <span class="m">26</span>  │ <span class="m">11</span>         │ <span class="m">4.78</span>    │
│ <span class="m">27</span>  │ <span class="m">12</span>         │ <span class="m">5.33</span>    │
│ <span class="m">28</span>  │ <span class="m">1</span>          │ <span class="m">-0.12</span>   │
│ <span class="m">29</span>  │ <span class="m">16</span>         │ <span class="m">0.74</span>    │
│ <span class="m">30</span>  │ <span class="m">9</span>          │ <span class="m">4.84</span>    │
│ <span class="m">31</span>  │ <span class="m">3</span>          │ <span class="m">1.07</span>    │

julia<span class="o">&gt;</span> plot<span class="p">(</span>phy<span class="p">,</span> <span class="o">:</span>RCall<span class="p">,</span> nodeLabel <span class="o">=</span> expectationsPlot<span class="p">(</span>ancStates<span class="p">));</span>

julia<span class="o">&gt;</span> predintPlot<span class="p">(</span>ancStates<span class="p">)</span> <span class="c1"># prediction intervals, in data frame, useful to plot</span>
<span class="m">31</span>×<span class="m">2</span> DataFrames.DataFrame
│ Row │ nodeNumber │ PredInt       │
├─────┼────────────┼───────────────┤
│ <span class="m">1</span>   │ <span class="m">-5</span>         │ <span class="p">[</span><span class="m">-0.31</span><span class="p">,</span> <span class="m">3.17</span><span class="p">]</span> │
│ <span class="m">2</span>   │ <span class="m">-8</span>         │ <span class="p">[</span><span class="m">-0.63</span><span class="p">,</span> <span class="m">3.33</span><span class="p">]</span> │
│ <span class="m">3</span>   │ <span class="m">-7</span>         │ <span class="p">[</span><span class="m">-0.11</span><span class="p">,</span> <span class="m">3.35</span><span class="p">]</span> │
│ <span class="m">4</span>   │ <span class="m">-6</span>         │ <span class="p">[</span><span class="m">-0.07</span><span class="p">,</span> <span class="m">3.16</span><span class="p">]</span> │
│ <span class="m">5</span>   │ <span class="m">-4</span>         │ <span class="p">[</span><span class="m">-0.07</span><span class="p">,</span> <span class="m">3.15</span><span class="p">]</span> │
│ <span class="m">6</span>   │ <span class="m">-3</span>         │ <span class="p">[</span><span class="m">-0.2</span><span class="p">,</span> <span class="m">3.5</span><span class="p">]</span>   │
│ <span class="m">7</span>   │ <span class="m">-13</span>        │ <span class="p">[</span><span class="m">3.9</span><span class="p">,</span> <span class="m">6.77</span><span class="p">]</span>   │
│ <span class="m">8</span>   │ <span class="m">-12</span>        │ <span class="p">[</span><span class="m">2.87</span><span class="p">,</span> <span class="m">6.23</span><span class="p">]</span>  │
⋮
│ <span class="m">23</span>  │ <span class="m">15</span>         │ <span class="m">0.54</span>          │
│ <span class="m">24</span>  │ <span class="m">7</span>          │ <span class="m">0.77</span>          │
│ <span class="m">25</span>  │ <span class="m">10</span>         │ <span class="m">6.95</span>          │
│ <span class="m">26</span>  │ <span class="m">11</span>         │ <span class="m">4.78</span>          │
│ <span class="m">27</span>  │ <span class="m">12</span>         │ <span class="m">5.33</span>          │
│ <span class="m">28</span>  │ <span class="m">1</span>          │ <span class="m">-0.12</span>         │
│ <span class="m">29</span>  │ <span class="m">16</span>         │ <span class="m">0.74</span>          │
│ <span class="m">30</span>  │ <span class="m">9</span>          │ <span class="m">4.84</span>          │
│ <span class="m">31</span>  │ <span class="m">3</span>          │ <span class="m">1.07</span>          │

julia<span class="o">&gt;</span> plot<span class="p">(</span>phy<span class="p">,</span> <span class="o">:</span>RCall<span class="p">,</span> nodeLabel <span class="o">=</span> predintPlot<span class="p">(</span>ancStates<span class="p">));</span>
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L2090-L2364' class='documenter-source'>source</a><br></p>
<p><code>ancestralStateReconstruction(fr::AbstractDataFrame, net::HybridNetwork; kwargs...)</code></p>
<p>Function to find the ancestral traits reconstruction on a network, given some data at the tips. Uses function <a href="./#PhyloNetworks.phyloNetworklm"><code>phyloNetworklm</code></a> to perform a phylogenetic regression of the data against an intercept (amounts to fitting an evolutionary model on the network, BM being the only option available for now).</p>
<p>See documentation on <a href="./#PhyloNetworks.phyloNetworklm"><code>phyloNetworklm</code></a> and <code>ancestralStateReconstruction(obj::PhyloNetworkLinearModel[, X_n::Matrix])</code> for further details.</p>
<p>Returns an object of type <a href="./#PhyloNetworks.ReconstructedStates"><code>ReconstructedStates</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L2385-L2397' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.expectations' href='#PhyloNetworks.expectations'>#</a>
<strong><code>PhyloNetworks.expectations</code></strong> &mdash; <em>Function</em>.</p>
<p><code>expectations(obj::ReconstructedStates)</code> Estimated reconstructed states at the nodes and tips.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1864-L1867' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.predint' href='#PhyloNetworks.predint'>#</a>
<strong><code>PhyloNetworks.predint</code></strong> &mdash; <em>Function</em>.</p>
<p><code>predint(obj::ReconstructedStates; level=0.95::Real)</code> Prediction intervals with level <code>level</code> for internal nodes and missing tips.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1902-L1905' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.expectationsPlot' href='#PhyloNetworks.expectationsPlot'>#</a>
<strong><code>PhyloNetworks.expectationsPlot</code></strong> &mdash; <em>Function</em>.</p>
<p><code>expectationsPlot(obj::ReconstructedStates)</code> Compute and format the expected reconstructed states for the plotting function. The resulting dataframe can be readily used as a <code>nodeLabel</code> argument to <code>plot</code> from package <a href="https://github.com/cecileane/PhyloPlots.jl"><code>PhyloPlots</code></a>. Keyword argument <code>markMissing</code> is a string that is appended to predicted tip values, so that they can be distinguished from the actual datapoints. Default to "*". Set to "" to remove any visual cue.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1872-L1880' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.predintPlot' href='#PhyloNetworks.predintPlot'>#</a>
<strong><code>PhyloNetworks.predintPlot</code></strong> &mdash; <em>Function</em>.</p>
<p><code>predintPlot(obj::ReconstructedStates; level=0.95::Real, withExp=false::Bool)</code> Compute and format the prediction intervals for the plotting function. The resulting dataframe can be readily used as a <code>nodeLabel</code> argument to <code>plot</code> from package <a href="https://github.com/cecileane/PhyloPlots.jl"><code>PhyloPlots</code></a>. Keyworks argument <code>level</code> control the confidence level of the prediction interval. If <code>withExp</code> is set to true, then the best predicted value is also shown along with the interval.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L1924-L1932' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.descendenceMatrix' href='#PhyloNetworks.descendenceMatrix'>#</a>
<strong><code>PhyloNetworks.descendenceMatrix</code></strong> &mdash; <em>Function</em>.</p>
<p><code>descendenceMatrix(net::HybridNetwork; checkPreorder=true::Bool)</code></p>
<p>This function computes the inciednce matrix between all the nodes of a network. It assumes that the network is in the pre-order. If checkPreorder is true (default), then it runs function <code>preoder</code> on the network beforehand.</p>
<p>Returns an object of type <a href="../internals/#PhyloNetworks.MatrixTopologicalOrder"><code>MatrixTopologicalOrder</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L340-L349' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.regressorShift' href='#PhyloNetworks.regressorShift'>#</a>
<strong><code>PhyloNetworks.regressorShift</code></strong> &mdash; <em>Function</em>.</p>
<p><code>regressorShift(node::Vector{Node}, net::HybridNetwork; checkPreorder=true::Bool)</code></p>
<p><code>regressorShift(edge::Vector{Edge}, net::HybridNetwork; checkPreorder=true::Bool)</code></p>
<p>Compute the regressor vectors associated with shifts on edges that are above nodes <code>node</code>, or on edges <code>edge</code>, on a network <code>net</code>. It uses function <a href="./#PhyloNetworks.descendenceMatrix"><code>descendenceMatrix</code></a>, so <code>net</code> might be modified to sort it in a pre-order. Return a <code>DataFrame</code> with as many rows as there are tips in net, and a column for each shift, each labelled according to the pattern shift_{number_of_edge}. It has an aditional column labelled <code>tipNames</code> to allow easy fitting afterward (see example).</p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span><span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">net</span> <span class="o">=</span> <span class="nt">readTopology</span><span class="o">(</span><span class="s2">&quot;(A:2.5,((B:1,#H1:0.5::0.4):1,(C:1,(D:0.5)#H1:0.5::0.6):1):0.5);&quot;</span><span class="o">);</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">preorder</span><span class="o">!(</span><span class="nt">net</span><span class="o">)</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">using</span> <span class="nt">PhyloPlots</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">plot</span><span class="o">(</span><span class="nt">net</span><span class="o">,</span> <span class="p">:</span><span class="nd">RCall</span><span class="o">,</span> <span class="nt">showNodeNumber</span><span class="o">=</span><span class="nt">true</span><span class="o">);</span> <span class="err">#</span> <span class="nt">to</span> <span class="nt">locate</span> <span class="nt">nodes</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">nodes_shifts</span> <span class="o">=</span> <span class="nt">indexin</span><span class="o">(</span><span class="cp">[</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mi">5</span><span class="cp">]</span><span class="o">,</span> <span class="cp">[</span><span class="nx">n.number</span> <span class="nx">for</span> <span class="n">n</span> <span class="k">in</span> <span class="nx">net.node</span><span class="cp">]</span><span class="o">)</span> <span class="err">#</span> <span class="nt">Put</span> <span class="nt">a</span> <span class="nt">shift</span> <span class="nt">on</span> <span class="nt">edges</span> <span class="nt">ending</span> <span class="nt">at</span> <span class="nt">nodes</span> <span class="nt">1</span> <span class="nt">and</span> <span class="nt">-5</span>
<span class="nt">2-element</span> <span class="nt">Array</span><span class="p">{</span><span class="err">Int64,1</span><span class="p">}</span><span class="o">:</span>
 <span class="nt">1</span>
 <span class="nt">7</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">params</span> <span class="o">=</span> <span class="nt">ParamsBM</span><span class="o">(</span><span class="nt">10</span><span class="o">,</span> <span class="nt">0</span><span class="p">.</span><span class="nc">1</span><span class="o">,</span> <span class="nt">ShiftNet</span><span class="o">(</span><span class="nt">net</span><span class="p">.</span><span class="nc">node</span><span class="cp">[</span><span class="nx">nodes_shifts</span><span class="cp">]</span><span class="o">,</span> <span class="cp">[</span><span class="mf">3.0</span><span class="p">,</span> <span class="o">-</span><span class="mf">3.0</span><span class="cp">]</span><span class="o">,</span>  <span class="nt">net</span><span class="o">))</span>
<span class="nt">PhyloNetworks</span><span class="p">.</span><span class="nc">ParamsBM</span><span class="o">:</span>
<span class="nt">Parameters</span> <span class="nt">of</span> <span class="nt">a</span> <span class="nt">BM</span> <span class="nt">with</span> <span class="nt">fixed</span> <span class="nt">root</span><span class="o">:</span>
<span class="nt">mu</span><span class="o">:</span> <span class="nt">10</span>
<span class="nt">Sigma2</span><span class="o">:</span> <span class="nt">0</span><span class="p">.</span><span class="nc">1</span>

<span class="nt">There</span> <span class="nt">are</span> <span class="nt">2</span> <span class="nt">shifts</span> <span class="nt">on</span> <span class="nt">the</span> <span class="nt">network</span><span class="o">:</span>
     <span class="nt">Edge</span> <span class="nt">Number</span> <span class="nt">Shift</span> <span class="nt">Value</span>
             <span class="nt">8</span><span class="p">.</span><span class="nc">0</span>        <span class="nt">-3</span><span class="p">.</span><span class="nc">0</span>
             <span class="nt">1</span><span class="p">.</span><span class="nc">0</span>         <span class="nt">3</span><span class="p">.</span><span class="nc">0</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">srand</span><span class="o">(</span><span class="nt">2468</span><span class="o">);</span> <span class="err">#</span> <span class="nt">sets</span> <span class="nt">the</span> <span class="nt">seed</span> <span class="nt">for</span> <span class="nt">reproducibility</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">sim</span> <span class="o">=</span> <span class="nt">simulate</span><span class="o">(</span><span class="nt">net</span><span class="o">,</span> <span class="nt">params</span><span class="o">);</span> <span class="err">#</span> <span class="nt">simulate</span> <span class="nt">a</span> <span class="nt">dataset</span> <span class="nt">with</span> <span class="nt">shifts</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">using</span> <span class="nt">DataFrames</span> <span class="err">#</span> <span class="nt">to</span> <span class="nt">handle</span> <span class="nt">data</span> <span class="nt">frames</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">dat</span> <span class="o">=</span> <span class="nt">DataFrame</span><span class="o">(</span><span class="nt">trait</span> <span class="o">=</span> <span class="nt">sim</span><span class="cp">[</span><span class="p">:</span><span class="nx">Tips</span><span class="cp">]</span><span class="o">,</span> <span class="nt">tipNames</span> <span class="o">=</span> <span class="nt">sim</span><span class="p">.</span><span class="nc">M</span><span class="p">.</span><span class="nc">tipNames</span><span class="o">)</span>
<span class="nt">4</span><span class="err">×</span><span class="nt">2</span> <span class="nt">DataFrames</span><span class="p">.</span><span class="nc">DataFrame</span>
<span class="err">│</span> <span class="nt">Row</span> <span class="err">│</span> <span class="nt">trait</span>   <span class="err">│</span> <span class="nt">tipNames</span> <span class="err">│</span>
<span class="err">├─────┼─────────┼──────────┤</span>
<span class="err">│</span> <span class="nt">1</span>   <span class="err">│</span> <span class="nt">13</span><span class="p">.</span><span class="nc">392</span>  <span class="err">│</span> <span class="nt">A</span>        <span class="err">│</span>
<span class="err">│</span> <span class="nt">2</span>   <span class="err">│</span> <span class="nt">9</span><span class="p">.</span><span class="nc">55741</span> <span class="err">│</span> <span class="nt">B</span>        <span class="err">│</span>
<span class="err">│</span> <span class="nt">3</span>   <span class="err">│</span> <span class="nt">7</span><span class="p">.</span><span class="nc">17704</span> <span class="err">│</span> <span class="nt">C</span>        <span class="err">│</span>
<span class="err">│</span> <span class="nt">4</span>   <span class="err">│</span> <span class="nt">7</span><span class="p">.</span><span class="nc">88906</span> <span class="err">│</span> <span class="nt">D</span>        <span class="err">│</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">dfr_shift</span> <span class="o">=</span> <span class="nt">regressorShift</span><span class="o">(</span><span class="nt">net</span><span class="p">.</span><span class="nc">node</span><span class="cp">[</span><span class="nx">nodes_shifts</span><span class="cp">]</span><span class="o">,</span> <span class="nt">net</span><span class="o">)</span> <span class="err">#</span> <span class="nt">the</span> <span class="nt">reressors</span> <span class="nt">matching</span> <span class="nt">the</span> <span class="nt">shifts</span><span class="o">.</span>
<span class="nt">4</span><span class="err">×</span><span class="nt">3</span> <span class="nt">DataFrames</span><span class="p">.</span><span class="nc">DataFrame</span>
<span class="err">│</span> <span class="nt">Row</span> <span class="err">│</span> <span class="nt">shift_1</span> <span class="err">│</span> <span class="nt">shift_8</span> <span class="err">│</span> <span class="nt">tipNames</span> <span class="err">│</span>
<span class="err">├─────┼─────────┼─────────┼──────────┤</span>
<span class="err">│</span> <span class="nt">1</span>   <span class="err">│</span> <span class="nt">1</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">A</span>        <span class="err">│</span>
<span class="err">│</span> <span class="nt">2</span>   <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">B</span>        <span class="err">│</span>
<span class="err">│</span> <span class="nt">3</span>   <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">1</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">C</span>        <span class="err">│</span>
<span class="err">│</span> <span class="nt">4</span>   <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">6</span>     <span class="err">│</span> <span class="nt">D</span>        <span class="err">│</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">dfr</span> <span class="o">=</span> <span class="nt">join</span><span class="o">(</span><span class="nt">dat</span><span class="o">,</span> <span class="nt">dfr_shift</span><span class="o">,</span> <span class="nt">on</span><span class="o">=</span><span class="p">:</span><span class="nd">tipNames</span><span class="o">);</span> <span class="err">#</span> <span class="nt">join</span> <span class="nt">data</span> <span class="nt">and</span> <span class="nt">regressors</span> <span class="nt">in</span> <span class="nt">a</span> <span class="nt">single</span> <span class="nt">dataframe</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">using</span> <span class="nt">StatsModels</span> <span class="err">#</span> <span class="nt">for</span> <span class="nt">statistical</span> <span class="nt">model</span> <span class="nt">formulas</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">fitBM</span> <span class="o">=</span> <span class="nt">phyloNetworklm</span><span class="o">(</span><span class="p">@</span><span class="k">formula</span><span class="o">(</span><span class="nt">trait</span> <span class="o">~</span> <span class="nt">shift_1</span> <span class="o">+</span> <span class="nt">shift_8</span><span class="o">),</span> <span class="nt">dfr</span><span class="o">,</span> <span class="nt">net</span><span class="o">)</span> <span class="err">#</span> <span class="nt">actual</span> <span class="nt">fit</span>
<span class="nt">StatsModels</span><span class="p">.</span><span class="nc">DataFrameRegressionModel</span><span class="p">{</span><span class="nt">PhyloNetworks</span><span class="p">.</span><span class="nc">PhyloNetworkLinearModel</span><span class="o">,</span><span class="nt">Array</span><span class="p">{</span><span class="err">Float64,2</span><span class="p">}}</span>

<span class="nt">Formula</span><span class="o">:</span> <span class="nt">trait</span> <span class="o">~</span> <span class="nt">1</span> <span class="o">+</span> <span class="nt">shift_1</span> <span class="o">+</span> <span class="nt">shift_8</span>

<span class="nt">Model</span><span class="o">:</span> <span class="nt">BM</span>

<span class="nt">Parameter</span><span class="o">(</span><span class="nt">s</span><span class="o">)</span> <span class="nt">Estimates</span><span class="o">:</span>
<span class="nt">Sigma2</span><span class="o">:</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0112618</span>

<span class="nt">Coefficients</span><span class="o">:</span>
             <span class="nt">Estimate</span> <span class="nt">Std</span><span class="p">.</span><span class="nc">Error</span>  <span class="nt">t</span> <span class="nt">value</span> <span class="nt">Pr</span><span class="o">(&gt;|</span><span class="nt">t</span><span class="o">|)</span>
<span class="o">(</span><span class="nt">Intercept</span><span class="o">)</span>   <span class="nt">9</span><span class="p">.</span><span class="nc">48238</span>  <span class="nt">0</span><span class="p">.</span><span class="nc">327089</span>  <span class="nt">28</span><span class="p">.</span><span class="nc">9902</span>   <span class="nt">0</span><span class="p">.</span><span class="nc">0220</span>
<span class="nt">shift_1</span>        <span class="nt">3</span><span class="p">.</span><span class="nc">9096</span>   <span class="nt">0</span><span class="p">.</span><span class="nc">46862</span>  <span class="nt">8</span><span class="p">.</span><span class="nc">34279</span>   <span class="nt">0</span><span class="p">.</span><span class="nc">0759</span>
<span class="nt">shift_8</span>       <span class="nt">-2</span><span class="p">.</span><span class="nc">4179</span>  <span class="nt">0</span><span class="p">.</span><span class="nc">422825</span> <span class="nt">-5</span><span class="p">.</span><span class="nc">71843</span>   <span class="nt">0</span><span class="p">.</span><span class="nc">1102</span>

<span class="nt">Log</span> <span class="nt">Likelihood</span><span class="o">:</span> <span class="nt">1</span><span class="p">.</span><span class="nc">8937302027</span>
<span class="nt">AIC</span><span class="o">:</span> <span class="nt">4</span><span class="p">.</span><span class="nc">2125395947</span>
</pre></div>


<p><strong>See also</strong></p>
<p><a href="./#PhyloNetworks.phyloNetworklm"><code>phyloNetworklm</code></a>, <a href="./#PhyloNetworks.descendenceMatrix"><code>descendenceMatrix</code></a>, <a href="./#PhyloNetworks.regressorHybrid"><code>regressorHybrid</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L386-L475' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.regressorHybrid' href='#PhyloNetworks.regressorHybrid'>#</a>
<strong><code>PhyloNetworks.regressorHybrid</code></strong> &mdash; <em>Function</em>.</p>
<p><code>regressorHybrid(net::HybridNetwork; checkPreorder=true::Bool)</code></p>
<p>Compute the regressor vectors associated with shifts on edges that imediatly below all hybrid nodes of <code>net</code>. It uses function <a href="./#PhyloNetworks.descendenceMatrix"><code>descendenceMatrix</code></a> through a call to <a href="./#PhyloNetworks.regressorShift"><code>regressorShift</code></a>, so <code>net</code> might be modified to sort it in a pre-order. Return a <code>DataFrame</code> with as many rows as there are tips in net, and a column for each hybrid, each labelled according to the pattern shift_{number_of_edge}. It has an aditional column labelled <code>tipNames</code> to allow easy fitting afterward (see example).</p>
<p>This function can be used to test for heterosis.</p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span><span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">using</span> <span class="nt">DataFrames</span> <span class="err">#</span> <span class="nt">Needed</span> <span class="nt">to</span> <span class="nt">handle</span> <span class="nt">data</span> <span class="nt">frames</span><span class="o">.</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">net</span> <span class="o">=</span> <span class="nt">readTopology</span><span class="o">(</span><span class="s2">&quot;(A:2.5,((B:1,#H1:0.5::0.4):1,(C:1,(D:0.5)#H1:0.5::0.6):1):0.5);&quot;</span><span class="o">);</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">preorder</span><span class="o">!(</span><span class="nt">net</span><span class="o">)</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">using</span> <span class="nt">PhyloPlots</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">plot</span><span class="o">(</span><span class="nt">net</span><span class="o">,</span> <span class="p">:</span><span class="nd">RCall</span><span class="o">,</span> <span class="nt">showNodeNumber</span><span class="o">=</span><span class="nt">true</span><span class="o">);</span> <span class="err">#</span> <span class="nt">to</span> <span class="nt">locate</span> <span class="nt">nodes</span><span class="o">:</span> <span class="nt">node</span> <span class="nt">5</span> <span class="nt">is</span> <span class="nt">child</span> <span class="nt">of</span> <span class="nt">hybrid</span> <span class="nt">node</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">nodes_hybrids</span> <span class="o">=</span> <span class="nt">indexin</span><span class="o">(</span><span class="cp">[</span><span class="mi">5</span><span class="cp">]</span><span class="o">,</span> <span class="cp">[</span><span class="nx">n.number</span> <span class="nx">for</span> <span class="n">n</span> <span class="k">in</span> <span class="nx">net.node</span><span class="cp">]</span><span class="o">)</span> <span class="err">#</span> <span class="nt">Put</span> <span class="nt">a</span> <span class="nt">shift</span> <span class="nt">on</span> <span class="nt">edges</span> <span class="nt">below</span> <span class="nt">hybrids</span>
<span class="nt">1-element</span> <span class="nt">Array</span><span class="p">{</span><span class="err">Int64,1</span><span class="p">}</span><span class="o">:</span>
 <span class="nt">5</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">params</span> <span class="o">=</span> <span class="nt">ParamsBM</span><span class="o">(</span><span class="nt">10</span><span class="o">,</span> <span class="nt">0</span><span class="p">.</span><span class="nc">1</span><span class="o">,</span> <span class="nt">ShiftNet</span><span class="o">(</span><span class="nt">net</span><span class="p">.</span><span class="nc">node</span><span class="cp">[</span><span class="nx">nodes_hybrids</span><span class="cp">]</span><span class="o">,</span> <span class="cp">[</span><span class="mf">3.0</span><span class="cp">]</span><span class="o">,</span>  <span class="nt">net</span><span class="o">))</span>
<span class="nt">PhyloNetworks</span><span class="p">.</span><span class="nc">ParamsBM</span><span class="o">:</span>
<span class="nt">Parameters</span> <span class="nt">of</span> <span class="nt">a</span> <span class="nt">BM</span> <span class="nt">with</span> <span class="nt">fixed</span> <span class="nt">root</span><span class="o">:</span>
<span class="nt">mu</span><span class="o">:</span> <span class="nt">10</span>
<span class="nt">Sigma2</span><span class="o">:</span> <span class="nt">0</span><span class="p">.</span><span class="nc">1</span>

<span class="nt">There</span> <span class="nt">are</span> <span class="nt">1</span> <span class="nt">shifts</span> <span class="nt">on</span> <span class="nt">the</span> <span class="nt">network</span><span class="o">:</span>
     <span class="nt">Edge</span> <span class="nt">Number</span> <span class="nt">Shift</span> <span class="nt">Value</span>
             <span class="nt">6</span><span class="p">.</span><span class="nc">0</span>         <span class="nt">3</span><span class="p">.</span><span class="nc">0</span>



<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">srand</span><span class="o">(</span><span class="nt">2468</span><span class="o">);</span> <span class="err">#</span> <span class="nt">sets</span> <span class="nt">the</span> <span class="nt">seed</span> <span class="nt">for</span> <span class="nt">reproducibility</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">sim</span> <span class="o">=</span> <span class="nt">simulate</span><span class="o">(</span><span class="nt">net</span><span class="o">,</span> <span class="nt">params</span><span class="o">);</span> <span class="err">#</span> <span class="nt">simulate</span> <span class="nt">a</span> <span class="nt">dataset</span> <span class="nt">with</span> <span class="nt">shifts</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">dat</span> <span class="o">=</span> <span class="nt">DataFrame</span><span class="o">(</span><span class="nt">trait</span> <span class="o">=</span> <span class="nt">sim</span><span class="cp">[</span><span class="p">:</span><span class="nx">Tips</span><span class="cp">]</span><span class="o">,</span> <span class="nt">tipNames</span> <span class="o">=</span> <span class="nt">sim</span><span class="p">.</span><span class="nc">M</span><span class="p">.</span><span class="nc">tipNames</span><span class="o">)</span>
<span class="nt">4</span><span class="err">×</span><span class="nt">2</span> <span class="nt">DataFrames</span><span class="p">.</span><span class="nc">DataFrame</span>
<span class="err">│</span> <span class="nt">Row</span> <span class="err">│</span> <span class="nt">trait</span>   <span class="err">│</span> <span class="nt">tipNames</span> <span class="err">│</span>
<span class="err">├─────┼─────────┼──────────┤</span>
<span class="err">│</span> <span class="nt">1</span>   <span class="err">│</span> <span class="nt">10</span><span class="p">.</span><span class="nc">392</span>  <span class="err">│</span> <span class="nt">A</span>        <span class="err">│</span>
<span class="err">│</span> <span class="nt">2</span>   <span class="err">│</span> <span class="nt">9</span><span class="p">.</span><span class="nc">55741</span> <span class="err">│</span> <span class="nt">B</span>        <span class="err">│</span>
<span class="err">│</span> <span class="nt">3</span>   <span class="err">│</span> <span class="nt">10</span><span class="p">.</span><span class="nc">177</span>  <span class="err">│</span> <span class="nt">C</span>        <span class="err">│</span>
<span class="err">│</span> <span class="nt">4</span>   <span class="err">│</span> <span class="nt">12</span><span class="p">.</span><span class="nc">6891</span> <span class="err">│</span> <span class="nt">D</span>        <span class="err">│</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">dfr_hybrid</span> <span class="o">=</span> <span class="nt">regressorHybrid</span><span class="o">(</span><span class="nt">net</span><span class="o">)</span> <span class="err">#</span> <span class="nt">the</span> <span class="nt">reressors</span> <span class="nt">matching</span> <span class="nt">the</span> <span class="nt">hybrids</span><span class="o">.</span>
<span class="nt">4</span><span class="err">×</span><span class="nt">3</span> <span class="nt">DataFrames</span><span class="p">.</span><span class="nc">DataFrame</span>
<span class="err">│</span> <span class="nt">Row</span> <span class="err">│</span> <span class="nt">shift_6</span> <span class="err">│</span> <span class="nt">tipNames</span> <span class="err">│</span> <span class="nt">sum</span> <span class="err">│</span>
<span class="err">├─────┼─────────┼──────────┼─────┤</span>
<span class="err">│</span> <span class="nt">1</span>   <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">A</span>        <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span> <span class="err">│</span>
<span class="err">│</span> <span class="nt">2</span>   <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">B</span>        <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span> <span class="err">│</span>
<span class="err">│</span> <span class="nt">3</span>   <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">C</span>        <span class="err">│</span> <span class="nt">0</span><span class="p">.</span><span class="nc">0</span> <span class="err">│</span>
<span class="err">│</span> <span class="nt">4</span>   <span class="err">│</span> <span class="nt">1</span><span class="p">.</span><span class="nc">0</span>     <span class="err">│</span> <span class="nt">D</span>        <span class="err">│</span> <span class="nt">1</span><span class="p">.</span><span class="nc">0</span> <span class="err">│</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">dfr</span> <span class="o">=</span> <span class="nt">join</span><span class="o">(</span><span class="nt">dat</span><span class="o">,</span> <span class="nt">dfr_hybrid</span><span class="o">,</span> <span class="nt">on</span><span class="o">=</span><span class="p">:</span><span class="nd">tipNames</span><span class="o">);</span> <span class="err">#</span> <span class="nt">join</span> <span class="nt">data</span> <span class="nt">and</span> <span class="nt">regressors</span> <span class="nt">in</span> <span class="nt">a</span> <span class="nt">single</span> <span class="nt">dataframe</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">using</span> <span class="nt">StatsModels</span>

<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">fitBM</span> <span class="o">=</span> <span class="nt">phyloNetworklm</span><span class="o">(</span><span class="p">@</span><span class="k">formula</span><span class="o">(</span><span class="nt">trait</span> <span class="o">~</span> <span class="nt">shift_6</span><span class="o">),</span> <span class="nt">dfr</span><span class="o">,</span> <span class="nt">net</span><span class="o">)</span> <span class="err">#</span> <span class="nt">actual</span> <span class="nt">fit</span>
<span class="nt">StatsModels</span><span class="p">.</span><span class="nc">DataFrameRegressionModel</span><span class="p">{</span><span class="nt">PhyloNetworks</span><span class="p">.</span><span class="nc">PhyloNetworkLinearModel</span><span class="o">,</span><span class="nt">Array</span><span class="p">{</span><span class="err">Float64,2</span><span class="p">}}</span>

<span class="nt">Formula</span><span class="o">:</span> <span class="nt">trait</span> <span class="o">~</span> <span class="nt">1</span> <span class="o">+</span> <span class="nt">shift_6</span>

<span class="nt">Model</span><span class="o">:</span> <span class="nt">BM</span>

<span class="nt">Parameter</span><span class="o">(</span><span class="nt">s</span><span class="o">)</span> <span class="nt">Estimates</span><span class="o">:</span>
<span class="nt">Sigma2</span><span class="o">:</span> <span class="nt">0</span><span class="p">.</span><span class="nc">041206</span>

<span class="nt">Coefficients</span><span class="o">:</span>
             <span class="nt">Estimate</span> <span class="nt">Std</span><span class="p">.</span><span class="nc">Error</span> <span class="nt">t</span> <span class="nt">value</span> <span class="nt">Pr</span><span class="o">(&gt;|</span><span class="nt">t</span><span class="o">|)</span>
<span class="o">(</span><span class="nt">Intercept</span><span class="o">)</span>    <span class="nt">10</span><span class="p">.</span><span class="nc">064</span>  <span class="nt">0</span><span class="p">.</span><span class="nc">277959</span> <span class="nt">36</span><span class="p">.</span><span class="nc">2068</span>   <span class="nt">0</span><span class="p">.</span><span class="nc">0008</span>
<span class="nt">shift_6</span>       <span class="nt">2</span><span class="p">.</span><span class="nc">72526</span>  <span class="nt">0</span><span class="p">.</span><span class="nc">315456</span> <span class="nt">8</span><span class="p">.</span><span class="nc">63912</span>   <span class="nt">0</span><span class="p">.</span><span class="nc">0131</span>

<span class="nt">Log</span> <span class="nt">Likelihood</span><span class="o">:</span> <span class="nt">-0</span><span class="p">.</span><span class="nc">7006021946</span>
<span class="nt">AIC</span><span class="o">:</span> <span class="nt">7</span><span class="p">.</span><span class="nc">4012043891</span>
</pre></div>


<p><strong>See also</strong></p>
<p><a href="./#PhyloNetworks.phyloNetworklm"><code>phyloNetworklm</code></a>, <a href="./#PhyloNetworks.descendenceMatrix"><code>descendenceMatrix</code></a>, <a href="./#PhyloNetworks.regressorShift"><code>regressorShift</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L520-L608' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.sharedPathMatrix' href='#PhyloNetworks.sharedPathMatrix'>#</a>
<strong><code>PhyloNetworks.sharedPathMatrix</code></strong> &mdash; <em>Function</em>.</p>
<p><code>sharedPathMatrix(net::HybridNetwork; checkPreorder=true::Bool)</code></p>
<p>This function computes the shared path matrix between all the nodes of a network. It assumes that the network is in the pre-order. If checkPreorder is true (default), then it runs function <code>preoder</code> on the network beforehand.</p>
<p>Returns an object of type <a href="../internals/#PhyloNetworks.MatrixTopologicalOrder"><code>MatrixTopologicalOrder</code></a>.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/traits.jl#L278-L287' class='documenter-source'>source</a><br></p>
<p><a id='discrete-trait-evolution-1'></a></p>
<h2 id="discrete-trait-evolution">discrete trait evolution</h2>
<p><a id='PhyloNetworks.parsimonySoftwired' href='#PhyloNetworks.parsimonySoftwired'>#</a>
<strong><code>PhyloNetworks.parsimonySoftwired</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>parsimonySoftwired(net, tipdata)
parsimonySoftwired(net, species, sequences)
</pre></div>


<p>Calculate the most parsimonious (MP) score of a network given a discrete character at the tips using the dynamic programming algorithm given in Fischer et al. (2015). The softwired parsimony concept is used: where the number of state transitions is minimized over all trees displayed in the network.</p>
<p>Data can given in one of the following:</p>
<ul>
<li><code>tipdata</code>: data frame for a single trait, in which case the taxon names</li>
</ul>
<p>are to appear in column 1 or in a column named "taxon" or "species", and trait values are to appear in column 2 or in a column named "trait".</p>
<ul>
<li><code>tipdata</code>: dictionary taxon =&gt; state, for a single trait.</li>
<li><code>species</code>: array of strings, and <code>sequences</code>: array of sequences,  in the order corresponding to the order of species names.</li>
</ul>
<p><strong>References</strong></p>
<ol>
<li>Fischer, M., van Iersel, L., Kelk, S., Scornavacca, C. (2015). On computing the Maximum Parsimony score of a phylogenetic network. SIAM J. Discrete Math., 29(1):559-585.</li>
</ol>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/parsimony.jl#L212-L235' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.parsimonyGF' href='#PhyloNetworks.parsimonyGF'>#</a>
<strong><code>PhyloNetworks.parsimonyGF</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>parsimonyGF(net, tip_dictionary, criterion=:softwired)
parsimonyGF(net, species, sequenceData, criterion=:softwired)
</pre></div>


<p>Calculate the most parsimonious score of a network given discrete characters at the tips using a general framework (Van Iersel et al. 2018) allowing for various parsimony criteria: softwired (default), hardwired, parental etc. The complexity of the algorithm is exponential in the level of the network, that is, the maximum number of hybridizations in a single blob (Fischer et al. 2015).</p>
<p>Data can given in one of the following:</p>
<ul>
<li><code>tipdata</code>: data frame for a single trait, in which case the taxon names</li>
</ul>
<p>are to appear in column 1 or in a column named "taxon" or "species", and trait values are to appear in column 2 or in a column named "trait".</p>
<ul>
<li><code>tipdata</code>: dictionary taxon =&gt; state, for a single trait.</li>
<li><code>species</code>: array of strings, and <code>sequences</code>: array of sequences,   in the order corresponding to the order of species names.</li>
</ul>
<p><strong>References</strong></p>
<ol>
<li>Leo Van Iersel, Mark Jones, Celine Scornavacca (2017). Improved Maximum Parsimony Models for Phylogenetic Networks, Systematic Biology, (https://doi.org/10.1093/sysbio/syx094).</li>
<li>Fischer, M., van Iersel, L., Kelk, S., Scornavacca, C. (2015). On computing the Maximum Parsimony score of a phylogenetic network. SIAM J. Discrete Math., 29(1):559-585.</li>
</ol>
<p>Use the recursive helper function <a href="../internals/#PhyloNetworks.parsimonyBottomUpGF!-Tuple{PhyloNetworks.Node,PhyloNetworks.Node,Integer,AbstractArray,AbstractArray,AbstractArray,AbstractArray}"><code>parsimonyBottomUpGF!</code></a>. Use the fields <code>isChild1</code>, <code>isExtBadTriangle</code> to know which nodes are at the root of a blob, and <code>fromBadDiamondI</code> to know which edges are cut (below the minor parent of each hybrid).</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/parsimony.jl#L390-L425' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.nStates' href='#PhyloNetworks.nStates'>#</a>
<strong><code>PhyloNetworks.nStates</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>nStates(model)
</pre></div>


<p>Number of character states for a given trait evolution model.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L138-L142' class='documenter-source'>source</a><br></p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span>julia&gt; m1 = BinaryTraitSubstitutionModel(1.0, 2.0)
julia&gt; nStates(m1)
2
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L145-L153' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.Q' href='#PhyloNetworks.Q'>#</a>
<strong><code>PhyloNetworks.Q</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>Q(model)
</pre></div>


<p>Substitution rate matrix for a given substitution model: Q[i,j] is the rate of transitioning from state i to state j.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L14-L19' class='documenter-source'>source</a><br></p>
<p>For a BinaryTraitSubstitutionModel, the rate matrix Q is of the form:</p>
<div class="codehilite"><pre><span></span>-α  α
 β -β
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L110-L115' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.P' href='#PhyloNetworks.P'>#</a>
<strong><code>PhyloNetworks.P</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>P(mod, t)
</pre></div>


<p>Probability transition matrix for a <a href="./#PhyloNetworks.TraitSubstitutionModel"><code>TraitSubstitutionModel</code></a>, of the form</p>
<div class="codehilite"><pre><span></span>P[1,1] ... P[1,k]
   .          .
   .          .
P[k,1] ... P[k,k]
</pre></div>


<p>where P[i,j] is the probability of ending in state j after time t, given that the process started in state i.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L52-L64' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>P(mod, t::Array{Float64})
</pre></div>


<p>When applied to a general substitution model, matrix exponentiation is used. The time argument <code>t</code> can be an array.</p>
<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L70-L75' class='documenter-source'>source</a><br></p>
<p><a id='PhyloNetworks.randomTrait' href='#PhyloNetworks.randomTrait'>#</a>
<strong><code>PhyloNetworks.randomTrait</code></strong> &mdash; <em>Function</em>.</p>
<div class="codehilite"><pre><span></span>randomTrait(model, t, start)
randomTrait!(end, model, t, start)
</pre></div>


<p>Simulate traits along one edge of length t. <code>start</code> must be a vector of integers, each representing the starting value of one trait. The bang version (ending with !) uses the vector <code>end</code> to store the simulated values.</p>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span>julia&gt; m1 = PhyloNetworks.BinaryTraitSubstitutionModel(1.0, 2.0) 
julia&gt; srand(12345);
julia&gt; randomTrait(m1, 0.2, [1,2,1,2,2])
 5-element Array{Int64,1}:
 1
 2
 1
 2
 2
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L260-L280' class='documenter-source'>source</a><br></p>
<div class="codehilite"><pre><span></span>randomTrait(mod, net; ntraits=1, keepInternal=true, checkPreorder=true)
</pre></div>


<p>Simulate evolution of discrete traits on a rooted evolutionary network based on the supplied evolutionary model. Trait sampling is uniform at the root.</p>
<p><strong>Arguments</strong></p>
<ul>
<li>ntraits: number of traits to be simulated (default: 1 trait).</li>
<li>keepInternal: if true, export character states at all nodes, including internal nodes. if false, export character states at tips only.</li>
</ul>
<p><strong>Output</strong></p>
<ul>
<li>matrix of character states with one row per trait, one column per node</li>
<li>array of node labels (for tips) or node numbers (for internal nodes) in the same order as columns in the character state matrix</li>
</ul>
<p><strong>Examples</strong></p>
<div class="codehilite"><pre><span></span><span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">m1</span> <span class="o">=</span> <span class="nt">PhyloNetworks</span><span class="p">.</span><span class="nc">BinaryTraitSubstitutionModel</span><span class="o">(</span><span class="nt">1</span><span class="p">.</span><span class="nc">0</span><span class="o">,</span> <span class="nt">2</span><span class="p">.</span><span class="nc">0</span><span class="o">)</span> 
<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">net</span> <span class="o">=</span> <span class="nt">readTopology</span><span class="o">(</span><span class="s2">&quot;(A:1.0,(B:1.0,(C:1.0,D:1.0):1.0):1.0);&quot;</span><span class="o">)</span>
<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">srand</span><span class="o">(</span><span class="nt">12345</span><span class="o">);</span>
<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">a</span><span class="o">,</span><span class="nt">b</span> <span class="o">=</span> <span class="nt">randomTrait</span><span class="o">(</span><span class="nt">m1</span><span class="o">,</span> <span class="nt">net</span><span class="o">)</span>
 <span class="o">(</span><span class="cp">[</span><span class="mi">1</span> <span class="mi">2</span> <span class="err">…</span> <span class="mi">1</span> <span class="mi">2</span><span class="cp">]</span><span class="o">,</span> <span class="nt">String</span><span class="cp">[</span><span class="s2">&quot;-2&quot;</span><span class="p">,</span> <span class="s2">&quot;-3&quot;</span><span class="p">,</span> <span class="s2">&quot;-4&quot;</span><span class="p">,</span> <span class="s2">&quot;D&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;B&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="cp">]</span><span class="o">)</span>
<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">a</span>
 <span class="nt">1</span><span class="err">×</span><span class="nt">7</span> <span class="nt">Array</span><span class="p">{</span><span class="err">Int64,2</span><span class="p">}</span><span class="o">:</span>
 <span class="nt">1</span>  <span class="nt">2</span>  <span class="nt">1</span>  <span class="nt">1</span>  <span class="nt">1</span>  <span class="nt">1</span>  <span class="nt">2</span>
<span class="nt">julia</span><span class="o">&gt;</span> <span class="nt">b</span>
 <span class="nt">7-element</span> <span class="nt">Array</span><span class="p">{</span><span class="err">String,1</span><span class="p">}</span><span class="o">:</span>
 <span class="s2">&quot;-2&quot;</span>
 <span class="s2">&quot;-3&quot;</span>
 <span class="s2">&quot;-4&quot;</span>
 <span class="s2">&quot;D&quot;</span>
 <span class="s2">&quot;C&quot;</span>
 <span class="s2">&quot;B&quot;</span>
 <span class="s2">&quot;A&quot;</span>
</pre></div>


<p><a target='_blank' href='https://github.com/crsl4/PhyloNetworks.jl/blob/eb4a37c6b696fb9c42053d7c9773910542e33b57/src/substitutionModels.jl#L296-L332' class='documenter-source'>source</a><br></p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../../man/trait_tree/" title="Continuous Trait Evolution" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Continuous Trait Evolution
              </span>
            </div>
          </a>
        
        
          <a href="../internals/" title="Internals" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Internals
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="http://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/application.8eb9be28.js"></script>
      
      <script>app.initialize({version:"0.17.3",url:{base:"../.."}})</script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../../assets/mathjaxhelper.js"></script>
      
    
    
      
    
  </body>
</html>