



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="PhyloNetworks is a Julia package for the manipulation, visualization and inference of phylogenetic networks.">
      
      
      
        <meta name="author" content="Claudia Sol&iacute;s-Lemus">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-0.17.4, mkdocs-material-2.9.2">
    
    
      
        <title>TICR pipeline - PhyloNetworks.jl</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/application.ba0fd1a6.css">
      
        <link rel="stylesheet" href="../../assets/stylesheets/application-palette.6079476c.css">
      
      
        
        
        <meta name="theme-color" content="#009688">
      
    
    
      <script src="../../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../../assets/fonts/material-icons.css">
    
      <link rel="stylesheet" href="../../assets/Documenter.css">
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="teal" data-md-color-accent="teal">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#ticr-pipeline" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../.." title="PhyloNetworks.jl" class="md-header-nav__button md-logo">
          
            <img src="../../snaq_small.png" width="24" height="24">
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                PhyloNetworks.jl
              </span>
              <span class="md-header-nav__topic">
                TICR pipeline
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/crsl4/PhyloNetworks.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <span class="md-nav__button md-logo">
      
        <img src="../../snaq_small.png" width="48" height="48">
      
    </span>
    PhyloNetworks.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/crsl4/PhyloNetworks.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../.." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" checked>
    
    <label class="md-nav__link" for="nav-2">
      Manual
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        Manual
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../installation/" title="Installation" class="md-nav__link">
      Installation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../inputdata/" title="Input Data for SNaQ" class="md-nav__link">
      Input Data for SNaQ
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        TICR pipeline
      </label>
    
    <a href="./" title="TICR pipeline" class="md-nav__link md-nav__link--active">
      TICR pipeline
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#to-run-mrbayes-you-already-have-alignments" title="To run MrBayes: You already have alignments" class="md-nav__link">
    To run MrBayes: You already have alignments
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#to-run-bucky-you-already-have-mrbayes-output" title="To run BUCKy: You already have MrBayes output" class="md-nav__link">
    To run BUCKy: You already have MrBayes output
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#to-run-mbsum-on-the-output-of-mrbayes-for-each-gene" title="To run mbsum on the output of MrBayes for each gene" class="md-nav__link">
    To run mbsum on the output of MrBayes for each gene
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#to-run-bucky-on-all-4-taxon-sets-you-already-have-the-mbsum-output" title="To run bucky on all 4-taxon sets: you already have the mbsum output" class="md-nav__link">
    To run bucky on all 4-taxon sets: you already have the mbsum output
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../snaq_plot/" title="Network estimation and display" class="md-nav__link">
      Network estimation and display
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../dist_reroot/" title="Network comparison and manipulation" class="md-nav__link">
      Network comparison and manipulation
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../fixednetworkoptim/" title="Candidate Networks" class="md-nav__link">
      Candidate Networks
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../expectedCFs/" title="Extract Expected CFs" class="md-nav__link">
      Extract Expected CFs
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../bootstrap/" title="Bootstrap" class="md-nav__link">
      Bootstrap
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../multiplealleles/" title="Multiple Alleles" class="md-nav__link">
      Multiple Alleles
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../trait_tree/" title="Continuous Trait Evolution" class="md-nav__link">
      Continuous Trait Evolution
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3">
    
    <label class="md-nav__link" for="nav-3">
      Library
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        Library
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../lib/public/" title="Public" class="md-nav__link">
      Public
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../lib/internals/" title="Internals" class="md-nav__link">
      Internals
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#to-run-mrbayes-you-already-have-alignments" title="To run MrBayes: You already have alignments" class="md-nav__link">
    To run MrBayes: You already have alignments
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#to-run-bucky-you-already-have-mrbayes-output" title="To run BUCKy: You already have MrBayes output" class="md-nav__link">
    To run BUCKy: You already have MrBayes output
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#to-run-mbsum-on-the-output-of-mrbayes-for-each-gene" title="To run mbsum on the output of MrBayes for each gene" class="md-nav__link">
    To run mbsum on the output of MrBayes for each gene
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#to-run-bucky-on-all-4-taxon-sets-you-already-have-the-mbsum-output" title="To run bucky on all 4-taxon sets: you already have the mbsum output" class="md-nav__link">
    To run bucky on all 4-taxon sets: you already have the mbsum output
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/crsl4/PhyloNetworks.jl/edit/master/docs/man/ticr_howtogetQuartetCFs.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <p><a id='TICR-pipeline-1'></a></p>
<h1 id="ticr-pipeline">TICR pipeline</h1>
<p>PhyloNetworks' <a href="https://github.com/crsl4/PhyloNetworks.jl/wiki/TICR:-from-alignments-to-quartet-concordance-factors">wiki</a> has a step-by-step tutorial, to go from multiple sequence alignments to a table of quartet gene frequencies (concordance factors: CFs), through BUCKy (to integrate out gene tree uncertainty) or through RAxML. To get the <code>raxml.pl</code> perl script to run RAxML on each gene, download the content of that wiki with</p>
<p><code>git clone https://github.com/crsl4/PhyloNetworks.jl.wiki.git</code></p>
<p>then go to the <code>script/</code> folder.   Below is more detail to insert data into the TICR pipeline at various stages. Full information and code for this pipeline is <a href="https://github.com/nstenz/TICR">here</a>.</p>
<p><a id='To-run-MrBayes:-You-already-have-alignments-1'></a></p>
<h2 id="to-run-mrbayes-you-already-have-alignments">To run MrBayes: You already have alignments</h2>
<p>If you don't need to run mdl.pl because you already have aligned gene sequences which you would like to run through MrBayes, you can simply create a tarball of the Nexus files (fasta won't work at this stage) you wish to use. This command assumes that you want to use all the files ending with ".nex" in the current directory, one file per locus:</p>
<p><code>tar czf my-genes.tar.gz *.nex</code></p>
<p>Once the tarball has been successfully generated, you can then specify this file as input for <a href="https://github.com/nstenz/TICR/blob/master/scripts/mb.pl">mb.pl</a> assuming you have a a valid MrBayes block located in the file "bayes.txt":</p>
<p><code>mb.pl my-genes.tar.gz -m bayes.txt -o my-genes-mb</code></p>
<p>If you get an error message like <code>mb.pl: Command not found</code>, it might be because <code>mb.pl</code> has no execute permission or your current directory is not in your path. An easy fix is to run this command instead:</p>
<p><code>perl mb.pl my-genes.tar.gz -m bayes.txt -o my-genes-mb</code></p>
<p>The resulting output tarball would now be located in my-genes-mb/my-genes.mb.tar, and can be used normally with <a href="https://github.com/nstenz/TICR/blob/master/scripts/bucky.pl">bucky.pl</a>, that is, like this:</p>
<p><code>bucky.pl my-genes-mb/my-genes.mb.tar -o mygenes-bucky</code></p>
<p>The output, with the table of concordance factors for all sets of 4 taxa, will be in a file named <code>my-genes.CFs.csv</code> inside a directory named <code>mygenes-bucky</code>. That's the file containing the quartet concordance factors to give to SNaQ as input.</p>
<p><a id='To-run-BUCKy:-You-already-have-MrBayes-output-1'></a></p>
<h2 id="to-run-bucky-you-already-have-mrbayes-output">To run BUCKy: You already have MrBayes output</h2>
<p><a id='To-run-mbsum-on-the-output-of-MrBayes-for-each-gene-1'></a></p>
<h3 id="to-run-mbsum-on-the-output-of-mrbayes-for-each-gene">To run mbsum on the output of MrBayes for each gene</h3>
<p>You must now run <code>mbsum</code> separately on each gene's MrBayes output. For a gene with output tree files named gene1.run1.t gene1.run2.t gene1.run3.t, and a desired burnin of 1000 trees per tree file:</p>
<p><code>mbsum -n 1000 -o gene1.in gene1.run1.t gene1.run2.t gene1.run3.t</code></p>
<p>This <code>mbsum</code> command will need to be executed for each gene. Now continue to the next section.</p>
<p><a id='To-run-bucky-on-all-4-taxon-sets:-you-already-have-the-mbsum-output-1'></a></p>
<h3 id="to-run-bucky-on-all-4-taxon-sets-you-already-have-the-mbsum-output">To run bucky on all 4-taxon sets: you already have the mbsum output</h3>
<p>If you have already run mbsum on each individual gene's MrBayes output, you can simply create a tarball containing all the mbsum output files. So if you had mbsum output in files named gene1.in, gene2.in, ... , gene100.in, you would want to run something similar to the following command:</p>
<p><code>tar czf my-genes-mbsum.tar.gz gene*.in</code></p>
<p>You can now use this tarball along with the -s option in <a href="https://github.com/nstenz/TICR/blob/master/scripts/bucky.pl">bucky.pl</a> like this:</p>
<p><code>bucky.pl my-genes-mbsum.tar.gz -s -o mygenes-bucky</code></p>
<p>Again, if you get an error like <code>bucky.pl: Command not found</code>, run instead</p>
<p><code>bucky.pl my-genes-mbsum.tar.gz -s -o mygenes-bucky</code></p>
<p>The output, with the table of concordance factors for all sets of 4 taxa, will be in a file named <code>my-genes.CFs.csv</code> inside directory <code>mygenes-bucky</code>. That's the file containing the quartet concordance factors to give to SNaQ as input.</p>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../inputdata/" title="Input Data for SNaQ" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Input Data for SNaQ
              </span>
            </div>
          </a>
        
        
          <a href="../snaq_plot/" title="Network estimation and display" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Network estimation and display
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/application.a59e2a89.js"></script>
      
      <script>app.initialize({version:"0.17.4",url:{base:"../.."}})</script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../../assets/mathjaxhelper.js"></script>
      
    
    
      
    
  </body>
</html>