<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Network comparison and manipulation · PhyloNetworks.jl</title><meta name="title" content="Network comparison and manipulation · PhyloNetworks.jl"/><meta property="og:title" content="Network comparison and manipulation · PhyloNetworks.jl"/><meta property="twitter:title" content="Network comparison and manipulation · PhyloNetworks.jl"/><meta name="description" content="Documentation for PhyloNetworks.jl."/><meta property="og:description" content="Documentation for PhyloNetworks.jl."/><meta property="twitter:description" content="Documentation for PhyloNetworks.jl."/><meta property="og:url" content="https://juliaphylo.github.io/PhyloNetworks.jl/stable/man/dist_reroot/"/><meta property="twitter:url" content="https://juliaphylo.github.io/PhyloNetworks.jl/stable/man/dist_reroot/"/><link rel="canonical" href="https://juliaphylo.github.io/PhyloNetworks.jl/stable/man/dist_reroot/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="PhyloNetworks.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="PhyloNetworks.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">PhyloNetworks.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Manual</span><ul><li><a class="tocitem" href="../installation/">Installation</a></li><li><a class="tocitem" href="../introduction/">Introduction</a></li><li><a class="tocitem" href="../netmanipulation/">Network manipulation</a></li><li><a class="tocitem" href="../net_plot/">Network visualization</a></li><li class="is-active"><a class="tocitem" href>Network comparison and manipulation</a><ul class="internal"><li><a class="tocitem" href="#Robinson-Foulds-distance-for-trees"><span>Robinson-Foulds distance for trees</span></a></li><li><a class="tocitem" href="#Re-rooting-trees-and-networks"><span>Re-rooting trees and networks</span></a></li><li><a class="tocitem" href="#Does-the-root-conflict-with-the-direction-of-a-reticulation?"><span>Does the root conflict with the direction of a reticulation?</span></a></li><li><a class="tocitem" href="#Extracting-the-major-tree"><span>Extracting the major tree</span></a></li><li><a class="tocitem" href="#Hardwired-cluster-distance"><span>Hardwired-cluster distance</span></a></li><li><a class="tocitem" href="#μ-distances"><span>μ distances</span></a></li><li><a class="tocitem" href="#Displayed-trees-and-subnetworks"><span>Displayed trees and subnetworks</span></a></li></ul></li><li><a class="tocitem" href="../network_support/">Network support</a></li><li><a class="tocitem" href="../parsimony/">Parsimony on networks</a></li><li><a class="tocitem" href="../nj/">Neighbour Joining</a></li><li><a class="tocitem" href="../expecteddata/">Pairwise and quartet data</a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../../lib/public/">Public</a></li><li><a class="tocitem" href="../../lib/internals/">Internals</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Manual</a></li><li class="is-active"><a href>Network comparison and manipulation</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Network comparison and manipulation</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/PhyloNetworks.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/JuliaPhylo/PhyloNetworks.jl/blob/master/docs/src/man/dist_reroot.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Comparing-and-manipulating-networks"><a class="docs-heading-anchor" href="#Comparing-and-manipulating-networks">Comparing and manipulating networks</a><a id="Comparing-and-manipulating-networks-1"></a><a class="docs-heading-anchor-permalink" href="#Comparing-and-manipulating-networks" title="Permalink"></a></h1><p>Examples below use networks available from the package. We can load them as follows. <code>astraltree</code> is a tree estimated using ASTRAL, <code>net0</code> is a network estimated using SNaQ under the constraint of 0 reticulation, so it&#39;s also a tree, and <code>net1</code> is a network estimated under the constraint of 1 reticulation, so not a tree.</p><p>All of them are to be interpreted as <em>semidirected</em> networks, because their root is not identifiable by the method used to estimate them. This affects how we want them.</p><pre><code class="language-julia hljs">astralfile = joinpath(dirname(pathof(PhyloNetworks)), &quot;..&quot;,&quot;examples&quot;,&quot;astral.tre&quot;)
astraltree = readmultinewick(astralfile)[102] # 102th tree = last tree here
net0 = readnewick(joinpath(dirname(pathof(PhyloNetworks)), &quot;..&quot;,&quot;examples&quot;,&quot;net0.out&quot;))
net1 = readnewick(joinpath(dirname(pathof(PhyloNetworks)), &quot;..&quot;,&quot;examples&quot;,&quot;net1.out&quot;))</code></pre><h2 id="Robinson-Foulds-distance-for-trees"><a class="docs-heading-anchor" href="#Robinson-Foulds-distance-for-trees">Robinson-Foulds distance for trees</a><a id="Robinson-Foulds-distance-for-trees-1"></a><a class="docs-heading-anchor-permalink" href="#Robinson-Foulds-distance-for-trees" title="Permalink"></a></h2><p>Do the 2 trees <code>astraltree</code> and <code>net0</code> have the same topology, unrooted? We can calculate the Robinson-Foulds distance between them, using a function that extends the Robinson-Foulds distance to general networks (more on this below).</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; hardwiredclusterdistance(astraltree, net0, false)</code><code class="nohighlight hljs ansi" style="display:block;">0</code></pre><p>The last option <code>false</code> is to consider topologies as unrooted. The RF distance is 0, so the two unrooted topologies are the same. If we had considered them as rooted, with whatever root they currently have in their internal representation, we would find a difference:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; hardwiredclusterdistance(astraltree, net0, true)</code><code class="nohighlight hljs ansi" style="display:block;">5</code></pre><h2 id="Re-rooting-trees-and-networks"><a class="docs-heading-anchor" href="#Re-rooting-trees-and-networks">Re-rooting trees and networks</a><a id="Re-rooting-trees-and-networks-1"></a><a class="docs-heading-anchor-permalink" href="#Re-rooting-trees-and-networks" title="Permalink"></a></h2><p>We can re-root our networks with the outgroup, O, and then re-compare the ASTRAL tree and the SNaQ tree as rooted topologies (and find no difference):</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; rootatnode!(astraltree, &quot;O&quot;)</code><code class="nohighlight hljs ansi" style="display:block;">HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: E, O, C, D, ...
(O,(((C,D)100.0:3.029,(B,A)100.0:1.324)100.0:0.464,E));</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; rootatnode!(net0, &quot;O&quot;)</code><code class="nohighlight hljs ansi" style="display:block;">HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, B, A, ...
(O,(E,((B,A):1.396,(C,D):10.0):0.485));</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; hardwiredclusterdistance(astraltree, net0, true)</code><code class="nohighlight hljs ansi" style="display:block;">0</code></pre><pre><code class="language-julia hljs">using PhyloPlots, RCall
R&quot;name &lt;- function(x) file.path(&#39;..&#39;, &#39;assets&#39;, &#39;figures&#39;, x)&quot;
R&quot;svg(name(&#39;net0_O.svg&#39;), width=4, height=4)&quot;
R&quot;par&quot;(mar=[0,0,0,0])
plot(net0);
R&quot;dev.off()&quot;</code></pre><p><img src="../../assets/figures/net0_O.svg" alt="net0_O"/></p><p>Note that we use the possibilities of <a href="https://juliainterop.github.io/RCall.jl/stable/gettingstarted/"><code>RCall</code></a> to save the plot. We only show these commands once, but they will be run behind the scene each time a plot is called.</p><p>After trees/networks are rooted with a correct outgroup, their visualization is more meaningful.</p><p>Networks can be re-rooted at a given node or along a given edge. Get help (type <code>?</code>) on the functions <code>rootatnode!</code> and <code>rootonedge!</code> for more info. There are examples in the <a href="../network_support/#Network-support">Network support</a> section.</p><p>If the network is plotted with crossing edges, you may identify ways to rotate the children edges at some nodes to untangle some crossing edges. This can be done using the function <code>rotate!</code>. See an example in the <a href="../network_support/#Network-support">Network support</a> section, or type <code>?</code> then <code>rotate!</code>.</p><h2 id="Does-the-root-conflict-with-the-direction-of-a-reticulation?"><a class="docs-heading-anchor" href="#Does-the-root-conflict-with-the-direction-of-a-reticulation?">Does the root conflict with the direction of a reticulation?</a><a id="Does-the-root-conflict-with-the-direction-of-a-reticulation?-1"></a><a class="docs-heading-anchor-permalink" href="#Does-the-root-conflict-with-the-direction-of-a-reticulation?" title="Permalink"></a></h2><p>With 1 hybridization or more, the direction of hybrid edges constrain the position of the root. The root cannot be downstream of hybrid edges. Any hybrid node has to be younger than, or of the same age as both of its parents. So time has to flow &quot;downwards&quot; of any hybrid node, and the root cannot be placed &quot;below&quot; a hybrid node. An attempt to re-root the network at a position incompatible with hybrid edges will fail, with a <code>RootMismatch</code> error. To show an example, let&#39;s use the network below. We plotted the edge numbers, because we will want to use them later to place the root.</p><pre><code class="language-julia hljs">net7taxa = readnewick(&quot;(C,D,((O,(E,#H7:::0.196):0.314):0.664,(((A1,A2))#H7:::0.804,B):10.0):10.0);&quot;)
plot(net7taxa, showgamma=true, showedgenumber=true, tipoffset=0.2);</code></pre><p><img src="../../assets/figures/reroot_net7taxa_1.svg" alt="reroot net7taxa 1"/></p><p>Let&#39;s imagine that A1 and A2 are our outgroups, and we estimated the network above. According to this network, time must flow from the hybrid node towards A1 and A2. So any attempt to reroot the network with A1 as outgroup, or with A2 as outgroup, or with the A clade (on edge 11), will fail with a <code>RootMismatch</code> error:</p><pre><code class="language-julia hljs">rootatnode!(net7taxa, &quot;A1&quot;); # ERROR: RootMismatch: non-leaf node 5 had 0 children. ...
rootatnode!(net7taxa, &quot;A2&quot;); # ERROR: RootMismatch (again)
rootonedge!(net7taxa, 10);   # ERROR: RootMismatch (again)</code></pre><p>In this case, however, it is possible to root the network on either parent edge of the hybrid node. These edges have numbers 12 and 5, based on the plot above. We get these 2 rooted versions of the network:</p><pre><code class="language-julia hljs">R&quot;layout(matrix(1:2,1,2))&quot;;
rootonedge!(net7taxa, 11);
rotate!(net7taxa, -5)
plot(net7taxa, showgamma=true, tipoffset=0.2, shownodenumber=true);
R&quot;mtext&quot;(&quot;rooted on hybrid edge 11 (major)&quot;, line=-1)
rootonedge!(net7taxa, 5);
plot(net7taxa, showgamma=true, tipoffset=0.2);
R&quot;mtext&quot;(&quot;rooted on hybrid edge 5 (minor)&quot;, line=-1);</code></pre><p><img src="../../assets/figures/reroot_net7taxa_2.svg" alt="reroot net7taxa 2"/></p><p>On the second plot, the A clade does not <em>appear</em> to be an outgroup, but this is just because the plot follows the major tree primarily, based the major hybrid edges (those with γ&gt;0.5). We can display the exact same network differently, by changing the γ inheritance values to invert the major/minor consideration of the hybrid edges.</p><pre><code class="language-julia hljs">net7taxa.edge[5] # just to check that it&#39;s one of the 2 hybrid edges of interest
setgamma!(net7taxa.edge[5], 0.501) # switch major/minor edges
plot(net7taxa, tipoffset=0.2); # not showing gamma values, because we changed them artificially
R&quot;mtext&quot;(&quot;rooted on hybrid edge 5 (considered major)&quot;, line=-1);</code></pre><p><img src="../../assets/figures/reroot_net7taxa_3.svg" alt="reroot net7taxa 3"/></p><p>Conclusion, in this particular example: it is possible to re-root the network to a place where the A clade is indeed an outgroup. But it did require some care, and we discovered that there are 2 acceptable rooting options. The first is more plausible, if we think that the <em>species tree</em> is the <em>major tree</em>, meaning that any gene flow or introgression event replaced less than 50% of the genes in the recipient population.</p><p>In other cases, it may not be possible to re-root the network with a known outgroup. It would be the case if A1 was the only outgroup, and if A2 was an ingroup taxon. In such a case, the outgroup knowledge tells us that our estimated network is wrong. One (or more) reticulation in the network must be incorrect. Its placement might be correct, but then its direction would be incorrect.</p><h2 id="Extracting-the-major-tree"><a class="docs-heading-anchor" href="#Extracting-the-major-tree">Extracting the major tree</a><a id="Extracting-the-major-tree-1"></a><a class="docs-heading-anchor-permalink" href="#Extracting-the-major-tree" title="Permalink"></a></h2><p>We can also compare the networks estimated with h=0 (<code>net0</code>) and h=1 (<code>net1</code>):</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; rootatnode!(net1, &quot;O&quot;); # the ; suppresses screen output</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; hardwiredclusterdistance(net0, net1, true)</code><code class="nohighlight hljs ansi" style="display:block;">2</code></pre><pre><code class="language-julia hljs">plot(net1, showgamma=true);</code></pre><p><img src="../../assets/figures/net1_O.svg" alt="net1_O"/></p><p>They differ by 2 clusters: that&#39;s because A is of hybrid descent in <code>net1</code> (descendant of each hybrid edge), not in <code>net0</code>.</p><p>To beyond this hybrid difference, we can extract the major tree from the network with 1 hybridization, that is, delete the hybrid edge supported by less than 50% of genes. Then we can compare this tree with the ASTRAL/SNaQ tree <code>net0</code>.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; tree1 = majortree(net1); # major tree from net1</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; hardwiredclusterdistance(net0, tree1, true)</code><code class="nohighlight hljs ansi" style="display:block;">0</code></pre><p>They are identical (at distance 0), so here the species network with 1 hybrid node is a refinement of the estimated species tree (this needs not be the case always).</p><h2 id="Hardwired-cluster-distance"><a class="docs-heading-anchor" href="#Hardwired-cluster-distance">Hardwired-cluster distance</a><a id="Hardwired-cluster-distance-1"></a><a class="docs-heading-anchor-permalink" href="#Hardwired-cluster-distance" title="Permalink"></a></h2><p>Is <code>net1</code>, the SNaQ network with 1 hybrid node, the same as the true network, the network that was initially used to simulate the data?</p><p>(Digression on this example&#39;s data: gene trees were simulated under the coalescent along some &quot;true&quot; network, then 500 base-pair alignments were simulated along each gene tree with the HKY model, gene trees were estimated from each alignment with RAxML, and these estimated gene trees served as input to both ASTRAL and SNaQ.)</p><p>The &quot;true&quot; network is shown below, correctly rooted at the outgroup O, and plotted with branch lengths proportional to their values in coalescence units:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; truenet = readnewick(&quot;((((D:0.4,C:0.4):4.8,((A:0.8,B:0.8):2.2)#H1:2.2::0.7):4.0,
       (#H1:0::0.3,E:3.0):6.2):2.0,O:11.2);&quot;);</code></pre><pre><code class="language-julia hljs">plot(truenet, useedgelength=true, showgamma=true);</code></pre><p><img src="../../assets/figures/truenet_sim.svg" alt="truenet"/></p><p>We can compare two networks using the hardwired-cluster dissimilarity.</p><ul><li>This dissimilarity counts the number of hardwired clusters present in one network but absent in the other.</li><li>The <strong>hardwired cluster</strong> associated with an edge is the set of <em>all</em> tips descendant from that edge, i.e. all tips that inherited at least <em>some</em> genetic material from that edge.</li><li>Each edge is associated with its hardwired cluster of descendants, and also with a &quot;tree&quot; tag or &quot;hybrid&quot; tag depending on the edge type.</li></ul><div class="admonition is-info" id="distance-versus-dissimilarity-d91864e9fa408254"><header class="admonition-header">distance versus dissimilarity<a class="admonition-anchor" href="#distance-versus-dissimilarity-d91864e9fa408254" title="Permalink"></a></header><div class="admonition-body"><p>When comparing level-1 networks, or tree-child networks more generally, the hardwired-cluster dissimilarity is a distance: d(N, N&#39;) = 0 exactly when N and N&#39; have the same topology. (<a href="https://doi.org/10.1109/TCBB.2024.3361390">Cardona et al. 2009</a>, <a href="https://doi.org/10.1016/j.mbs.2021.108537">Bai et al.</a> <a href="https://doi.org/10.1109/TCBBIO.2025.3534780">Maxfield, Xu &amp; Ané 2025</a>). Unfortunately, this is not generally true for complex network: there are networks at hardwired-cluster &#39;distance&#39; 0, that have different topologies. But no dissimilarity measure can both be fast to calculate and be a distance on the full space of phylogenetic networks (<a href="https://doi.org/10.1155/2014/254279">Cardona et al. 2014</a>).</p></div></div><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; hardwiredclusterdistance(net1, truenet, true) # true: nets considered as rooted</code><code class="nohighlight hljs ansi" style="display:block;">4</code></pre><p>Our estimated network is at distance 4 (not 0), so it is <em>different</em> from the true network (there was estimation error). The 4 cluster differences correspond to the hybrid edges in <code>net1</code> (whose cluster is A), the hybrid edges in <code>truenet</code> (whose cluster is AB); and the parent of their minor hybrid edges: with cluster EA in <code>net1</code> versus EAB in <code>truenet</code>.</p><p>From the plots, we see that:</p><ul><li>the underlying tree is correctly estimated</li><li>the origin of gene flow is correctly estimated: E</li><li>the target of gene flow is <em>not</em> correctly estimated: it was the lineage ancestral to (A,B), but it is estimated to be A only.</li></ul><h2 id="μ-distances"><a class="docs-heading-anchor" href="#μ-distances">μ distances</a><a id="μ-distances-1"></a><a class="docs-heading-anchor-permalink" href="#μ-distances" title="Permalink"></a></h2><p>The node-based μ-representation of rooted networks and the edge-based μ-representation of semidirected networks lead to μ-distances, which generalize the RF-distance on trees. See <a href="https://doi.org/10.1109/TCBB.2024.3361390">Cardona et al. (2024)</a> and <a href="https://doi.org/10.1109/TCBBIO.2025.3534780">Maxfield, Xu &amp; Ané (2025)</a></p><p>Each μ-representation lists a number of counts: counting the number of paths from each node / edge to each leaf (and to some/any hybrid node). From these path numbers, we can get the hardwired cluster associated to node or leaf: the tips that can be reached via 1 or more paths.   So then the μ-distance is more discriminant (larger) than the hardwired-cluster distance.</p><p>For semidirected networks, the calculation of the μ-distance is much faster than the calculation of the hardwired-cluster distance, because it requires a single traversal of the network.</p><div class="admonition is-info" id="distance-versus-dissimilarity-92e30a6a4beed050"><header class="admonition-header">distance versus dissimilarity<a class="admonition-anchor" href="#distance-versus-dissimilarity-92e30a6a4beed050" title="Permalink"></a></header><div class="admonition-body"><p>Again, the μ-distances are only dissimilarities generally. They are known to provided a distance, with d(N,N&#39;) = 0 exactly when N and N&#39; share the same topology, when N and N&#39; are tree-child rooted or semidirected networks.</p></div></div><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; mudistance_rooted(net1, truenet)</code><code class="nohighlight hljs ansi" style="display:block;">8</code></pre><p>Here both networks have an edge leading to the cluster ABCDE (the stem of the clade), but these edges have different number of paths to B: only 1 path in <code>net1</code> versus 2 paths in <code>truenet</code>. So these edges did not count towards the hardwired-cluster distance, but do count towards the rooted μ-distance now.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; mudistance_semidirected(net1, truenet)</code><code class="nohighlight hljs ansi" style="display:block;">12</code></pre><p>When considered as undirected, the μ-distance is greater here: this is because each edge is considered in each direction it can take if its direction may change depending on the root. To math edges between 2 networks, the edges&#39; number of paths to taxa need to match – in both directions (unlike when the networks are considered rooted). Here, each network has 6 internal edges (after suppressing the root) and none of them match between the 2 networks.</p><p>But the distance does not depend on the root position (which may be unknown):</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; net1_E = deepcopy(net1); rootatnode!(net1_E, &quot;E&quot;);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; mudistance_semidirected(net1_E, truenet) # same as with net1</code><code class="nohighlight hljs ansi" style="display:block;">12</code></pre><p>and if we ignore B the μ-distance drops to 0, consistent with the 2 subnetworks being equal:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; net1_noB    = deepcopy(net1);    deleteleaf!(net1_noB, &quot;B&quot;);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; truenet_noB = deepcopy(truenet); deleteleaf!(truenet_noB, &quot;B&quot;);</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; mudistance_semidirected(net1_noB, truenet_noB)</code><code class="nohighlight hljs ansi" style="display:block;">0</code></pre><p>To get the μ-representations themselves, see <a href="../../lib/internals/#PhyloNetworks.node_murepresentation"><code>PhyloNetworks.node_murepresentation</code></a> and <a href="../../lib/internals/#PhyloNetworks.edge_murepresentation"><code>PhyloNetworks.edge_murepresentation</code></a>.</p><h2 id="Displayed-trees-and-subnetworks"><a class="docs-heading-anchor" href="#Displayed-trees-and-subnetworks">Displayed trees and subnetworks</a><a id="Displayed-trees-and-subnetworks-1"></a><a class="docs-heading-anchor-permalink" href="#Displayed-trees-and-subnetworks" title="Permalink"></a></h2><p>We can extract all trees displayed in a network. These trees are obtained by picking one parent hybrid edge at each hybrid node, and dropping the other parent hybrid edge. We can choose to pick the &quot;important&quot; hybrid edges only, with heritability γ at or above a threshold. Below we use a γ threshold of 0, so we get all displayed trees:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; t = displayedtrees(net1, 0.0) # list of trees displayed in network</code><code class="nohighlight hljs ansi" style="display:block;">2-element Vector{HybridNetwork}:
 HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,(((B,A):10.0,(C,D):10.0):0.664,E));

 HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,((E,A):0.314,((C,D):10.0,B):0.664));</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; writenewick(t[1], round=true)</code><code class="nohighlight hljs ansi" style="display:block;">&quot;(O,(((B,A):10.0,(C,D):10.0):0.664,E));&quot;</code><br/><code class="language-julia-repl hljs" style="display:block;">julia&gt; writenewick(t[2], round=true)</code><code class="nohighlight hljs ansi" style="display:block;">&quot;(O,((E,A):0.314,((C,D):10.0,B):0.664));&quot;</code></pre><p>If we decide to keep edges with γ&gt;0.2 only, then we are left with a single tree in the list (the major tree). This is because our example has 1 hybrid node with minor γ=0.196.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; t = displayedtrees(net1, 0.2)</code><code class="nohighlight hljs ansi" style="display:block;">1-element Vector{HybridNetwork}:
 HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,(((B,A):10.0,(C,D):10.0):0.664,E));</code></pre><p>We can also delete all &quot;non-important&quot; reticulations, those with a minor heritability γ below some threshold. The function below changes our network <code>net1</code>, as indicated by its name ending with a <code>!</code>.</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; deletehybridthreshold!(net1, 0.1)</code><code class="nohighlight hljs ansi" style="display:block;">HybridNetwork, Rooted Network
13 edges
13 nodes: 6 tips, 1 hybrid nodes, 6 internal tree nodes.
tip labels: C, D, O, E, ...
(O,((E,#H7:::0.196):0.314,((B,(A)#H7:::0.804):10.0,(C,D):10.0):0.664));</code></pre><p>Nothing happened to our network: because its γ is above 0.1. But if we set the threshold to 0.3, then our reticulation disappears:</p><pre><code class="language-julia-repl hljs" style="display:block;">julia&gt; deletehybridthreshold!(net1, 0.3)</code><code class="nohighlight hljs ansi" style="display:block;">HybridNetwork, Rooted Network
10 edges
11 nodes: 6 tips, 0 hybrid nodes, 5 internal tree nodes.
tip labels: C, D, O, E, ...
(O,(((B,A):10.0,(C,D):10.0):0.664,E));</code></pre><p>See also function <code>displayednetworkat!</code> to get the network with a single reticulation of interest, and eliminate all other reticulations.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../net_plot/">« Network visualization</a><a class="docs-footer-nextpage" href="../network_support/">Network support »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.16.1 on <span class="colophon-date" title="Thursday 4 December 2025 16:25">Thursday 4 December 2025</span>. Using Julia version 1.12.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
