<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>fitting DNA on a network · PhyloNetworks.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../"><img class="logo" src="../../assets/logo.png" alt="PhyloNetworks.jl logo"/></a><h1>PhyloNetworks.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">Manual</span><ul><li><a class="toctext" href="../installation/">Installation</a></li><li><a class="toctext" href="../inputdata/">Input Data for SNaQ</a></li><li><a class="toctext" href="../ticr_howtogetQuartetCFs/">TICR pipeline</a></li><li><a class="toctext" href="../snaq_plot/">Network estimation and display</a></li><li><a class="toctext" href="../dist_reroot/">Network comparison and manipulation</a></li><li><a class="toctext" href="../fixednetworkoptim/">Candidate Networks</a></li><li><a class="toctext" href="../expectedCFs/">Extract Expected CFs</a></li><li><a class="toctext" href="../bootstrap/">Bootstrap</a></li><li><a class="toctext" href="../multiplealleles/">Multiple Alleles</a></li><li><a class="toctext" href="../trait_tree/">Continuous Trait Evolution</a></li><li><a class="toctext" href="../parsimony/">Parsimony on networks</a></li><li><a class="toctext" href="../fitDiscrete/">Discrete Trait Evolution</a></li></ul></li><li><span class="toctext">Library</span><ul><li><a class="toctext" href="../../lib/public/">Public</a></li><li><a class="toctext" href="../../lib/internals/">Internals</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li><a href>fitting DNA on a network</a></li></ul><a class="edit-page" href="https://github.com/crsl4/PhyloNetworks.jl/blob/master/docs/src/man/fitdiscreteDNA.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>fitting DNA on a network</span><a class="fa fa-bars" href="#"></a></div></header><div></div><h1><a class="nav-anchor" id="fitting-DNA-on-a-network-1" href="#fitting-DNA-on-a-network-1">fitting DNA on a network</a></h1><p>The methods below model each DNA site as a trait, assuming that sites are unlinked, that is, they evolve independently of each other. In other words, this is a &quot;concatenation&quot; approach where sites from the same locus do not share information about their evolutionary path. This is appropriate if recombination is assumed to have occurred within genes.</p><h2><a class="nav-anchor" id="DNA-evolution:-data-and-models-1" href="#DNA-evolution:-data-and-models-1">DNA evolution: data and models</a></h2><h3><a class="nav-anchor" id="reading-in-an-alignment-1" href="#reading-in-an-alignment-1">reading in an alignment</a></h3><p>As for trait evolution, <a href="../../lib/public/#PhyloNetworks.fitdiscrete"><code>fitdiscrete</code></a> can be used. It can be given data in a variety of ways. For DNA, this is one way:</p><pre><code class="language-julia-repl">julia&gt; # read in network
       dna_net = readTopology(&quot;((((((((((((((Ae_caudata_Tr275:1.0,Ae_caudata_Tr276:1.0):1.0,Ae_caudata_Tr139:1.0):1.0)#H1:1.0::0.6,((((((Ae_longissima_Tr241:1.0,Ae_longissima_Tr242:1.0):1.0,Ae_longissima_Tr355:1.0):1.0,(Ae_sharonensis_Tr265:1.0,Ae_sharonensis_Tr264:1.0):1.0):1.0,((Ae_bicornis_Tr408:1.0,Ae_bicornis_Tr407:1.0):1.0,Ae_bicornis_Tr406:1.0):1.0):1.0,((Ae_searsii_Tr164:1.0,Ae_searsii_Tr165:1.0):1.0,Ae_searsii_Tr161:1.0):1.0):1.0)#H2:1.0::0.6):1.0,(((Ae_umbellulata_Tr266:1.0,Ae_umbellulata_Tr257:1.0):1.0,Ae_umbellulata_Tr268:1.0):1.0,#H1:1.0::0.4):1.0):1.0,((Ae_comosa_Tr271:1.0,Ae_comosa_Tr272:1.0):1.0,(((Ae_uniaristata_Tr403:1.0,Ae_uniaristata_Tr357:1.0):1.0,Ae_uniaristata_Tr402:1.0):1.0,Ae_uniaristata_Tr404:1.0):1.0):1.0):1.0,(((Ae_tauschii_Tr352:1.0,Ae_tauschii_Tr351:1.0):1.0,(Ae_tauschii_Tr180:1.0,Ae_tauschii_Tr125:1.0):1.0):1.0,(#H2:1.0::0.4,((((Ae_mutica_Tr237:1.0,Ae_mutica_Tr329:1.0):1.0,Ae_mutica_Tr244:1.0):1.0,Ae_mutica_Tr332:1.0):1.0)#H4:1.0::0.6):1.0):1.0):1.0,(((T_boeoticum_TS8:1.0,(T_boeoticum_TS10:1.0,T_boeoticum_TS3:1.0):1.0):1.0,T_boeoticum_TS4:1.0):1.0,((T_urartu_Tr315:1.0,T_urartu_Tr232:1.0):1.0,(T_urartu_Tr317:1.0,T_urartu_Tr309:1.0):1.0):1.0):1.0):1.0,(((((Ae_speltoides_Tr320:1.0,Ae_speltoides_Tr323:1.0):1.0,Ae_speltoides_Tr223:1.0):1.0,Ae_speltoides_Tr251:1.0):1.0):1.0,#H4:1.0::0.4):1.0):1.0):1.0,Ta_caputMedusae_TB2:1.0):1.0,S_vavilovii_Tr279:1.0):1.0,Er_bonaepartis_TB1:1.0):1.0,H_vulgare_HVens23:1.0);&quot;);

julia&gt; # read in alignment in FASTA format
       fastafile = joinpath(dirname(pathof(PhyloNetworks)), &quot;..&quot;,&quot;examples&quot;,&quot;Ae_bicornis_Tr406_Contig10132.aln&quot;);

julia&gt; dna_dat, dna_weights = readfastatodna(fastafile, true);

julia&gt; dna_dat
22×210 DataFrames.DataFrame. Omitted printing of 205 columns
│ Row │ taxon                │ x1        │ x2        │ x3        │ x4        │
│     │ String               │ BioSymbo… │ BioSymbo… │ BioSymbo… │ BioSymbo… │
├─────┼──────────────────────┼───────────┼───────────┼───────────┼───────────┤
│ 1   │ Ae_uniaristata_Tr357 │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 2   │ Ae_mutica_Tr237      │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 3   │ Ae_tauschii_Tr352    │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 4   │ Ae_searsii_Tr165     │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
│ 5   │ Ae_sharonensis_Tr265 │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 6   │ Ae_bicornis_Tr407    │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
│ 7   │ Ae_longissima_Tr241  │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
⋮
│ 15  │ Ae_longissima_Tr242  │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 16  │ Ae_comosa_Tr271      │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 17  │ Ae_searsii_Tr164     │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
│ 18  │ Ae_speltoides_Tr251  │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 19  │ Ae_bicornis_Tr408    │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │
│ 20  │ Ae_tauschii_Tr351    │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
│ 21  │ Ae_longissima_Tr355  │ DNA_T     │ DNA_C     │ DNA_A     │ DNA_G     │
│ 22  │ Ae_caudata_Tr139     │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │ DNA_Gap   │

julia&gt; dna_weights
209-element Array{Float64,1}:
 23.0
 18.0
 13.0
 16.0
  1.0
  1.0
  1.0
  1.0
  1.0
  9.0
  ⋮
  1.0
  1.0
  1.0
  1.0
  1.0
  1.0
  1.0
  2.0
  3.0</code></pre><p>Here, <code>dna_dat</code> is a single data frame containing both species names and trait data (site patterns). The alignment was summarized by listing each observed site pattern only once in <code>dna_dat</code>. <code>dna_weights</code> is a vector of weights, containing the number of times that each site pattern was observed.</p><h3><a class="nav-anchor" id="sequence-substitution-models-1" href="#sequence-substitution-models-1">sequence substitution models</a></h3><p>DNA-specific substitution models have 4 states: the 4 nucleotides from <a href="https://github.com/BioJulia/BioSymbols.jl">BioSymbols</a> (listed <a href="http://biojulia.net/BioSymbols.jl/stable/nucleicacids/">here</a>). Each model has a relative and an absolute version.</p><ul><li><code>:JC69</code> Jukes &amp; Cantor 1969 model: one single rate for all transitions. The relative version has values -1 along the diagonal of the rate matrix (1 expected transition / unit of time). The absolute version has an extra parameter to scale the rate matrix.</li><li><code>:HKY85</code> Hasegawa, Kishino &amp; Yano 1985: treats transitions differently from transversions. The relative is scaled to predict an average of 1 transition / unit of time.</li></ul><p>We may allow for rate variation across sites using the <code>:RV</code> option.</p><h3><a class="nav-anchor" id="likelihood-of-a-fixed-network-1" href="#likelihood-of-a-fixed-network-1">likelihood of a fixed network</a></h3><p>In the examples below, none of the rate parameters are optimized, so we get to see the default starting values.</p><pre><code class="language-julia-repl">julia&gt; d1 = fitdiscrete(dna_net, :JC69, dna_dat, dna_weights, :RV; optimizeQ=false, optimizeRVAS=false)
┌ Warning: the network contains taxa with no data: those will be pruned
└ @ PhyloNetworks ~/build/crsl4/PhyloNetworks.jl/src/traitsLikDiscrete.jl:681
┌ Warning: resetting edge numbers to be from 1 to 53
└ @ PhyloNetworks ~/build/crsl4/PhyloNetworks.jl/src/manipulateNet.jl:919
PhyloNetworks.StatisticalSubstitutionModel:
Jukes and Cantor 69 Substitution Model,
relative rate version
off-diagonal rates equal to 1/3
rate matrix Q:
               A       C       G       T
       A       *  0.3333  0.3333  0.3333
       C  0.3333       *  0.3333  0.3333
       G  0.3333  0.3333       *  0.3333
       T  0.3333  0.3333  0.3333       *
209 traits, 22 species
variable rates across sites ~ discretized gamma with
 alpha=1.0
 4 categories
 rate multipliers: [0.14578, 0.51313, 1.07083, 2.27025]
on a network with 3 reticulations
log-likelihood: -12792.52577

julia&gt; d2 = fitdiscrete(dna_net, :HKY85, dna_dat, dna_weights, :RV; optimizeQ=false, optimizeRVAS=false)
┌ Warning: the network contains taxa with no data: those will be pruned
└ @ PhyloNetworks ~/build/crsl4/PhyloNetworks.jl/src/traitsLikDiscrete.jl:681
┌ Warning: resetting edge numbers to be from 1 to 53
└ @ PhyloNetworks ~/build/crsl4/PhyloNetworks.jl/src/manipulateNet.jl:919
PhyloNetworks.StatisticalSubstitutionModel:
HKY85 Substitution Model base frequencies: [0.29100263035741913, 0.2682964567538295, 0.23638403218319665, 0.2043168807055547]
relative rate version with transition/tranversion ratio kappa = 1.0,
 scaled so that there is one substitution per unit time
rate matrix Q:
               A       C       G       T
       A       *  0.3598  0.3170  0.2740
       C  0.3902       *  0.3170  0.2740
       G  0.3902  0.3598       *  0.2740
       T  0.3902  0.3598  0.3170       *
209 traits, 22 species
variable rates across sites ~ discretized gamma with
 alpha=1.0
 4 categories
 rate multipliers: [0.14578, 0.51313, 1.07083, 2.27025]
on a network with 3 reticulations
log-likelihood: -12767.36686</code></pre><p>When allowing for rate variation across sites, the default α is 1.</p><p>In the more interesting examples below, we optimize the evolutionary rates and the way rates vary across sites (which is the default).</p><pre><code class="language-julia-repl">julia&gt; d3 = fitdiscrete(dna_net, :JC69, dna_dat, dna_weights, :RV; ftolAbs=0.1, xtolAbs=0.01)
┌ Warning: the network contains taxa with no data: those will be pruned
└ @ PhyloNetworks ~/build/crsl4/PhyloNetworks.jl/src/traitsLikDiscrete.jl:681
┌ Warning: resetting edge numbers to be from 1 to 53
└ @ PhyloNetworks ~/build/crsl4/PhyloNetworks.jl/src/manipulateNet.jl:919
PhyloNetworks.StatisticalSubstitutionModel:
Jukes and Cantor 69 Substitution Model,
relative rate version
off-diagonal rates equal to 1/3
rate matrix Q:
               A       C       G       T
       A       *  0.3333  0.3333  0.3333
       C  0.3333       *  0.3333  0.3333
       G  0.3333  0.3333       *  0.3333
       T  0.3333  0.3333  0.3333       *
209 traits, 22 species
variable rates across sites ~ discretized gamma with
 alpha=0.06860668132448229
 4 categories
 rate multipliers: [0.0, 2.0e-5, 0.02709, 3.9729]
on a network with 3 reticulations
log-likelihood: -2214.09485</code></pre><p>Lenient tolerance parameters <code>ftolAbs</code> etc. have been chosen here to make this example faster. Note that the fitted object contains a separate version of the input network, where any taxon without data has been pruned, and where branch length numbers may have been modified.</p><pre><code class="language-julia-repl">julia&gt; d3.net
HybridNetwork, Rooted Network
53 edges
51 nodes: 22 tips, 3 hybrid nodes, 26 internal tree nodes.
tip labels: Ae_caudata_Tr275, Ae_caudata_Tr139, Ae_longissima_Tr241, Ae_longissima_Tr242, ...
(H_vulgare_HVens23:1.0,((((Ae_speltoides_Tr251:2.0):1.0,#H4:1.0::0.4):1.0,((((((Ae_caudata_Tr139:1.0,Ae_caudata_Tr275:2.0):1.0)#H1:1.0::0.6,((((((Ae_longissima_Tr241:1.0,Ae_longissima_Tr242:1.0):1.0,Ae_longissima_Tr355:1.0):1.0,Ae_sharonensis_Tr265:2.0):1.0,((Ae_bicornis_Tr408:1.0,Ae_bicornis_Tr407:1.0):1.0,Ae_bicornis_Tr406:1.0):1.0):1.0,(Ae_searsii_Tr164:1.0,Ae_searsii_Tr165:1.0):2.0):1.0)#H2:1.0::0.6):1.0,#H1:2.0::0.4):1.0,((Ae_comosa_Tr271:1.0,Ae_comosa_Tr272:1.0):1.0,((Ae_uniaristata_Tr403:1.0,Ae_uniaristata_Tr357:1.0):1.0,Ae_uniaristata_Tr402:1.0):2.0):1.0):1.0,(((Ae_tauschii_Tr352:1.0,Ae_tauschii_Tr351:1.0):1.0,Ae_tauschii_Tr125:2.0):1.0,(#H2:1.0::0.4,(Ae_mutica_Tr237:4.0)#H4:1.0::0.6):1.0):1.0):2.0):1.0):4.0);</code></pre><footer><hr/></footer></article></body></html>
