<!DOCTYPE html>
<!--[if lt IE 7 ]><html class="no-js ie6"><![endif]-->
<!--[if IE 7 ]><html class="no-js ie7"><![endif]-->
<!--[if IE 8 ]><html class="no-js ie8"><![endif]-->
<!--[if IE 9 ]><html class="no-js ie9"><![endif]-->
<!--[if (gt IE 9)|!(IE)]><!--> <html class="no-js"> <!--<![endif]-->
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,user-scalable=no,initial-scale=1,maximum-scale=1">
    
      
        <title>PhyloNetworks.jl - PhyloNetworks is a Julia package for the manipulation, visualization and inference of phylogenetic networks.</title>
      
      
        <meta name="description" content="PhyloNetworks is a Julia package for the manipulation, visualization and inference of phylogenetic networks.">
      
      
      
        <meta name="author" content="Claudia Sol&iacute;s-Lemus">
      
    
    <meta property="og:url" content="None">
    <meta property="og:title" content="PhyloNetworks.jl">
    <meta property="og:image" content="None/./snaq_small.png">
    <meta name="apple-mobile-web-app-title" content="PhyloNetworks.jl">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
    
      <link rel="apple-touch-icon" href="./snaq_small.png">
    
    
    <link rel="shortcut icon" type="image/x-icon" href="./assets/images/favicon-e565ddfa3b.ico">
    <link rel="icon" type="image/x-icon" href="./assets/images/favicon-e565ddfa3b.ico">
    <style>
      @font-face {
      	font-family: 'Icon';
      	src: url('./assets/fonts/icon.eot?52m981');
      	src: url('./assets/fonts/icon.eot?#iefix52m981')
               format('embedded-opentype'),
      		   url('./assets/fonts/icon.woff?52m981')
               format('woff'),
      		   url('./assets/fonts/icon.ttf?52m981')
               format('truetype'),
      		   url('./assets/fonts/icon.svg?52m981#icon')
               format('svg');
      	font-weight: normal;
      	font-style: normal;
      }
    </style>
    <link rel="stylesheet" href="./assets/stylesheets/application-a422ff04cc.css">
    
      <link rel="stylesheet" href="./assets/stylesheets/palettes-05ab2406df.css">
    
    
      
      
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Ubuntu:400,700|Ubuntu+Mono">
      <style>
        body, input {
          font-family: 'Ubuntu', Helvetica, Arial, sans-serif;
        }
        pre, code {
          font-family: 'Ubuntu Mono', 'Courier New', 'Courier', monospace;
        }
      </style>
    
    
      <link rel="stylesheet" href="./assets/Documenter.css">
    
    <script src="./assets/javascripts/modernizr-4ab42b99fd.js"></script>
    
  </head>
  
  
  
  <body class="palette-primary-teal palette-accent-teal">
    
      
      
    
    <div class="backdrop">
      <div class="backdrop-paper"></div>
    </div>
    <input class="toggle" type="checkbox" id="toggle-drawer">
    <input class="toggle" type="checkbox" id="toggle-search">
    <label class="toggle-button overlay" for="toggle-drawer"></label>
    <header class="header">
      <nav aria-label="Header">
  <div class="bar default">
    <div class="button button-menu" role="button" aria-label="Menu">
      <label class="toggle-button icon icon-menu" for="toggle-drawer">
        <span></span>
      </label>
    </div>
    <div class="stretch">
      <div class="title">
        
          <span class="path">
            
          </span>
        
        PhyloNetworks.jl
      </div>
    </div>
    
    
    <div class="button button-search" role="button" aria-label="Search">
      <label class="toggle-button icon icon-search" title="Search" for="toggle-search"></label>
    </div>
  </div>
  <div class="bar search">
    <div class="button button-close" role="button" aria-label="Close">
      <label class="toggle-button icon icon-back" for="toggle-search"></label>
    </div>
    <div class="stretch">
      <div class="field">
        <input class="query" type="text" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck>
      </div>
    </div>
    <div class="button button-reset" role="button" aria-label="Search">
      <button class="toggle-button icon icon-close" id="reset-search"></button>
    </div>
  </div>
</nav>
    </header>
    <main class="main">
      
      <div class="drawer">
        <nav aria-label="Navigation">
  
  <a href="https://github.com/crsl4/PhyloNetworks.jl" class="project">
    <div class="banner">
      
        <div class="logo">
          <img src="./snaq_small.png">
        </div>
      
      <div class="name">
        <strong>
          PhyloNetworks.jl
          <span class="version">
            
          </span>
        </strong>
        
          <br>
          crsl4/PhyloNetworks.jl
        
      </div>
    </div>
  </a>
  <div class="scrollable">
    <div class="wrapper">
      
        <ul class="repo">
          <li class="repo-download">
            
            <a href="https://github.com/crsl4/PhyloNetworks.jl/archive/master.zip" target="_blank" title="Download" data-action="download">
              <i class="icon icon-download"></i> Download
            </a>
          </li>
          <li class="repo-stars">
            <a href="https://github.com/crsl4/PhyloNetworks.jl/stargazers" target="_blank" title="Stargazers" data-action="star">
              <i class="icon icon-star"></i> Stars
              <span class="count">&ndash;</span>
            </a>
          </li>
        </ul>
        <hr>
      
      <div class="toc">
        <ul>
          
            
  <li>
    <a class="current" title="Home" href=".">
      Home
    </a>
    
      
        
      
      
        <ul>
          
            <li class="anchor">
              <a title="How to get help" href="#how-to-get-help">
                How to get help
              </a>
            </li>
          
            <li class="anchor">
              <a title="Manual Outline" href="#manual-outline">
                Manual Outline
              </a>
            </li>
          
            <li class="anchor">
              <a title="Library Outline" href="#library-outline">
                Library Outline
              </a>
            </li>
          
            <li class="anchor">
              <a title="Index" href="#index">
                Index
              </a>
            </li>
          
        </ul>
      
    
  </li>

          
            
  <li>
    <span class="section">Manual</span>
    <ul>
      
        
  <li>
    <a class="" title="Installation" href="man/installation/">
      Installation
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Input Data for SNaQ" href="man/inputdata/">
      Input Data for SNaQ
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="TICR pipeline" href="man/ticr_howtogetQuartetCFs/">
      TICR pipeline
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Network estimation and display" href="man/snaq_plot/">
      Network estimation and display
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Fixed Network" href="man/fixednetworkoptim/">
      Fixed Network
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Extract Expected CFs" href="man/expectedCFs/">
      Extract Expected CFs
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Bootstrap" href="man/bootstrap/">
      Bootstrap
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Multiple Alleles" href="man/multiplealleles/">
      Multiple Alleles
    </a>
    
  </li>

      
    </ul>
  </li>

          
            
  <li>
    <span class="section">Library</span>
    <ul>
      
        
  <li>
    <a class="" title="Public" href="lib/public/">
      Public
    </a>
    
  </li>

      
        
  <li>
    <a class="" title="Internals" href="lib/internals/">
      Internals
    </a>
    
  </li>

      
    </ul>
  </li>

          
        </ul>
        
      </div>
    </div>
  </div>
</nav>
      </div>
      <article class="article">
        <div class="wrapper">
          
          <p><a id='PhyloNetworks.jl-1'></a></p>
<h1 id="phylonetworksjl">PhyloNetworks.jl</h1>
<p>PhyloNetworks is a Julia package for the manipulation, visualization and inference of phylogenetic networks.  SNaQ implements the statistical inference method in <a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1005896">Sol&iacute;s-Lemus and An&eacute; 2016</a>. The procedure involves a numerical optimization of branch lengths and inheritance probabilities and a heuristic search in the space of phylogenetic networks.</p>
<p><a id='How-to-get-help-1'></a></p>
<h2 id="how-to-get-help">How to get help</h2>
<ul>
<li>the package <a href="https://github.com/crsl4/PhyloNetworks.jl/wiki">wiki</a> has a step-by-step   tutorial, done for the Evolution 2016 conference, with background on networks and   explanations.</li>
<li>the <a href="https://groups.google.com/forum/#!forum/phylonetworks-users">google group</a>   has answers to common questions.</li>
<li>the Manual below has a quick tutorial on SNaQ and PhyloNetworks.</li>
<li>the <a href=".#main-index-1">Index</a> further below has the full list of documented functions.</li>
</ul>
<p><a id='Manual-Outline-1'></a></p>
<h2 id="manual-outline">Manual Outline</h2>
<ul>
<li><a href="man/installation/#Installation-1">Installation</a><ul>
<li><a href="man/installation/#Installation-of-Julia-1">Installation of Julia</a></li>
<li><a href="man/installation/#Installation-of-the-package-PhyloNetworks-1">Installation of the package PhyloNetworks</a></li>
<li><a href="man/installation/#Test-example-1">Test example</a></li>
<li><a href="man/installation/#Julia-types-1">Julia types</a></li>
</ul>
</li>
<li><a href="man/inputdata/#Input-for-SNaQ-1">Input for SNaQ</a><ul>
<li><a href="man/inputdata/#Tutorial-data:-gene-trees-1">Tutorial data: gene trees</a></li>
<li><a href="man/inputdata/#Tutorial-data:-quartet-CFs-1">Tutorial data: quartet CFs</a></li>
</ul>
</li>
<li><a href="man/ticr_howtogetQuartetCFs/#TICR-pipeline-1">TICR pipeline</a><ul>
<li><a href="man/ticr_howtogetQuartetCFs/#To-run-MrBayes:-You-already-have-alignments-1">To run MrBayes: You already have alignments</a></li>
<li><a href="man/ticr_howtogetQuartetCFs/#To-run-BUCKy:-You-already-have-MrBayes-output-1">To run BUCKy: You already have MrBayes output</a></li>
</ul>
</li>
<li><a href="man/snaq_plot/#Getting-a-Network-1">Getting a Network</a><ul>
<li><a href="man/snaq_plot/#Network-Estimation-1">Network Estimation</a></li>
<li><a href="man/snaq_plot/#Network-Visualization-1">Network Visualization</a></li>
<li><a href="man/snaq_plot/#Re-rooting-networks-1">Re-rooting networks</a></li>
<li><a href="man/snaq_plot/#Candidate-Network-Evaluation-1">Candidate Network Evaluation</a></li>
</ul>
</li>
<li><a href="man/fixednetworkoptim/#Fixed-Network-1">Fixed Network</a><ul>
<li><a href="man/fixednetworkoptim/#Optimizing-branch-lengths-and-inheritance-probabilities-for-a-given-network-1">Optimizing branch lengths and inheritance probabilities for a given network</a></li>
<li><a href="man/fixednetworkoptim/#Network-Score-with-no-optimization-1">Network Score with no optimization</a></li>
</ul>
</li>
<li><a href="man/expectedCFs/#Extract-Expected-CFs-1">Extract Expected CFs</a></li>
<li><a href="man/bootstrap/#Bootstrap-1">Bootstrap</a><ul>
<li><a href="man/bootstrap/#Running-a-bootstrap-analysis-1">Running a bootstrap analysis</a></li>
<li><a href="man/bootstrap/#Summarizing-bootstrap-on-the-main-tree-1">Summarizing bootstrap on the main tree</a></li>
<li><a href="man/bootstrap/#Summarizing-bootstrap-on-reticulations-1">Summarizing bootstrap on reticulations</a></li>
</ul>
</li>
<li><a href="man/multiplealleles/#Multiple-alleles-per-species-1">Multiple alleles per species</a></li>
</ul>
<p><a id='Library-Outline-1'></a></p>
<h2 id="library-outline">Library Outline</h2>
<ul>
<li><a href="lib/public/#Public-Documentation-1">Public Documentation</a><ul>
<li><a href="lib/public/#Contents-1">Contents</a></li>
<li><a href="lib/public/#Index-1">Index</a></li>
<li><a href="lib/public/#Types-1">Types</a></li>
<li><a href="lib/public/#Utilities-1">Utilities</a></li>
<li><a href="lib/public/#Data-and-Topology-read/write-1">Data and Topology read/write</a></li>
<li><a href="lib/public/#Network-inference-1">Network inference</a></li>
<li><a href="lib/public/#Network-Comparisons-1">Network Comparisons</a></li>
</ul>
</li>
<li><a href="lib/internals/#Internal-Documentation-1">Internal Documentation</a><ul>
<li><a href="lib/internals/#Contents-1">Contents</a></li>
<li><a href="lib/internals/#Index-1">Index</a></li>
<li><a href="lib/internals/#Types-1">Types</a></li>
<li><a href="lib/internals/#Functions-and-methods-1">Functions and methods</a></li>
</ul>
</li>
</ul>
<p><a id='main-index-1'></a></p>
<h2 id="index">Index</h2>
<p><a id='Functions-1'></a></p>
<h3 id="functions">Functions</h3>
<ul>
<li><a href="lib/public/#Gadfly.plot"><code>Gadfly.plot</code></a></li>
<li><a href="lib/public/#PhyloNetworks.bootsnaq"><code>PhyloNetworks.bootsnaq</code></a></li>
<li><a href="lib/public/#PhyloNetworks.cladewiseorder!"><code>PhyloNetworks.cladewiseorder!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.deleteHybridThreshold!"><code>PhyloNetworks.deleteHybridThreshold!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.deleteleaf!"><code>PhyloNetworks.deleteleaf!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.directEdges!"><code>PhyloNetworks.directEdges!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.displayedNetworkAt!"><code>PhyloNetworks.displayedNetworkAt!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.displayedTrees"><code>PhyloNetworks.displayedTrees</code></a></li>
<li><a href="lib/public/#PhyloNetworks.fittedQuartetCF"><code>PhyloNetworks.fittedQuartetCF</code></a></li>
<li><a href="lib/public/#PhyloNetworks.hardwiredCluster"><code>PhyloNetworks.hardwiredCluster</code></a></li>
<li><a href="lib/public/#PhyloNetworks.hardwiredClusterDistance"><code>PhyloNetworks.hardwiredClusterDistance</code></a></li>
<li><a href="lib/public/#PhyloNetworks.hardwiredClusters"><code>PhyloNetworks.hardwiredClusters</code></a></li>
<li><a href="lib/public/#PhyloNetworks.hybridBootstrapSupport"><code>PhyloNetworks.hybridBootstrapSupport</code></a></li>
<li><a href="lib/public/#PhyloNetworks.hybridatnode!"><code>PhyloNetworks.hybridatnode!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.majorTree"><code>PhyloNetworks.majorTree</code></a></li>
<li><a href="lib/public/#PhyloNetworks.mapAllelesCFtable"><code>PhyloNetworks.mapAllelesCFtable</code></a></li>
<li><a href="lib/public/#PhyloNetworks.minorTreeAt"><code>PhyloNetworks.minorTreeAt</code></a></li>
<li><a href="lib/public/#PhyloNetworks.preorder!"><code>PhyloNetworks.preorder!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.printEdges"><code>PhyloNetworks.printEdges</code></a></li>
<li><a href="lib/public/#PhyloNetworks.printNodes"><code>PhyloNetworks.printNodes</code></a></li>
<li><a href="lib/public/#PhyloNetworks.readBootstrapTrees"><code>PhyloNetworks.readBootstrapTrees</code></a></li>
<li><a href="lib/public/#PhyloNetworks.readInputTrees"><code>PhyloNetworks.readInputTrees</code></a></li>
<li><a href="lib/public/#PhyloNetworks.readMultiTopology"><code>PhyloNetworks.readMultiTopology</code></a></li>
<li><a href="lib/public/#PhyloNetworks.readSnaqNetwork"><code>PhyloNetworks.readSnaqNetwork</code></a></li>
<li><a href="lib/public/#PhyloNetworks.readTableCF"><code>PhyloNetworks.readTableCF</code></a></li>
<li><a href="lib/public/#PhyloNetworks.readTableCF!"><code>PhyloNetworks.readTableCF!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.readTopology"><code>PhyloNetworks.readTopology</code></a></li>
<li><a href="lib/public/#PhyloNetworks.readTopologyLevel1"><code>PhyloNetworks.readTopologyLevel1</code></a></li>
<li><a href="lib/public/#PhyloNetworks.readTrees2CF"><code>PhyloNetworks.readTrees2CF</code></a></li>
<li><a href="lib/public/#PhyloNetworks.rootatnode!"><code>PhyloNetworks.rootatnode!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.rootonedge!"><code>PhyloNetworks.rootonedge!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.rotate!"><code>PhyloNetworks.rotate!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.setGamma!"><code>PhyloNetworks.setGamma!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.setLength!"><code>PhyloNetworks.setLength!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.snaq!"><code>PhyloNetworks.snaq!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.snaqDebug"><code>PhyloNetworks.snaqDebug</code></a></li>
<li><a href="lib/public/#PhyloNetworks.sorttaxa!"><code>PhyloNetworks.sorttaxa!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.summarizeDataCF"><code>PhyloNetworks.summarizeDataCF</code></a></li>
<li><a href="lib/public/#PhyloNetworks.tipLabels"><code>PhyloNetworks.tipLabels</code></a></li>
<li><a href="lib/public/#PhyloNetworks.topologyMaxQPseudolik!"><code>PhyloNetworks.topologyMaxQPseudolik!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.topologyQPseudolik!"><code>PhyloNetworks.topologyQPseudolik!</code></a></li>
<li><a href="lib/public/#PhyloNetworks.treeEdgesBootstrap"><code>PhyloNetworks.treeEdgesBootstrap</code></a></li>
<li><a href="lib/public/#PhyloNetworks.writeMultiTopology"><code>PhyloNetworks.writeMultiTopology</code></a></li>
<li><a href="lib/public/#PhyloNetworks.writeTopology"><code>PhyloNetworks.writeTopology</code></a></li>
<li><a href="lib/internals/#PhyloNetworks.assignhybridnames!"><code>PhyloNetworks.assignhybridnames!</code></a></li>
<li><a href="lib/internals/#PhyloNetworks.deleteNode!"><code>PhyloNetworks.deleteNode!</code></a></li>
<li><a href="lib/internals/#PhyloNetworks.sampleBootstrapTrees"><code>PhyloNetworks.sampleBootstrapTrees</code></a></li>
<li><a href="lib/internals/#PhyloNetworks.sampleCFfromCI"><code>PhyloNetworks.sampleCFfromCI</code></a></li>
<li><a href="lib/internals/#PhyloNetworks.setNonIdBL!"><code>PhyloNetworks.setNonIdBL!</code></a></li>
</ul>
<p><a id='Types-1'></a></p>
<h3 id="types">Types</h3>
<ul>
<li><a href="lib/public/#PhyloNetworks.DataCF"><code>PhyloNetworks.DataCF</code></a></li>
<li><a href="lib/public/#PhyloNetworks.HybridNetwork"><code>PhyloNetworks.HybridNetwork</code></a></li>
<li><a href="lib/public/#PhyloNetworks.Quartet"><code>PhyloNetworks.Quartet</code></a></li>
<li><a href="lib/internals/#PhyloNetworks.ANode"><code>PhyloNetworks.ANode</code></a></li>
</ul>
          <aside class="copyright" role="note">
            
            Documentation built with
            <a href="http://www.mkdocs.org" target="_blank">MkDocs</a>
            using the
            <a href="http://squidfunk.github.io/mkdocs-material/" target="_blank">
              Material
            </a>
            theme.
          </aside>
          
            <footer class="footer">
              
  <nav class="pagination" aria-label="Footer">
    <div class="previous">
      
    </div>
    <div class="next">
      
        <a href="man/installation/" title="Installation">
          <span class="direction">
            Next
          </span>
          <div class="page">
            <div class="stretch">
              <div class="title">
                Installation
              </div>
            </div>
            <div class="button button-next" role="button" aria-label="Next">
              <i class="icon icon-forward"></i>
            </div>
          </div>
        </a>
      
    </div>
  </nav>

            </footer>
          
        </div>
      </article>
      <div class="results" role="status" aria-live="polite">
        <div class="scrollable">
          <div class="wrapper">
            <div class="meta"></div>
            <div class="list"></div>
          </div>
        </div>
      </div>
    </main>
    <script>
      var base_url = '.';
      var repo_id  = 'crsl4/PhyloNetworks.jl';
    </script>
    <script src="./assets/javascripts/application-997097ee0c.js"></script>
    
      <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
      <script src="./assets/mathjaxhelper.js"></script>
    
    
  </body>
</html>